// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-app-header/lux-app-header-subcomponents/lux-app-header-action-nav/lux-app-header-action-nav-item/lux-app-header-action-nav-item.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var LuxAppHeaderActionNavItemComponent = /** @class */ (function () {
    function LuxAppHeaderActionNavItemComponent() {
        this.luxClicked = new EventEmitter();
    }
    LuxAppHeaderActionNavItemComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxLabel", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxIconName", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxColor", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxDisabled", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxTagId", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxAppHeaderActionNavItemComponent.prototype, "luxClicked", void 0);
    LuxAppHeaderActionNavItemComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-app-header-action-nav-item',
            template: "<lux-button\n  [ngClass]=\"{ 'lux-button-with-label': !!luxLabel }\"\n  [luxLabel]=\"luxLabel\"\n  [luxIconName]=\"luxIconName\"\n  [luxColor]=\"luxColor\"\n  [luxRaised]=\"false\"\n  [luxRounded]=\"false\"\n  [luxDisabled]=\"luxDisabled\"\n  [luxTagId]=\"luxTagId\"\n  (luxClicked)=\"luxClicked.emit($event)\"\n>\n</lux-button>\n",
            styles: ["lux-button ::ng-deep button{padding:0;min-width:30px;max-height:30px!important;min-height:30px!important}lux-button.lux-button-with-label ::ng-deep button{padding:0 4px;max-height:45px!important;min-height:45px!important}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxAppHeaderActionNavItemComponent);
    return LuxAppHeaderActionNavItemComponent;
}());
export { LuxAppHeaderActionNavItemComponent };
