// @ihk-gfi/lux-components/esm5/src/app/modules/lux-action/lux-button/lux-button.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, HostBinding, Input, OnDestroy, OnInit } from '@angular/core';
import { LuxComponentsConfigService } from '../../lux-components-config/lux-components-config.service';
import { LuxActionComponentBaseClass } from '../lux-action-model/lux-action-component-base.class';
var LuxButtonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LuxButtonComponent, _super);
    function LuxButtonComponent(elementRef, componentsConfigService) {
        var _this = _super.call(this) || this;
        _this.elementRef = elementRef;
        _this.componentsConfigService = componentsConfigService;
        _this.iconSize = '2x';
        _this.luxType = 'button';
        return _this;
    }
    LuxButtonComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.configSubscription = this.componentsConfigService.config.subscribe(function () {
            // Hintergrund: LuxLink, LuxSideNavItem und LuxMenuItem benutzen alle unter der Haube
            // den LuxButton. Wenn diese nun als Ausnahmen f端r Uppercase in der Config eingetragen werden,
            // darf eine Uppercase-Einstellung f端r den LuxButton diese nicht 端berschreiben.
            // Deshalb pr端ft der LuxButton hier, ob er Teil einer dieser Komponenten ist.
            _this.detectParent();
        });
    };
    LuxButtonComponent.prototype.ngOnDestroy = function () {
        this.configSubscription.unsubscribe();
    };
    LuxButtonComponent.prototype.clicked = function (event) {
        this.luxClicked.emit(event);
    };
    LuxButtonComponent.prototype.detectParent = function () {
        var className = this.elementRef.nativeElement.className;
        var selector = 'lux-button';
        if (className.indexOf('lux-link') > -1) {
            selector = 'lux-link';
        }
        else if (className.indexOf('lux-side-nav-item') > -1) {
            selector = 'lux-side-nav-item';
        }
        else if (className.indexOf('lux-menu-item') > -1) {
            selector = 'lux-menu-item';
        }
        this.labelUppercase = this.componentsConfigService.isLabelUppercaseForSelector(selector);
    };
    LuxButtonComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: LuxComponentsConfigService }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxButtonComponent.prototype, "luxType", void 0);
    tslib_1.__decorate([
        HostBinding('class.lux-uppercase'),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxButtonComponent.prototype, "labelUppercase", void 0);
    LuxButtonComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-button',
            template: "<button\n  *ngIf=\"luxRaised && !luxRounded\"\n  mat-raised-button\n  [color]=\"luxColor\"\n  (click)=\"clicked($event)\"\n  [disabled]=\"luxDisabled\"\n  [type]=\"luxType\"\n  class=\"lux-button lux-bold lux-align-center\"\n  luxTagIdHandler\n  [luxTagId]=\"luxTagId\"\n  [ngClass]=\"{ 'lux-align-baseline': luxIconAlignWithLabel && !!luxIconName, 'lux-icon-button': !!luxIconName }\"\n  [attr.aria-label]=\"luxLabel ? luxLabel : ''\"\n>\n  <ng-container *ngTemplateOutlet=\"luxIconName ? iconTemplate : noIconTemplate\"></ng-container>\n  <span class=\"lux-button-label\" [ngClass]=\"{ 'lux-ml-3': !!luxIconName }\" *ngIf=\"luxLabel\">\n    {{ luxLabel }}\n  </span>\n</button>\n\n<button\n  *ngIf=\"!luxRaised && !luxRounded\"\n  mat-button\n  [color]=\"luxColor\"\n  (click)=\"clicked($event)\"\n  [disabled]=\"luxDisabled\"\n  [type]=\"luxType\"\n  class=\"lux-button lux-bold lux-align-center\"\n  luxTagIdHandler\n  [luxTagId]=\"luxTagId\"\n  [ngClass]=\"{ 'lux-align-baseline': luxIconAlignWithLabel && !!luxIconName, 'lux-icon-button': !!luxIconName }\"\n  [attr.aria-label]=\"luxLabel ? luxLabel : ''\"\n>\n  <ng-container *ngTemplateOutlet=\"luxIconName ? iconTemplate : noIconTemplate\"></ng-container>\n  <span class=\"lux-button-label\" [ngClass]=\"{ 'lux-ml-3': !!luxIconName }\" *ngIf=\"luxLabel\">\n    {{ luxLabel }}\n  </span>\n</button>\n\n<button\n  *ngIf=\"luxRounded\"\n  mat-fab\n  [color]=\"luxColor\"\n  (click)=\"clicked($event)\"\n  [disabled]=\"luxDisabled\"\n  [type]=\"luxType\"\n  class=\"lux-button lux-bold lux-align-center lux-button-rounded mat-elevation-z1\"\n  luxTagIdHandler\n  [luxTagId]=\"luxTagId\"\n  [ngClass]=\"{ 'lux-align-baseline': luxIconAlignWithLabel && !!luxIconName, 'lux-icon-button': !!luxIconName }\"\n  [attr.aria-label]=\"luxLabel ? luxLabel : ''\"\n>\n  <ng-container *ngTemplateOutlet=\"luxIconName ? iconTemplate : noIconTemplate\"></ng-container>\n  <span class=\"lux-button-label\" [ngClass]=\"{ 'lux-ml-3': !!luxIconName }\" *ngIf=\"luxLabel\">\n    {{ luxLabel }}\n  </span>\n</button>\n<ng-template #iconTemplate>\n  <lux-icon [luxIconName]=\"luxIconName\" [luxIconSize]=\"iconSize\" luxMargin=\"0\" luxPadding=\"0\"></lux-icon>\n</ng-template>\n<ng-template #noIconTemplate></ng-template>\n",
            styles: [".lux-align-baseline{padding-bottom:8px!important}.lux-align-baseline ::ng-deep .mat-button-wrapper{-webkit-box-align:end;-ms-flex-align:end;align-items:flex-end}.lux-button-label{line-height:1}.lux-icon-button{height:45px;padding-top:4px;padding-bottom:2px}.lux-icon-button.mat-fab{width:45px}.lux-icon-button.mat-fab ::ng-deep lux-icon i,.lux-icon-button.mat-fab ::ng-deep lux-icon mat-icon{font-size:1.5em}.lux-icon-button.mat-fab ::ng-deep .mat-button-wrapper{padding:0}:host ::ng-deep .mat-button-wrapper{height:100%;display:-webkit-inline-box;display:-ms-inline-flexbox;display:inline-flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center}:host ::ng-deep lux-icon{line-height:1}:host ::ng-deep lux-iconi,:host ::ng-deep lux-iconmat-icon{line-height:1}:host ::ng-deep lux-iconi{display:block}"]
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, LuxComponentsConfigService])
    ], LuxButtonComponent);
    return LuxButtonComponent;
}(LuxActionComponentBaseClass));
export { LuxButtonComponent };
