// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-file-preview/lux-file-preview-base/lux-file-preview-base.js
import * as tslib_1 from "tslib";
import { ESCAPE } from '@angular/cdk/keycodes';
import { HostListener, Inject } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { LuxUtil } from '../../lux-util/lux-util';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
import { LuxFilePreviewRef } from '../lux-file-preview-ref';
let LuxFilePreviewBase = class LuxFilePreviewBase {
    constructor(previewRef, previewData, sanitizer) {
        this.previewRef = previewRef;
        this.previewData = previewData;
        this.sanitizer = sanitizer;
        this.urls = [];
        this.paddingWith = 100;
        this.paddingHeight = 150;
        this.startPhase = true;
        this.startDurationMs = 250;
        this.loading = true;
        this.downloadIconName = 'fas fa-download';
        this.downloadTagId = 'file-preview-download-btn';
        this.downloadAriaLabel = 'Datei herunterladen';
        this.closeIconName = 'fas fa-window-close';
        this.closeTagId = 'file-preview-close-btn';
        this.closeAriaLabel = 'Dateivorschau schließen';
    }
    handleKeydown(event) {
        if (event.keyCode === ESCAPE) {
            this.onClose();
        }
    }
    windowResize() {
        this.updateWidthAndHeight();
    }
    ngOnInit() {
        this.loadingTimer = setTimeout(() => {
            this.startPhase = false;
        }, this.startDurationMs);
        this.updateWidthAndHeight();
        setTimeout(() => {
            let myBlob;
            if ('string' === typeof this.previewData.fileObject.content) {
                myBlob = new Blob([LuxUtil.base64ToArrayBuffer(this.previewData.fileObject.content.split(',')[1])], {
                    type: this.previewData.fileObject.type
                });
            }
            else {
                myBlob = this.previewData.fileObject.content;
            }
            this.url = window.URL.createObjectURL(myBlob);
            this.urls.push(this.url);
        });
    }
    ngOnDestroy() {
        this.urls.forEach(url => {
            window.URL.revokeObjectURL(url);
        });
    }
    onDownload() {
        this.previewData.fileComponent.downloadFile(this.previewData.fileObject);
        this.previewRef.close();
    }
    onClose() {
        this.previewRef.close();
    }
    loadingFinished() {
        this.loading = false;
    }
    clearFocus() {
        if (document.activeElement instanceof HTMLElement) {
            document.activeElement.blur();
        }
    }
    updateWidthAndHeight() {
        this.width = window.innerWidth - this.paddingWith;
        this.height = window.innerHeight - this.paddingHeight;
    }
};
tslib_1.__decorate([
    HostListener('document:keydown', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [KeyboardEvent]),
    tslib_1.__metadata("design:returntype", void 0)
], LuxFilePreviewBase.prototype, "handleKeydown", null);
tslib_1.__decorate([
    HostListener('window:resize'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], LuxFilePreviewBase.prototype, "windowResize", null);
LuxFilePreviewBase = tslib_1.__decorate([
    tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
    tslib_1.__metadata("design:paramtypes", [LuxFilePreviewRef, Object, DomSanitizer])
], LuxFilePreviewBase);
export { LuxFilePreviewBase };
