// @ihk-gfi/lux-components/esm5/src/app/modules/lux-form/lux-chips/lux-chips-subcomponents/lux-chip-group.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
var LuxChipGroupComponent = /** @class */ (function () {
    function LuxChipGroupComponent() {
        this._luxColor = 'primary';
        this.luxChipClicked = new EventEmitter();
        this.luxChipAdded = new EventEmitter();
        this.luxChipRemoved = new EventEmitter();
        this.luxChipSelected = new EventEmitter();
        this.luxLabels = [];
        this.luxDisabled = false;
        this.luxRemovable = true;
        this.luxSelected = true;
    }
    Object.defineProperty(LuxChipGroupComponent.prototype, "luxColor", {
        get: function () {
            return this._luxColor;
        },
        set: function (color) {
            if (color !== 'primary' && color !== 'accent' && color !== 'warn') {
                color = undefined;
            }
            this._luxColor = color;
        },
        enumerable: true,
        configurable: true
    });
    LuxChipGroupComponent.prototype.ngOnInit = function () { };
    LuxChipGroupComponent.prototype.add = function (label) {
        this.luxLabels.push(label);
        this.luxChipAdded.emit(label);
    };
    LuxChipGroupComponent.prototype.remove = function (index) {
        this.luxChipRemoved.emit(index);
        this.luxLabels.splice(index, 1);
    };
    LuxChipGroupComponent.prototype.click = function (index) {
        this.luxChipClicked.emit(index);
    };
    LuxChipGroupComponent.prototype.select = function (selected, index) {
        this.luxChipSelected.emit({ index: index, selected: selected });
    };
    tslib_1.__decorate([
        ContentChild(TemplateRef, { static: false }),
        tslib_1.__metadata("design:type", TemplateRef)
    ], LuxChipGroupComponent.prototype, "tempRef", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipGroupComponent.prototype, "luxChipClicked", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipGroupComponent.prototype, "luxChipAdded", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipGroupComponent.prototype, "luxChipRemoved", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipGroupComponent.prototype, "luxChipSelected", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], LuxChipGroupComponent.prototype, "luxLabels", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipGroupComponent.prototype, "luxDisabled", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipGroupComponent.prototype, "luxRemovable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipGroupComponent.prototype, "luxSelected", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxChipGroupComponent.prototype, "luxColor", null);
    LuxChipGroupComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-chip-group',
            template: ''
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxChipGroupComponent);
    return LuxChipGroupComponent;
}());
export { LuxChipGroupComponent };
