// @ihk-gfi/lux-components/esm5/src/app/modules/lux-form/lux-chips/lux-chips-subcomponents/lux-chip.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
var LuxChipComponent = /** @class */ (function () {
    function LuxChipComponent() {
        this.removeClicked = false;
        this._luxColor = 'primary';
        this.luxChipRemoved = new EventEmitter();
        this.luxChipClicked = new EventEmitter();
        this.luxChipSelected = new EventEmitter();
        this.luxDisabled = false;
        this.luxRemovable = true;
        this.luxSelected = true;
    }
    Object.defineProperty(LuxChipComponent.prototype, "luxColor", {
        get: function () {
            return this._luxColor;
        },
        set: function (color) {
            if (color !== 'primary' && color !== 'accent' && color !== 'warn') {
                color = undefined;
            }
            this._luxColor = color;
        },
        enumerable: true,
        configurable: true
    });
    LuxChipComponent.prototype.ngOnInit = function () { };
    LuxChipComponent.prototype.remove = function (index) {
        this.luxChipRemoved.emit(index);
        this.removeClicked = true;
    };
    LuxChipComponent.prototype.click = function (index) {
        if (!this.removeClicked) {
            this.luxChipClicked.emit(index);
        }
    };
    LuxChipComponent.prototype.select = function (selected, index) {
        this.luxSelected = selected;
        this.luxChipSelected.emit({ index: index, selected: selected });
    };
    tslib_1.__decorate([
        ViewChild(TemplateRef, { static: true }),
        tslib_1.__metadata("design:type", TemplateRef)
    ], LuxChipComponent.prototype, "templateRef", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipComponent.prototype, "luxChipRemoved", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipComponent.prototype, "luxChipClicked", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", Object)
    ], LuxChipComponent.prototype, "luxChipSelected", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipComponent.prototype, "luxDisabled", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipComponent.prototype, "luxRemovable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxChipComponent.prototype, "luxSelected", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxChipComponent.prototype, "luxColor", null);
    LuxChipComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-chip',
            template: "\n    <ng-template>\n      <ng-content></ng-content>\n    </ng-template>\n  "
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxChipComponent);
    return LuxChipComponent;
}());
export { LuxChipComponent };
