// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-popups/lux-dialog/lux-dialog-preset/lux-dialog-preset.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { LuxDialogRef } from '../lux-dialog-model/lux-dialog-ref.class';
/**
 * Diese Component wird von dem LuxDialogService zur Darstellung via "open" genutzt und nimmt ein Konfig-Objekt
 * vom Typ LuxDialogConfig entgegen.
 */
let LuxDialogPresetComponent = class LuxDialogPresetComponent {
    constructor(dialogRef) {
        this.dialogRef = dialogRef;
    }
    ngOnInit() {
        this.data = this.dialogRef.data;
    }
    /**
     * Schließt den Dialog beim "positiven" Beenden des Dialogs mit dem Wert true.
     */
    onConfirmClick() {
        this.dialogRef.closeDialog(true);
    }
    /**
     * Schließt den Dialog beim "negativen" Beenden des Dialogs mit dem Wert false.
     */
    onDeclineClick() {
        this.dialogRef.closeDialog(false);
    }
};
LuxDialogPresetComponent.ctorParameters = () => [
    { type: LuxDialogRef }
];
LuxDialogPresetComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-dialog-preset',
        template: "<lux-dialog-structure *ngIf=\"data\">\n  <lux-dialog-title>\n    {{ data.title }}\n  </lux-dialog-title>\n  <lux-dialog-content>\n    <ng-container *ngIf=\"data.contentTemplate\">\n      <ng-container *ngTemplateOutlet=\"data.contentTemplate\"></ng-container>\n    </ng-container>\n    <ng-container *ngIf=\"data.content && !data.contentTemplate\">\n      {{ data.content }}\n    </ng-container>\n  </lux-dialog-content>\n  <lux-dialog-actions fxLayout=\"row wrap\" fxLayoutAlign=\"end center\" fxLayoutGap=\"8px grid\">\n    <lux-button\n      class=\"lux-mr-3 lux-dialog-preset-decline\"\n      luxTabIndex=\"0\"\n      [luxPotentialChildren]=\"['button']\"\n      [luxLabel]=\"data.declineAction.label\"\n      [luxColor]=\"data.declineAction.color\"\n      [luxRaised]=\"data.declineAction.raised\"\n      [luxIconName]=\"data.declineAction.iconName\"\n      [luxTagId]=\"data.declineAction.tagId\"\n      [luxDisabled]=\"data.declineAction.disabled\"\n      [luxRounded]=\"data.declineAction.rounded\"\n      (luxClicked)=\"onDeclineClick()\"\n      *ngIf=\"data.declineAction\"\n    >\n    </lux-button>\n    <lux-button\n      class=\"lux-mr-3 lux-dialog-preset-confirm\"\n      luxTabIndex=\"0\"\n      [luxPotentialChildren]=\"['button']\"\n      [luxLabel]=\"data.confirmAction.label\"\n      [luxColor]=\"data.confirmAction.color\"\n      [luxRaised]=\"data.confirmAction.raised\"\n      [luxIconName]=\"data.confirmAction.iconName\"\n      [luxTagId]=\"data.confirmAction.tagId\"\n      [luxDisabled]=\"data.confirmAction.disabled\"\n      [luxRounded]=\"data.confirmAction.rounded\"\n      (luxClicked)=\"onConfirmClick()\"\n      *ngIf=\"data.confirmAction\"\n    >\n    </lux-button>\n  </lux-dialog-actions>\n</lux-dialog-structure>\n",
        styles: [""]
    }),
    tslib_1.__metadata("design:paramtypes", [LuxDialogRef])
], LuxDialogPresetComponent);
export { LuxDialogPresetComponent };
