// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-popups/lux-snackbar/lux-snackbar.service.js
import * as tslib_1 from "tslib";
var LuxSnackbarService_1;
import { Injectable } from '@angular/core';
import { MatSnackBar, MatSnackBarDismiss } from '@angular/material';
import { LuxSnackbarComponent } from './lux-snackbar-component/lux-snackbar.component';
/**
 * Über den LuxSnackbarService können einfach Snackbarinfos angezeigt werden.
 */
let LuxSnackbarService = LuxSnackbarService_1 = class LuxSnackbarService {
    constructor(snackBar) {
        this.snackBar = snackBar;
    }
    /**
     * Diese Methode öffnet eine Snackbar mit einem Text.
     *
     * @param message Ein Text.
     * @param duration Eine Dauer in Msec (z.B. 2000 = 2 Sekunden). Wenn die Snackbar permanent angezeigt werden soll,
     * muss der Wert 0 angegeben werden.
     * @param actionName Eine Bezeichnung (z.B. Wiederherstellen) für die Action. Dieser Parameter ist optional und
     * muss nicht angegeben werden.
     */
    openText(message, duration, actionName) {
        this.snackBar.open(message, actionName, {
            duration: duration,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    }
    /**
     * Diese Methode öffnet eine Snackbar, in der die übergebene Komponente angezeigt wird.
     *
     * @param component Die Komponente, die angezeigt werden soll.
     * @param duration Eine Dauer in Msec (z.B. 2000 = 2 Sekunden). Wenn die Snackbar permanent angezeigt werden soll,
     * muss der Wert 0 angegeben werden.
     * @param data Ein Datenobjekt, das an die Komponente weitergereicht wird. Mit dem Code folgenden Code, können die
     * Daten verwendet werden. constructor(@Inject(MAT_SNACK_BAR_DATA) public data: any).
     */
    openComponent(component, duration = 0, data) {
        this.snackBar.openFromComponent(component, {
            duration: duration,
            data: data,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    }
    /**
     * Oeffnet eine Snackbar anhand der uebergebenen Konfiguration.
     * Ermoeglicht eine genaue Konfiguration der Snackbar.
     * @param config
     */
    open(duration, config) {
        this.snackBar.openFromComponent(LuxSnackbarComponent, {
            duration: duration,
            data: config,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    }
    /**
     * Diese Methode liefert ein Observable zurück, das den Aufrufer benachrichtigt, wenn die Action in der Snackbar
     * geklickt wird.
     */
    onAction() {
        if (this.snackBar._openedSnackBarRef.instance instanceof LuxSnackbarComponent) {
            return this.snackBar._openedSnackBarRef.instance.onAction();
        }
        return this.snackBar._openedSnackBarRef.onAction();
    }
    /**
     * Diese Methode liefert ein Observable zurück, das den Aufrufer benachrichtigt, wenn die Action in der Snackbar
     * geklickt wird.
     */
    afterDismissed() {
        return this.snackBar._openedSnackBarRef.afterDismissed();
    }
    /**
     * Diese Methode blendet die Snackbar aus.
     */
    dismiss() {
        this.snackBar.dismiss();
    }
};
LuxSnackbarService.VERTICAL_POSITION = 'top';
LuxSnackbarService.HORIZONTAL_POSITION = 'end';
LuxSnackbarService.ctorParameters = () => [
    { type: MatSnackBar }
];
LuxSnackbarService = LuxSnackbarService_1 = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [MatSnackBar])
], LuxSnackbarService);
export { LuxSnackbarService };
