// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-stepper/lux-stepper-model/lux-stepper-parent.class.js
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output, QueryList, ViewChild, ViewChildren, ViewContainerRef } from '@angular/core';
/**
 * Parent-Klasse für den LuxStepperVertical und LuxStepperHorizontal, beide Komponenten werden ausschließlich von
 * dem LuxStepper zur Darstellung genutzt.
 */
export class LuxStepperParent {
    constructor() {
        // Diese Outputs werden bei den Klicks auf die Stepper-eigenen Navigations-Buttons augelöst und informieren die
        // LuxStepperComponent
        this.luxFinButtonClicked = new EventEmitter();
        this.luxNextButtonClicked = new EventEmitter();
        this.luxPrevButtonClicked = new EventEmitter();
        // Wird beim Wechsel des Steps (über Header oder Button) aufgerufen
        this.luxStepChanged = new EventEmitter();
        this.luxCheckValidation = new EventEmitter();
        // Wird beim AfterViewInit und bei jeder Aktualisierung der MatStepLabels aufgerufen um die LuxStepperComponent
        // über die aktuellen Elemente informiert zu halten
        this.luxMatStepperLoaded = new EventEmitter();
        this.luxMatStepLabelsLoaded = new EventEmitter();
    }
    ngAfterViewInit() {
        // Sobald die Component initialisiert ist, dem Parent (luxStepper) den
        // MatStepper und die MatStepLabels mitteilen
        this.luxMatStepperLoaded.emit(this.matStepper);
        this.luxMatStepLabelsLoaded.emit(this.matStepLabels.toArray());
        this.matStepLabels.changes.subscribe(() => {
            this.luxMatStepLabelsLoaded.emit(this.matStepLabels.toArray());
        });
    }
}
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxFinButtonClicked", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxNextButtonClicked", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxPrevButtonClicked", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxStepChanged", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxCheckValidation", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxMatStepperLoaded", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxStepperParent.prototype, "luxMatStepLabelsLoaded", void 0);
tslib_1.__decorate([
    ViewChild('stepper', { static: true }),
    tslib_1.__metadata("design:type", Object)
], LuxStepperParent.prototype, "matStepper", void 0);
tslib_1.__decorate([
    ViewChildren('matStepLabel', { read: ViewContainerRef }),
    tslib_1.__metadata("design:type", QueryList)
], LuxStepperParent.prototype, "matStepLabels", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], LuxStepperParent.prototype, "luxStepperConfig", void 0);
