// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-accordion/lux-accordion.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { MatAccordion } from '@angular/material';
import { Subject } from 'rxjs';
var LuxAccordionComponent = /** @class */ (function () {
    function LuxAccordionComponent() {
        this.changed$ = new Subject();
        this.luxMode = 'default';
        this.luxMulti = false;
        this._luxDisabled = false;
        this._luxHideToggle = false;
    }
    Object.defineProperty(LuxAccordionComponent.prototype, "luxDisabled", {
        get: function () {
            return this._luxDisabled;
        },
        set: function (disabled) {
            this._luxDisabled = disabled;
            this.changed$.next('luxDisabled');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxAccordionComponent.prototype, "luxHideToggle", {
        get: function () {
            return this._luxHideToggle;
        },
        set: function (hideToggle) {
            this._luxHideToggle = hideToggle;
            this.changed$.next('luxHideToggle');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxAccordionComponent.prototype, "luxExpandedHeaderHeight", {
        get: function () {
            return this._luxExpandedHeaderHeight;
        },
        set: function (height) {
            this._luxExpandedHeaderHeight = height;
            this.changed$.next('luxExpandedHeaderHeight');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxAccordionComponent.prototype, "luxCollapsedHeaderHeight", {
        get: function () {
            return this._luxCollapsedHeaderHeight;
        },
        set: function (height) {
            this._luxCollapsedHeaderHeight = height;
            this.changed$.next('luxCollapsedHeaderHeight');
        },
        enumerable: true,
        configurable: true
    });
    LuxAccordionComponent.prototype.ngOnInit = function () { };
    LuxAccordionComponent.prototype.ngOnDestroy = function () {
        this.changed$.complete();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxAccordionComponent.prototype, "luxMode", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxAccordionComponent.prototype, "luxMulti", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxAccordionComponent.prototype, "luxDisabled", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxAccordionComponent.prototype, "luxHideToggle", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxAccordionComponent.prototype, "luxExpandedHeaderHeight", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxAccordionComponent.prototype, "luxCollapsedHeaderHeight", null);
    tslib_1.__decorate([
        ViewChild(MatAccordion, { static: true }),
        tslib_1.__metadata("design:type", MatAccordion)
    ], LuxAccordionComponent.prototype, "matAccordion", void 0);
    LuxAccordionComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-accordion',
            template: "<mat-accordion [displayMode]=\"luxMode\" [multi]=\"luxMulti\">\n  <ng-content select=\"lux-panel\"></ng-content>\n</mat-accordion>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxAccordionComponent);
    return LuxAccordionComponent;
}());
export { LuxAccordionComponent };
