// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-error/lux-error-page/lux-error-page.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { LuxErrorStoreService } from './lux-error-services/lux-error-store.service';
let LuxErrorPageComponent = class LuxErrorPageComponent {
    constructor(router, errorStore) {
        this.router = router;
        this.errorStore = errorStore;
    }
    get error() {
        return this.errorStore.error;
    }
    get errorConfig() {
        return this.errorStore.config;
    }
    ngOnInit() { }
    /**
     * Navigiert ueber den Router zum eingetragenen Home-Pfad
     */
    clickHomeRedirect() {
        this.router.navigate([this.errorConfig.homeRedirectUrl]);
    }
};
LuxErrorPageComponent.ctorParameters = () => [
    { type: Router },
    { type: LuxErrorStoreService }
];
LuxErrorPageComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-error-page',
        template: "<lux-card ngClass.gt-sm=\"lux-error-page-desktop\" id=\"ErrorPage\" fxFlexFill>\n  <lux-card-content fxFlex=\"noshrink\" fxLayout=\"column\" fxLayoutAlign=\"center center\">\n    <lux-icon\n      [luxIconName]=\"errorConfig.iconName\"\n      [luxIconSize]=\"errorConfig.iconSize\"\n      class=\"lux-semi-transparent lux-black-semi-transparent\"\n      fxFlex=\"0 0 auto\"\n      ngClass.gt-sm=\"lux-error-icon\"\n      ngClass.sm=\"lux-error-icon-sm\"\n      ngClass.xs=\"lux-error-icon-xs\"\n    ></lux-icon>\n    <h2 class=\"lux-ml-3 lux-mr-3\">{{ errorConfig.errorText }}</h2>\n    <lux-button\n      luxColor=\"primary\"\n      [luxLabel]=\"errorConfig.homeRedirectText\"\n      (luxClicked)=\"clickHomeRedirect()\"\n      fxFlex=\"0 0 auto\"\n    >\n    </lux-button>\n    <ng-container *ngIf=\"error\">\n      <p class=\"lux-semi-transparent lux-black-semi-transparent\" fxFlex=\"0 0 auto\" *ngIf=\"error.errorId\">\n        Fehlernummer: {{ error.errorId }}\n      </p>\n      <lux-panel *ngIf=\"error.errorMessage\">\n        <lux-panel-header-title>Fehlermeldung</lux-panel-header-title>\n        <lux-panel-content>\n          {{ error.errorMessage }}\n        </lux-panel-content>\n      </lux-panel>\n    </ng-container>\n  </lux-card-content>\n</lux-card>\n",
        styles: [":host{display:-webkit-box;display:-ms-flexbox;display:flex;max-height:100%;height:100%}.lux-error-icon{font-size:5em}.lux-error-icon-sm{font-size:4em}.lux-error-icon-xs{font-size:3em}.lux-error-page-container{border:0 solid transparent}lux-card.lux-error-page-desktop ::ng-deep h2{margin-left:50px;margin-right:0}lux-card.lux-error-page-desktop ::ng-deep lux-card-content>*{max-width:600px}lux-card.lux-error-page-desktop ::ng-deep mat-card{height:100%;margin:0!important}lux-card.lux-error-page-desktop ::ng-deep mat-card>.lux-card-content{-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:center;-ms-flex-pack:center;justify-content:center}lux-card:not(.lux-error-page-desktop) ::ng-deep mat-card{box-shadow:none}lux-card:not(.lux-error-page-desktop) ::ng-deep lux-card-content>*{max-width:100%}"]
    }),
    tslib_1.__metadata("design:paramtypes", [Router, LuxErrorStoreService])
], LuxErrorPageComponent);
export { LuxErrorPageComponent };
