// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-tooltip/lux-tooltip.directive.js
import * as tslib_1 from "tslib";
import { AriaDescriber, FocusMonitor } from '@angular/cdk/a11y';
import { Directionality } from '@angular/cdk/bidi';
import { Overlay } from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { ScrollDispatcher } from '@angular/cdk/scrolling';
import { Directive, ElementRef, HostListener, Inject, Input, NgZone, OnChanges, Optional, SimpleChanges, ViewContainerRef } from '@angular/core';
import { MAT_TOOLTIP_DEFAULT_OPTIONS, MAT_TOOLTIP_SCROLL_STRATEGY, MatTooltip } from '@angular/material';
let LuxTooltipDirective = class LuxTooltipDirective extends MatTooltip {
    constructor(luxOverlay, luxElementRef, luxScrollDispatcher, luxViewContainerRef, luxNgZone, luxPlatform, luxAriaDescriber, luxFocusMonitor, luxScrollStrategy, luxDir, luxDefaultOptions) {
        super(luxOverlay, luxElementRef, luxScrollDispatcher, luxViewContainerRef, luxNgZone, luxPlatform, luxAriaDescriber, luxFocusMonitor, luxScrollStrategy, luxDir, luxDefaultOptions);
        this.luxOverlay = luxOverlay;
        this.luxElementRef = luxElementRef;
        this.luxScrollDispatcher = luxScrollDispatcher;
        this.luxViewContainerRef = luxViewContainerRef;
        this.luxNgZone = luxNgZone;
        this.luxPlatform = luxPlatform;
        this.luxAriaDescriber = luxAriaDescriber;
        this.luxFocusMonitor = luxFocusMonitor;
        this.luxScrollStrategy = luxScrollStrategy;
        this.luxDir = luxDir;
        this.luxDefaultOptions = luxDefaultOptions;
        this.luxTooltip = '???';
        this.luxTooltipHideDelay = 0;
        this.luxTooltipShowDelay = 0;
        this.luxTooltipPosition = 'above';
        this.luxTooltipDisabled = false;
    }
    _handleLongPress() {
        super.show(this.luxTooltipShowDelay);
    }
    _handleKeydown(event) {
        super._handleKeydown(event);
    }
    _handleTouchend() {
        super._handleTouchend();
    }
    ngOnChanges(simpleChanges) {
        this.message = this.luxTooltip;
        this.hideDelay = this.luxTooltipHideDelay;
        this.showDelay = this.luxTooltipShowDelay;
        this.position = this.luxTooltipPosition;
        this.disabled = this.luxTooltipDisabled;
    }
};
LuxTooltipDirective.ctorParameters = () => [
    { type: Overlay },
    { type: ElementRef },
    { type: ScrollDispatcher },
    { type: ViewContainerRef },
    { type: NgZone },
    { type: Platform },
    { type: AriaDescriber },
    { type: FocusMonitor },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_TOOLTIP_SCROLL_STRATEGY,] }] },
    { type: Directionality, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_TOOLTIP_DEFAULT_OPTIONS,] }] }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxTooltipDirective.prototype, "luxTooltip", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxTooltipDirective.prototype, "luxTooltipHideDelay", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxTooltipDirective.prototype, "luxTooltipShowDelay", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxTooltipDirective.prototype, "luxTooltipPosition", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxTooltipDirective.prototype, "luxTooltipDisabled", void 0);
tslib_1.__decorate([
    HostListener('longpress'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], LuxTooltipDirective.prototype, "_handleLongPress", null);
tslib_1.__decorate([
    HostListener('keydown', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], LuxTooltipDirective.prototype, "_handleKeydown", null);
tslib_1.__decorate([
    HostListener('touchend'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], LuxTooltipDirective.prototype, "_handleTouchend", null);
LuxTooltipDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxTooltip]',
        exportAs: 'luxTooltip'
    }),
    tslib_1.__param(8, Inject(MAT_TOOLTIP_SCROLL_STRATEGY)),
    tslib_1.__param(9, Optional()),
    tslib_1.__param(10, Optional()),
    tslib_1.__param(10, Inject(MAT_TOOLTIP_DEFAULT_OPTIONS)),
    tslib_1.__metadata("design:paramtypes", [Overlay,
        ElementRef,
        ScrollDispatcher,
        ViewContainerRef,
        NgZone,
        Platform,
        AriaDescriber,
        FocusMonitor, Object, Directionality, Object])
], LuxTooltipDirective);
export { LuxTooltipDirective };
