// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-form/lux-input/lux-input.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ContentChild, ElementRef, Input, Optional, ViewChild } from '@angular/core';
import { ControlContainer } from '@angular/forms';
import { LuxInputPrefixComponent } from './lux-input-subcomponents/lux-input-prefix.component';
import { LuxInputSuffixComponent } from './lux-input-subcomponents/lux-input-suffix.component';
import { LuxConsoleService } from '../../lux-util/lux-console.service';
import { LuxFormInputBaseClass } from '../lux-form-model/lux-form-input-base.class';
import { LuxComponentsConfigService } from '../../lux-components-config/lux-components-config.service';
let LuxInputComponent = class LuxInputComponent extends LuxFormInputBaseClass {
    constructor(controlContainer, cdr, logger, config) {
        super(controlContainer, cdr, logger, config);
        this.symbolRegExp = /[,.]/;
        this.luxType = 'text';
        this.luxNumberAlignLeft = false;
    }
    /**
     * Wird bei jedem Tastendruck auf dem Inputfeld aufgerufen.
     * @param $event
     */
    onKeyDown($event) {
        // Soll nur für number-Inputs greifen
        if (this.inputElement) {
            const value = this.inputElement.nativeElement.value;
            // Doppelte Punkt-/Komma-Setzung und E's vermeiden
            if (value &&
                this.symbolRegExp.test($event.key) &&
                (this.inputElement.nativeElement.value.match(this.symbolRegExp) || []).length > 0) {
                $event.preventDefault();
            }
        }
    }
};
LuxInputComponent.ctorParameters = () => [
    { type: ControlContainer, decorators: [{ type: Optional }] },
    { type: ChangeDetectorRef },
    { type: LuxConsoleService },
    { type: LuxComponentsConfigService }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxInputComponent.prototype, "luxType", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxInputComponent.prototype, "luxNumberAlignLeft", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxInputComponent.prototype, "luxMaxLength", void 0);
tslib_1.__decorate([
    ContentChild(LuxInputPrefixComponent, { static: false }),
    tslib_1.__metadata("design:type", LuxInputPrefixComponent)
], LuxInputComponent.prototype, "inputPrefix", void 0);
tslib_1.__decorate([
    ContentChild(LuxInputSuffixComponent, { static: false }),
    tslib_1.__metadata("design:type", LuxInputSuffixComponent)
], LuxInputComponent.prototype, "inputSuffix", void 0);
tslib_1.__decorate([
    ViewChild('input', { read: ElementRef, static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], LuxInputComponent.prototype, "inputElement", void 0);
LuxInputComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-input',
        template: "<lux-form-control [luxFormComponent]=\"this\" [formGroup]=\"formGroup\" *ngIf=\"formGroup\">\n  <div class=\"lux-input-row\" fxLayout=\"row\">\n    <span class=\"lux-input-prefix-container\" fxFlex=\"auto\" matPrefix *ngIf=\"inputPrefix\">\n      <ng-content select=\"lux-input-prefix\"></ng-content>\n    </span>\n    <input\n      fxFlex=\"auto\"\n      matInput\n      luxTagIdHandler\n      [luxTagId]=\"luxTagId\"\n      [placeholder]=\"luxPlaceholder\"\n      [type]=\"luxType\"\n      [luxMaxLengthAttr]=\"luxMaxLength\"\n      [formControl]=\"formControl\"\n      [required]=\"luxRequired\"\n      [readonly]=\"luxReadonly\"\n      [autocomplete]=\"luxAutocomplete\"\n      (blur)=\"luxBlur.emit($event)\"\n      (focus)=\"luxFocus.emit($event)\"\n      (focusin)=\"luxFocusIn.emit($event)\"\n      (focusout)=\"luxFocusOut.emit($event)\"\n      [luxAriaDescribedby]=\"uid + '-error ' + uid + '-hint'\"\n      [id]=\"uid\"\n      #input\n      *ngIf=\"luxType !== 'number'\"\n    />\n    <input\n      fxFlex=\"auto\"\n      matInput\n      luxTagIdHandler\n      [luxTagId]=\"luxTagId\"\n      [placeholder]=\"luxPlaceholder\"\n      type=\"number\"\n      [luxMaxLengthAttr]=\"luxMaxLength\"\n      [formControl]=\"formControl\"\n      [required]=\"luxRequired\"\n      [readonly]=\"luxReadonly\"\n      [autocomplete]=\"luxAutocomplete\"\n      [ngClass]=\"{\n        'lux-number ': luxType === 'number',\n        'lux-number-left': luxType !== 'number' || (luxType === 'number' && luxNumberAlignLeft)\n      }\"\n      (blur)=\"luxBlur.emit($event)\"\n      (focus)=\"luxFocus.emit($event)\"\n      (focusin)=\"luxFocusIn.emit($event)\"\n      (focusout)=\"luxFocusOut.emit($event)\"\n      [luxAriaDescribedby]=\"uid + '-error ' + uid + '-hint'\"\n      (keydown)=\"onKeyDown($event)\"\n      [id]=\"uid\"\n      #input\n      *ngIf=\"luxType === 'number'\"\n    />\n    <span class=\"lux-input-suffix-container\" fxFlex=\"auto\" matSuffix *ngIf=\"inputSuffix\">\n      <ng-content select=\"lux-input-suffix\"></ng-content>\n    </span>\n  </div>\n</lux-form-control>\n",
        styles: ["@charset \"UTF-8\";.lux-label{display:block;margin-bottom:2px}input.lux-number{text-align:right!important}input.lux-number-left{text-align:left!important}input.lux-number::-webkit-inner-spin-button,input.lux-number::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}input.lux-number::-ms-clear{display:none}input[type=number]{-moz-appearance:textfield}@-webkit-keyframes autofill{to{color:#666;background:0 0}}input:-webkit-autofill{-webkit-animation-name:autofill;-webkit-animation-fill-mode:both}"]
    }),
    tslib_1.__param(0, Optional()),
    tslib_1.__metadata("design:paramtypes", [ControlContainer,
        ChangeDetectorRef,
        LuxConsoleService,
        LuxComponentsConfigService])
], LuxInputComponent);
export { LuxInputComponent };
