// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-form/lux-textarea/lux-textarea.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, OnInit, Optional } from '@angular/core';
import { ControlContainer } from '@angular/forms';
import { LuxConsoleService } from '../../lux-util/lux-console.service';
import { LuxFormInputBaseClass } from '../lux-form-model/lux-form-input-base.class';
import { LuxComponentsConfigService } from '../../lux-components-config/lux-components-config.service';
let LuxTextareaComponent = class LuxTextareaComponent extends LuxFormInputBaseClass {
    constructor(controlContainer, cdr, logger, config) {
        super(controlContainer, cdr, logger, config);
        this.luxMaxRows = -1;
        this.luxMinRows = 0;
    }
    ngOnInit() {
        super.ngOnInit();
    }
};
LuxTextareaComponent.ctorParameters = () => [
    { type: ControlContainer, decorators: [{ type: Optional }] },
    { type: ChangeDetectorRef },
    { type: LuxConsoleService },
    { type: LuxComponentsConfigService }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxTextareaComponent.prototype, "luxMaxRows", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxTextareaComponent.prototype, "luxMinRows", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxTextareaComponent.prototype, "luxMaxLength", void 0);
LuxTextareaComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-textarea',
        template: "<lux-form-control [luxFormComponent]=\"this\" [luxScalableHeight]=\"true\" [formGroup]=\"formGroup\" *ngIf=\"formGroup\">\n  <textarea\n    matInput\n    luxTagIdHandler\n    [luxTagId]=\"luxTagId\"\n    [placeholder]=\"luxPlaceholder\"\n    [formControl]=\"formControl\"\n    [luxMaxLengthAttr]=\"luxMaxLength\"\n    matTextareaAutosize\n    [matAutosizeMinRows]=\"luxMinRows\"\n    [matAutosizeMaxRows]=\"luxMaxRows\"\n    [required]=\"luxRequired\"\n    [readonly]=\"luxReadonly\"\n    (blur)=\"luxBlur.emit($event)\"\n    (focus)=\"luxFocus.emit($event)\"\n    (focusin)=\"luxFocusIn.emit($event)\"\n    (focusout)=\"luxFocusOut.emit($event)\"\n    [id]=\"uid\"\n    [luxAriaDescribedby]=\"uid + '-error ' + uid + '-hint'\"\n  ></textarea>\n</lux-form-control>\n",
        styles: [""]
    }),
    tslib_1.__param(0, Optional()),
    tslib_1.__metadata("design:paramtypes", [ControlContainer,
        ChangeDetectorRef,
        LuxConsoleService,
        LuxComponentsConfigService])
], LuxTextareaComponent);
export { LuxTextareaComponent };
