// @ihk-gfi/lux-components/esm5/src/app/modules/lux-form/lux-form-model/lux-form-checkable-base.class.js
import * as tslib_1 from "tslib";
import { Validators } from '@angular/forms';
import { EventEmitter, Input, Output } from '@angular/core';
import { LuxFormComponentBase } from './lux-form-component-base.class';
/**
 * Basis-Klasse f체r FormComponents, die einen 채hnlichen Grundaufbau f체r das Aktivieren eines Boolean-Wertes besitzen
 * (LuxToggle und LuxCheckbox z.B.).
 */
var LuxFormCheckableBaseClass = /** @class */ (function (_super) {
    tslib_1.__extends(LuxFormCheckableBaseClass, _super);
    function LuxFormCheckableBaseClass(controlContainer, cdr, logger, config) {
        var _this = _super.call(this, controlContainer, cdr, logger, config) || this;
        _this.luxCheckedChange = new EventEmitter();
        return _this;
    }
    Object.defineProperty(LuxFormCheckableBaseClass.prototype, "luxChecked", {
        get: function () {
            return this.getValue();
        },
        set: function (checked) {
            this.setValue(checked);
        },
        enumerable: true,
        configurable: true
    });
    // region Overridden methods
    LuxFormCheckableBaseClass.prototype.notifyFormValueChanged = function (formValue) {
        // Aktualisierungen an dem FormControl-Value sollen auch via EventEmitter bekannt gemacht werden
        this.luxCheckedChange.emit(formValue);
        // Bei luxRequired = true && einem false-Wert entsprechend einen Fehler setzen
        if (formValue === false && this.luxRequired && this.formControl.errors === null) {
            this.formControl.setErrors({ required: true });
        }
    };
    LuxFormCheckableBaseClass.prototype.errorMessageModifier = function (value, errors) {
        if (errors.required) {
            return 'Das ist ein Pflichtfeld';
        }
        return undefined;
    };
    LuxFormCheckableBaseClass.prototype.checkValidatorsContainRequired = function (validators) {
        // Fall: required = true, aber neue Validatoren werden gesetzt
        if (this.luxRequired === true) {
            // Sind es mehrere Validatoren, aber kein .requiredTrue? Dann wird er erg채nzt
            if (Array.isArray(validators) && validators.indexOf(Validators.requiredTrue) === -1) {
                validators.push(Validators.requiredTrue);
            }
            else if (!Array.isArray(validators) && validators !== Validators.requiredTrue) {
                // Ist es nur ein einzelner Validator und nicht .requiredTrue? Dann Array erstellen und beide kombinieren
                validators = [validators, Validators.requiredTrue];
            }
        }
        else if (this.luxRequired === false) {
            if (Array.isArray(validators)) {
                validators = validators.filter(function (validator) { return validator !== Validators.requiredTrue; });
            }
            else if (validators === Validators.requiredTrue) {
                validators = undefined;
            }
        }
        return validators;
    };
    LuxFormCheckableBaseClass.prototype.triggerOutputPatternCheck = function () {
        this.checkOutputPatternViolation(this.luxCheckedChange.observers);
    };
    LuxFormCheckableBaseClass.prototype.triggerInputPatternCheck = function (simpleChanges) {
        this.checkInputPatternViolation(simpleChanges.luxChecked);
    };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxFormCheckableBaseClass.prototype, "luxCheckedChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxFormCheckableBaseClass.prototype, "luxTagId", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxFormCheckableBaseClass.prototype, "luxReadonly", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxFormCheckableBaseClass.prototype, "luxChecked", null);
    return LuxFormCheckableBaseClass;
}(LuxFormComponentBase));
export { LuxFormCheckableBaseClass };
