// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-tabindex/lux-tab-index.directive.js
import * as tslib_1 from "tslib";
import { AfterViewInit, Directive, ElementRef, Input, OnChanges, OnDestroy, Renderer2, SimpleChanges } from '@angular/core';
import { ReplaySubject } from 'rxjs';
let LuxTabIndexDirective = class LuxTabIndexDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.changes$ = new ReplaySubject(1);
        this.luxTabIndex = '0';
        this.luxApplyToParent = false;
        this.luxApplyToChildren = true;
        this.luxPotentialChildren = ['input', 'textarea', 'a', 'button', 'mat-select'];
    }
    ngOnChanges(simpleChanges) {
        this.changes$.next(simpleChanges);
    }
    ngAfterViewInit() {
        this.updateElementsWithTabIndex();
        // Dadurch verhindern wir, dass Changes vor afterViewInit zu etwaigen Fehlern führen könnten und erst danach aufgelöst werden
        this.changesSubscription = this.changes$.subscribe(this.onChanges.bind(this));
    }
    ngOnDestroy() {
        if (this.changesSubscription) {
            this.changesSubscription.unsubscribe();
        }
    }
    /**
     * Eigener Listener für Änderungen, der erst nach dem AfterViewInit-Lifecycle greift.
     * Sicherheitshalber, um evtl. noch unaufgelösten Elementen zu entgehen.
     * @param simpleChanges
     */
    onChanges(simpleChanges) {
        if (simpleChanges.luxTabIndex) {
            this.updateElementsWithTabIndex();
        }
    }
    /**
     * Aktualisiert den Tabindex für
     *  - das Zielelement, wenn luxApplyToParent === true (default = false) ist
     *  - potentielle Kindelemente, wenn luxApplyToChildren === true (default) ist
     */
    updateElementsWithTabIndex() {
        if (this.luxApplyToParent) {
            this.setTabIndexForElement(this.elementRef.nativeElement);
        }
        if (this.luxApplyToChildren) {
            this.luxPotentialChildren.forEach((childQuery) => this.setTabIndexByQuery(childQuery));
        }
    }
    setTabIndexByQuery(elementQuery) {
        const elements = this.elementRef.nativeElement.querySelectorAll(elementQuery);
        for (let i = 0; i < elements.length; i++) {
            const element = elements[i];
            this.setTabIndexForElement(element);
        }
    }
    setTabIndexForElement(element) {
        this.renderer.setAttribute(element, 'tabIndex', this.luxTabIndex);
    }
};
LuxTabIndexDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxTabIndexDirective.prototype, "luxTabIndex", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxTabIndexDirective.prototype, "luxApplyToParent", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxTabIndexDirective.prototype, "luxApplyToChildren", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], LuxTabIndexDirective.prototype, "luxPotentialChildren", void 0);
LuxTabIndexDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxTabIndex]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxTabIndexDirective);
export { LuxTabIndexDirective };
