// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-stepper/lux-stepper-subcomponents/lux-step-header.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { LuxComponentsConfigService } from '../../../lux-components-config/lux-components-config.service';
let LuxStepHeaderComponent = class LuxStepHeaderComponent {
    constructor(componentsConfigService) {
        this.componentsConfigService = componentsConfigService;
    }
    ngOnInit() {
        this.configSubscription = this.componentsConfigService.config.subscribe(() => {
            this.labelUppercase = this.componentsConfigService.isLabelUppercaseForSelector('lux-step');
        });
    }
    ngOnDestroy() {
        this.configSubscription.unsubscribe();
    }
};
LuxStepHeaderComponent.ctorParameters = () => [
    { type: LuxComponentsConfigService }
];
LuxStepHeaderComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-step-header',
        template: `
    <div [ngClass]="{ 'lux-uppercase': labelUppercase }">
      <ng-content></ng-content>
    </div>
  `,
        styles: ['']
    }),
    tslib_1.__metadata("design:paramtypes", [LuxComponentsConfigService])
], LuxStepHeaderComponent);
export { LuxStepHeaderComponent };
