// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-stepper/lux-stepper-subcomponents/lux-step.component.js
import * as tslib_1 from "tslib";
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { FormGroup } from '@angular/forms';
let LuxStepComponent = class LuxStepComponent {
    constructor() {
        this._iconChange = new BehaviorSubject(false);
        this._luxIconName = undefined;
        this.luxIconSize = '1x';
        this.luxOptional = false;
        this.luxEditable = true;
        this.luxCompleted = true;
    }
    get luxIconName() {
        return this._luxIconName;
    }
    set luxIconName(iconName) {
        this._luxIconName = iconName;
        this._iconChange.next(true);
    }
    getIconChangeObsv() {
        return this._iconChange.asObservable();
    }
    /**
     * Gibt an, ob der Step als abgeschlossen gilt.
     *
     * Möglichkeiten:
     *  1. Hat ein luxStepControl, welches valid ist
     *  2. Der Wert luxCompleted ist true
     *  3. Der Wert luxOptional ist true
     */
    isCompleted() {
        if (this.luxStepControl) {
            return this.luxStepControl.valid;
        }
        return this.luxCompleted || this.luxOptional;
    }
};
tslib_1.__decorate([
    ViewChild('header', { static: true }),
    tslib_1.__metadata("design:type", TemplateRef)
], LuxStepComponent.prototype, "headerTemplate", void 0);
tslib_1.__decorate([
    ViewChild('content', { static: true }),
    tslib_1.__metadata("design:type", TemplateRef)
], LuxStepComponent.prototype, "contentTemplate", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxStepComponent.prototype, "luxIconSize", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxStepComponent.prototype, "luxOptional", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxStepComponent.prototype, "luxEditable", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxStepComponent.prototype, "luxCompleted", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", FormGroup)
], LuxStepComponent.prototype, "luxStepControl", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxStepComponent.prototype, "luxIconName", null);
LuxStepComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-step',
        template: `
    <ng-template #header>
      <ng-content select="lux-step-header"></ng-content>
    </ng-template>
    <ng-template #content>
      <ng-content select="lux-step-content"></ng-content>
    </ng-template>
  `
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxStepComponent);
export { LuxStepComponent };
