// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-list/lux-list.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChildren, EventEmitter, HostBinding, HostListener, Input, Output, QueryList } from '@angular/core';
import { LuxListItemComponent } from './lux-list-subcomponents/lux-list-item.component';
import { FocusKeyManager } from '@angular/cdk/a11y';
import { DOWN_ARROW, END, ENTER, HOME, SPACE, UP_ARROW } from '@angular/cdk/keycodes';
var LuxListComponent = /** @class */ (function () {
    function LuxListComponent() {
        this._luxSelectedPosition = 0;
        this.clickSubscriptions = [];
        this.keyManager = new FocusKeyManager([]);
        this.luxFocusedItemChange = new EventEmitter();
        this.luxFocusedPositionChange = new EventEmitter();
        this.luxSelectedPositionChange = new EventEmitter();
        this.luxEmptyIconName = 'fas fa-info-circle';
        this.luxEmptyIconSize = '5x';
        this.luxEmptyLabel = 'Keine Einträge vorhanden';
        this.role = 'listbox';
        this.tabindex = '0';
        this.ariaMulti = 'true';
    }
    LuxListComponent.prototype.onKeydown = function ($event) {
        this.keydown($event);
    };
    Object.defineProperty(LuxListComponent.prototype, "luxSelectedPosition", {
        get: function () {
            return this._luxSelectedPosition;
        },
        set: function (position) {
            if (position === this.luxSelectedPosition) {
                return;
            }
            this.focus(position);
            this.select(position);
            this.scroll(position);
        },
        enumerable: true,
        configurable: true
    });
    LuxListComponent.prototype.ngOnInit = function () { };
    LuxListComponent.prototype.isEmpty = function () {
        return !this.luxItems || this.luxItems.length === 0;
    };
    LuxListComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        // Click-Events der LuxListItems erhalten
        this.listItemsSubscription = this.luxItems.changes.subscribe(function () {
            _this.listenForClicks();
            _this.keyManager = new FocusKeyManager(_this.luxItems);
            if (_this.previousFocusedPosition) {
                _this.keyManager.setActiveItem(_this.previousFocusedPosition);
            }
        });
        this.listenForClicks();
        this.keyManager = new FocusKeyManager(this.luxItems);
        if (this.luxSelectedPosition) {
            this.keyManager.setActiveItem(this.luxSelectedPosition);
        }
    };
    LuxListComponent.prototype.ngOnDestroy = function () {
        var e_1, _a;
        // Sämtliche Subscriptions auflösen
        if (this.listItemsSubscription) {
            this.listItemsSubscription.unsubscribe();
        }
        try {
            for (var _b = tslib_1.__values(this.clickSubscriptions), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subscription = _c.value;
                subscription.unsubscribe();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * Wird beim Drücken einer Taste ausgeführt und handelt die Aktionen bei speziellen Tasten
     * (UP_ARROW || DOWN_ARROW werden vom KeyManager selbstständig gepflegt)
     * @param $event
     */
    LuxListComponent.prototype.keydown = function ($event) {
        var keyCode = $event.keyCode;
        switch (keyCode) {
            case SPACE:
            case ENTER:
                this.select(this.keyManager.activeItemIndex);
                $event.preventDefault();
                break;
            case HOME:
            case END:
                keyCode === HOME ? this.keyManager.setFirstItemActive() : this.keyManager.setLastItemActive();
                this.focus(this.keyManager.activeItemIndex);
                $event.preventDefault();
                break;
            case UP_ARROW:
                this.keyManager.setPreviousItemActive();
                this.focus(this.keyManager.activeItemIndex);
                $event.preventDefault();
                break;
            case DOWN_ARROW:
                this.keyManager.setNextItemActive();
                this.focus(this.keyManager.activeItemIndex);
                $event.preventDefault();
                break;
            default:
                this.keyManager.onKeydown($event);
        }
    };
    /**
     * Auf Click-Events der hier bekannten LuxListItems hören und entsprechend das selektierte ListItem aktualisieren.
     */
    LuxListComponent.prototype.listenForClicks = function () {
        var _this = this;
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.clickSubscriptions), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subscription = _c.value;
                subscription.unsubscribe();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        this.clickSubscriptions = [];
        this.luxItems.forEach(function (listItem, index) {
            _this.clickSubscriptions.push(listItem.luxClicked.subscribe(function () {
                _this.focus(index);
                _this.select(index);
            }));
        });
    };
    /**
     * Merkt sich die position als Selektions-Position und aktualisiert den luxSelected-Wert
     * aller luxItems, die hier bekannt sind.
     * @param position
     */
    LuxListComponent.prototype.select = function (position) {
        this._luxSelectedPosition = position;
        this.previousFocusedPosition = position;
        this.luxSelectedPositionChange.emit(position);
        if (this.luxItems) {
            this.luxItems.forEach(function (listItem) {
                listItem.luxSelected = false;
            });
            var selectedListItem = this.findListItem(position);
            if (selectedListItem) {
                selectedListItem.luxSelected = true;
            }
        }
    };
    /**
     * Merkt sich die position als Fokus-Position und aktualisiert die CSS-Klassen der ListItems.
     * @param position
     */
    LuxListComponent.prototype.focus = function (position) {
        this.keyManager.setActiveItem(position);
        this.luxFocusedPositionChange.emit(position);
        this.luxFocusedItemChange.emit(this.keyManager.activeItem);
        this.previousFocusedPosition = position;
    };
    /**
     * Scrollt zu dem Element an der position.
     * @param position
     */
    LuxListComponent.prototype.scroll = function (position) {
        if (this.luxItems) {
            var listItem = this.keyManager.activeItem;
            if (listItem) {
                listItem.elementRef.nativeElement.scrollIntoView(true);
            }
        }
    };
    /**
     * Gibt das ListItem an der position zurück bzw. null wenn die luxItems undefined/null sind.
     * @param position
     */
    LuxListComponent.prototype.findListItem = function (position) {
        return this.luxItems
            ? this.luxItems.find(function (listItem, index) { return index === position; })
            : null;
    };
    tslib_1.__decorate([
        ContentChildren(LuxListItemComponent),
        tslib_1.__metadata("design:type", QueryList)
    ], LuxListComponent.prototype, "luxItems", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxListComponent.prototype, "luxFocusedItemChange", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxListComponent.prototype, "luxFocusedPositionChange", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxListComponent.prototype, "luxSelectedPositionChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "luxEmptyIconName", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "luxEmptyIconSize", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "luxEmptyLabel", void 0);
    tslib_1.__decorate([
        HostBinding('attr.role'),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "role", void 0);
    tslib_1.__decorate([
        HostBinding('attr.tabindex'),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "tabindex", void 0);
    tslib_1.__decorate([
        HostBinding('attr.aria-multiselectable'),
        tslib_1.__metadata("design:type", Object)
    ], LuxListComponent.prototype, "ariaMulti", void 0);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], LuxListComponent.prototype, "onKeydown", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], LuxListComponent.prototype, "luxSelectedPosition", null);
    LuxListComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-list',
            template: "<!-- die tats\u00E4chlichen LuxListItems via Content-Projection einbinden -->\n<ng-content *ngIf=\"!isEmpty()\"></ng-content>\n<!-- wenn keine gefunden werden konnten, Empty-Icon und Empty-Label anzeigen -->\n<div class=\"lux-list-empty\" fxFill *ngIf=\"isEmpty()\">\n  <div fxFlexFill fxLayout=\"column\" fxLayoutAlign=\"center center\">\n    <lux-icon class=\"lux-list-empty-icon\" [luxIconName]=\"luxEmptyIconName\" [luxIconSize]=\"luxEmptyIconSize\"> </lux-icon>\n    <span class=\"lux-list-empty-icon-text\">\n      <strong>{{ luxEmptyLabel }}</strong>\n    </span>\n  </div>\n</div>\n",
            styles: ["@charset \"UTF-8\";lux-side-nav .lux-side-nav-item,lux-side-nav lux-link{box-sizing:border-box}lux-side-nav .lux-side-nav-item.lux-side-nav-item-selected,lux-side-nav lux-link.lux-side-nav-item-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}lux-side-nav .lux-side-nav-item.lux-side-nav-item-selected:after,lux-side-nav lux-link.lux-side-nav-item-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}lux-side-nav .lux-side-nav-item button .mat-button-focus-overlay,lux-side-nav lux-link button .mat-button-focus-overlay{display:none}lux-side-nav .lux-side-nav-item button:focus,lux-side-nav lux-link button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-side-nav .lux-side-nav-item button:hover,lux-side-nav lux-link button:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-list{border:4px solid transparent}lux-list:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-list lux-list-item:focus{outline-width:0}lux-list lux-list-item:focus mat-card{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-list lux-list-item lux-card.lux-list-item-selected mat-card{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}lux-list lux-list-item lux-card.lux-list-item-selected mat-card:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}lux-list lux-list-item lux-card mat-card{box-sizing:border-box}lux-list lux-list-item lux-card mat-card:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .lux-form-control-label>*{color:#335c85}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .cdk-focused .mat-radio-container,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-checkbox-inner-container,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-slide-toggle-thumb{position:relative}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .cdk-focused .mat-radio-container:after,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-checkbox-inner-container:after,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-slide-toggle-thumb:after{content:\"\";border:1px solid rgba(112,141,170,.2);background-color:rgba(112,141,170,.2);display:block;top:-4px;left:-4px;right:-4px;bottom:-4px;border-radius:15px;z-index:1;position:absolute}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .lux-form-control-container:after{content:'';display:block;box-sizing:border-box;left:0;right:0;position:absolute;height:2px;border-bottom:2px solid #0073b3}lux-datepicker mat-datepicker-toggle button:focus:after{content:\"\";border:1px solid rgba(112,141,170,.2);background-color:rgba(112,141,170,.2);display:block;top:-4px;left:-4px;right:-4px;bottom:-4px;border-radius:15px;z-index:1;position:absolute}.lux-datepicker-panel .mat-calendar-header button .mat-button-focus-overlay,lux-datepicker mat-datepicker-toggle .mat-button-focus-overlay{display:none}.lux-datepicker-panel .mat-calendar-header button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-datepicker-panel .mat-calendar-header button:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}.lux-datepicker-panel .mat-calendar-content td.mat-calendar-body-cell:hover>div{background-color:rgba(0,0,0,.2)!important;border-radius:999px}.lux-datepicker-panel .mat-calendar-content td.mat-calendar-body-cell.mat-calendar-body-active>div{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:999px}lux-table tr:hover>td,lux-table tr:hover>th{border-radius:0;background-color:#cfcfcf!important}lux-table th button.mat-sort-header-button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-select-panel mat-option{border-left:0;box-sizing:border-box}.lux-select-panel mat-option.mat-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}.lux-select-panel mat-option.mat-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}.lux-select-panel mat-option:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-select-panel mat-option.mat-active,.lux-select-panel mat-option:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-button button:not([disabled]) .mat-button-focus-overlay{display:none}lux-button button:not([disabled]):not(.mat-fab):not(.mat-raised-button):focus{background-color:rgba(112,141,170,.2)!important}lux-button button:not([disabled]):not(.mat-fab):not(.mat-raised-button):hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-button button:not([disabled]).mat-fab:focus,lux-button button:not([disabled]).mat-raised-button:focus{background-color:rgba(112,141,170,.2)!important}lux-button button:not([disabled]).mat-fab:focus.mat-accent,lux-button button:not([disabled]).mat-raised-button:focus.mat-accent{background-color:#b6d184!important}lux-button button:not([disabled]).mat-fab:focus.mat-warn,lux-button button:not([disabled]).mat-raised-button:focus.mat-warn{background-color:#ea2524!important}lux-button button:not([disabled]).mat-fab:focus.mat-primary,lux-button button:not([disabled]).mat-raised-button:focus.mat-primary{background-color:#4b82ba!important}.lux-autocomplete-panel mat-option{border-left:0;box-sizing:border-box}.lux-autocomplete-panel mat-option:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-autocomplete-panel mat-option.mat-active,.lux-autocomplete-panel mat-option:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-tabs mat-ink-bar{background-color:#0073b3!important}lux-tabs .mat-tab-label:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-panel mat-expansion-panel-header:hover,lux-stepper mat-step-header:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-panel mat-expansion-panel-header:focus,lux-stepper mat-step-header:focus,lux-stepper mat-step-header[aria-selected=true],lux-tile mat-card:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-master-detail lux-master-header .lux-master-header-collapse button.lux-button:not(.mat-fab):not(.mat-raised-button):focus{background-color:rgba(0,0,0,.2)!important}lux-file-list .lux-file-list{border:4px solid transparent}lux-file-list .lux-file-list:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-file-list .lux-file-list .lux-file-list-entry{border:4px solid transparent}lux-file-list .lux-file-list .lux-file-list-entry:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-file-list .lux-file-list mat-card{margin:4px}.lux-menu-panel .lux-menu-item:not([disabled=true]):focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-menu-panel .lux-menu-item:not([disabled=true]):hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-dialog-structure lux-dialog-content:focus,lux-dialog-structure lux-dialog-title:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}example-root .example-base-components-list .example-component-list-item,example-root .example-base-components-list mat-expansion-panel-header{box-sizing:border-box}example-root .example-base-components-list .example-component-list-item.example-component-list-item-selected,example-root .example-base-components-list mat-expansion-panel-header.example-component-list-item-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}example-root .example-base-components-list .example-component-list-item.example-component-list-item-selected:after,example-root .example-base-components-list mat-expansion-panel-header.example-component-list-item-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}example-root .example-base-components-list .example-component-list-item:focus,example-root .example-base-components-list mat-expansion-panel-header:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}example-root .example-base-components-list .example-component-list-item:hover,example-root .example-base-components-list mat-expansion-panel-header:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}.lux-app-container,body,html{height:100%}.lux-app-container{-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;box-sizing:border-box;display:-webkit-box;display:-ms-flexbox;display:flex}body{-webkit-overflow-scrolling:touch}div{font-family:Roboto,Helvetica Neue,sans-serif}.cdk-overlay-backdrop.cdk-overlay-backdrop-showing.lux-file-preview-backdrop{background:#000;opacity:.8!important}lux-card.lux-card-grow{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-flex:1;-ms-flex:1 1 auto;flex:1 1 auto;width:400px;min-width:0}lux-card.lux-card-grow mat-card,lux-card.lux-card-grow mat-card-content{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-flex:1;-ms-flex:1 1 auto;flex:1 1 auto;min-width:0}.lux-crop{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.lux-label{color:rgba(0,0,0,.54);font-size:75%}.lux-hint{color:rgba(0,0,0,.6);font-size:12px}.lux-display-none{display:none}.lux-hidden{visibility:hidden}.lux-nowrap{white-space:nowrap}.lux-cursor{cursor:pointer}.lux-block-pointer-events{pointer-events:none}.lux-vertical-align-baseline{vertical-align:baseline!important}.lux-black-semi-transparent{color:rgba(0,0,0,.87)}.lux-semi-transparent{opacity:.6}.lux-overflow-y-hidden{overflow-y:hidden}.lux-overflow-y-auto{overflow-y:auto}.lux-overflow-wrap-break-word{overflow-wrap:break-word}.lux-bold{font-weight:700!important}.lux-uppercase{text-transform:uppercase!important}.lux-lowercase{text-transform:lowercase!important}.lux-hyphenate{overflow-wrap:break-word;word-wrap:break-word;word-break:normal;-webkit-hyphens:auto;-ms-hyphens:auto;hyphens:auto;overflow-x:auto}.lux-p-footer,.lux-p-header{padding:8px}.lux-snackbar{top:68px;right:8px;position:absolute;margin:0!important;min-width:unset!important;box-shadow:0 3px 1px -2px rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.14),0 1px 5px 0 rgba(0,0,0,.12)}.mat-snack-bar-handset .lux-snackbar{left:8px;width:auto!important}.lux-mt-0{margin-top:0}.lux-pt-0{padding-top:0}.lux-mt-1{margin-top:2px}.lux-pt-1{padding-top:2px}.lux-mt-2{margin-top:4px}.lux-pt-2{padding-top:4px}.lux-mt-3{margin-top:8px}.lux-pt-3{padding-top:8px}.lux-mt-4{margin-top:16px}.lux-pt-4{padding-top:16px}.lux-mb-0{margin-bottom:0}.lux-pb-0{padding-bottom:0}.lux-mb-1{margin-bottom:2px}.lux-pb-1{padding-bottom:2px}.lux-mb-2{margin-bottom:4px}.lux-pb-2{padding-bottom:4px}.lux-mb-3{margin-bottom:8px}.lux-pb-3{padding-bottom:8px}.lux-mb-4{margin-bottom:16px}.lux-pb-4{padding-bottom:16px}.lux-ml-0{margin-left:0}.lux-pl-0{padding-left:0}.lux-ml-1{margin-left:2px}.lux-pl-1{padding-left:2px}.lux-ml-2{margin-left:4px}.lux-pl-2{padding-left:4px}.lux-ml-3{margin-left:8px}.lux-pl-3{padding-left:8px}.lux-ml-4{margin-left:16px}.lux-pl-4{padding-left:16px}.lux-m-0{margin:0}.lux-mr-0{margin-right:0}.lux-p-0{padding:0}.lux-pr-0{padding-right:0}.lux-m-1{margin:2px}.lux-mr-1{margin-right:2px}.lux-p-1{padding:2px}.lux-pr-1{padding-right:2px}.lux-m-2{margin:4px}.lux-mr-2{margin-right:4px}.lux-p-2{padding:4px}.lux-pr-2{padding-right:4px}.lux-m-3{margin:8px}.lux-mr-3{margin-right:8px}.lux-p-3{padding:8px}.lux-pr-3{padding-right:8px}.lux-m-4{margin:16px}.lux-mr-4{margin-right:16px}.lux-p-4{padding:16px}.lux-pr-4{padding-right:16px}.lux-flex-layout-gap-10[style*=row]>:not(:last-child){margin-right:10px}.lux-flex-layout-gap-10[style*=column]>:not(:last-child){margin-bottom:10px}.lux-flex-layout-gap-20[style*=row]>:not(:last-child){margin-right:20px}.lux-flex-layout-gap-20[style*=column]>:not(:last-child){margin-bottom:20px}.lux-flex-layout-gap-30[style*=row]>:not(:last-child){margin-right:30px}.lux-flex-layout-gap-30[style*=column]>:not(:last-child){margin-bottom:30px}.lux-z-index-100{z-index:100}.lux-z-index-500{z-index:500}.lux-z-index-1000{z-index:1000}.lux-uppercase>lux-button .lux-button-label,lux-button.lux-uppercase .lux-button-label{text-transform:uppercase}.lux-bg-color-red{background:#b01211}.lux-bg-color-green{background:#95bc4c}.lux-bg-color-yellow{background:#ebc000}.lux-bg-color-blue{background:#335c85}.lux-bg-color-gray{background:#c8c8c8}.lux-bg-color-orange{background:#ec9c1b}.lux-bg-color-brown{background:#783f04}.lux-font-color-black{color:#000}.lux-font-color-white{color:#fff}.lux-badge-notification.mat-badge-large .mat-badge-content{font-size:15px}.lux-badge-notification.lux-badge-color-default .mat-badge-content{color:#000;background:#fff;border:1px solid rgba(0,0,0,.87)}:host{position:relative;height:100%;width:100%;display:block;overflow-y:auto;box-sizing:border-box}.lux-list-empty{position:absolute;top:0;left:0;right:0;bottom:0}.lux-list-empty-icon,.lux-list-empty-icon-text{opacity:.6;color:rgba(0,0,0,.87);padding:0 8px}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxListComponent);
    return LuxListComponent;
}());
export { LuxListComponent };
