// @ihk-gfi/lux-components/esm5/src/app/modules/lux-common/lux-badge/lux-badge.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { LuxBackgroundColorsEnum } from '../../lux-util/lux-colors.enum';
import { LuxUtil } from '../../lux-util/lux-util';
var LuxBadgeComponent = /** @class */ (function () {
    function LuxBadgeComponent() {
        this.ICON_SIZE = '1x';
        this.DEFAULT_BADGE_COLOR = LuxBackgroundColorsEnum.gray;
        this.luxUppercase = true;
        this.luxIconName = '';
        this._backgroundCSSClass = '';
        this._fontCSSClass = '';
        this.luxColor = this.DEFAULT_BADGE_COLOR;
    }
    Object.defineProperty(LuxBadgeComponent.prototype, "backgroundCSSClass", {
        get: function () {
            return this._backgroundCSSClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxBadgeComponent.prototype, "fontCSSClass", {
        get: function () {
            return this._fontCSSClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxBadgeComponent.prototype, "luxColor", {
        set: function (color) {
            var result = LuxUtil.getColorsByBgColorsEnum(color);
            this._backgroundCSSClass = result.backgroundCSSClass;
            this._fontCSSClass = result.fontCSSClass;
            if (!LuxBackgroundColorsEnum[color]) {
                this.luxColor = this.DEFAULT_BADGE_COLOR;
            }
        },
        enumerable: true,
        configurable: true
    });
    LuxBadgeComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxBadgeComponent.prototype, "luxUppercase", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxBadgeComponent.prototype, "luxIconName", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxBadgeComponent.prototype, "luxColor", null);
    LuxBadgeComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-badge',
            template: "<div\n  class=\"lux-badge lux-mt-1 lux-mb-1 lux-ml-1 lux-mr-1 lux-pt-1 lux-pb-1 lux-pl-3 lux-pr-3\"\n  [ngClass]=\"[backgroundCSSClass, fontCSSClass, luxUppercase ? 'lux-badge-uppercase' : '']\"\n  *ngIf=\"backgroundCSSClass\"\n>\n  <lux-icon [luxIconName]=\"luxIconName\" [luxIconSize]=\"ICON_SIZE\" luxMargin=\"0 8px 0 0\" *ngIf=\"luxIconName\"></lux-icon>\n  <ng-content select=\"lux-label\"></ng-content>\n</div>\n",
            styles: [".lux-badge{border-radius:8px;display:-webkit-inline-box;display:-ms-inline-flexbox;display:inline-flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-flex:initial;-ms-flex:initial;flex:initial}.lux-badge.lux-badge-uppercase ::ng-deep *{text-transform:uppercase}.lux-badge>lux-icon ::ng-deep *{padding:0!important}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxBadgeComponent);
    return LuxBadgeComponent;
}());
export { LuxBadgeComponent };
