// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-file-preview/lux-file-preview-imgviewer/lux-file-preview-imgviewer.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, Inject, ViewChild } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { LuxFilePreviewBase } from '../lux-file-preview-base/lux-file-preview-base';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
import { LuxFilePreviewRef } from '../lux-file-preview-ref';
let LuxFilePreviewImgViewerComponent = class LuxFilePreviewImgViewerComponent extends LuxFilePreviewBase {
    constructor(previewRef, previewData, sanitizer) {
        super(previewRef, previewData, sanitizer);
        this.previewRef = previewRef;
        this.previewData = previewData;
        this.sanitizer = sanitizer;
        this.zoomActive = false;
        this.zoomWidth = 0;
        this.zoomStep = 250;
    }
    ngAfterViewInit() {
        setTimeout(() => {
            this.zoomActive =
                this.previewImg && this.previewImg.nativeElement.naturalWidth - this.paddingWith > window.innerWidth;
        });
    }
    onLoad(event) {
        this.zoomActive =
            this.previewImg && this.previewImg.nativeElement.naturalWidth - this.paddingWith > window.innerWidth;
        this.loadingFinished();
    }
    onZoomIn() {
        this.zoomWidth += this.zoomStep;
        this.clearFocus();
    }
    onZoomOut() {
        this.zoomWidth -= this.zoomStep;
        this.clearFocus();
    }
};
LuxFilePreviewImgViewerComponent.ctorParameters = () => [
    { type: LuxFilePreviewRef },
    { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] },
    { type: DomSanitizer }
];
tslib_1.__decorate([
    ViewChild('previewImg', { static: false }),
    tslib_1.__metadata("design:type", ElementRef)
], LuxFilePreviewImgViewerComponent.prototype, "previewImg", void 0);
LuxFilePreviewImgViewerComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-file-preview-imgviewer',
        template: "<lux-file-preview-toolbar>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-search-plus\"\n    *ngIf=\"zoomActive\"\n    (luxClicked)=\"onZoomIn()\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-zoomin-btn\"\n    luxAriaLabel=\"Hinein zoomen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-search-minus\"\n    (luxClicked)=\"onZoomOut()\"\n    *ngIf=\"zoomActive\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-zoomout-btn\"\n    luxAriaLabel=\"Heraus zoomen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    [luxIconName]=\"downloadIconName\"\n    (luxClicked)=\"onDownload()\"\n    luxColor=\"accent\"\n    luxRaised=\"true\"\n    [luxTagId]=\"downloadTagId\"\n    [luxAriaLabel]=\"downloadAriaLabel\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    [luxIconName]=\"closeIconName\"\n    (luxClicked)=\"onClose()\"\n    luxRaised=\"true\"\n    [luxTagId]=\"closeTagId\"\n    [luxAriaLabel]=\"closeAriaLabel\"\n  ></lux-button>\n</lux-file-preview-toolbar>\n\n<div class=\"overlay-content\" [ngStyle]=\"{ 'max-width': width + 'px', 'max-height': height + 'px' }\">\n  <img\n    [src]=\"this.sanitizer.bypassSecurityTrustResourceUrl(this.url)\"\n    *ngIf=\"url\"\n    (load)=\"onLoad($event)\"\n    [ngStyle]=\"{\n      'max-width': width + zoomWidth + 'px',\n      'max-height': height + zoomWidth + 'px',\n      display: loading ? 'none' : 'block'\n    }\"\n    #previewImg\n  />\n  <div *ngIf=\"loading && !startPhase\" class=\"message-box\">\n    Lade...\n  </div>\n</div>\n",
        styles: [".overlay-content{overflow:auto}img{height:auto}"]
    }),
    tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
    tslib_1.__metadata("design:paramtypes", [LuxFilePreviewRef, Object, DomSanitizer])
], LuxFilePreviewImgViewerComponent);
export { LuxFilePreviewImgViewerComponent };
