// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-tabs/lux-tabs-subcomponents/lux-tab.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, Input, TemplateRef } from '@angular/core';
var LuxTabComponent = /** @class */ (function () {
    function LuxTabComponent() {
        this.luxCounterCap = 10;
        this.luxDisabled = false;
    }
    LuxTabComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTabComponent.prototype, "luxTitle", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTabComponent.prototype, "luxIconName", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], LuxTabComponent.prototype, "luxCounter", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], LuxTabComponent.prototype, "luxCounterCap", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxTabComponent.prototype, "luxShowNotification", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTabComponent.prototype, "luxDisabled", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTabComponent.prototype, "luxTagIdHeader", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTabComponent.prototype, "luxTagIdContent", void 0);
    tslib_1.__decorate([
        ContentChild(TemplateRef, { static: false }),
        tslib_1.__metadata("design:type", TemplateRef)
    ], LuxTabComponent.prototype, "contentTemplate", void 0);
    LuxTabComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-tab',
            template: ''
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxTabComponent);
    return LuxTabComponent;
}());
export { LuxTabComponent };
