// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-divider/lux-divider.component.js
import * as tslib_1 from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
var LuxDividerComponent = /** @class */ (function () {
    function LuxDividerComponent() {
        this._luxVertical = false;
        this.luxInset = false;
        this.classes = '';
    }
    Object.defineProperty(LuxDividerComponent.prototype, "luxVertical", {
        get: function () {
            return this._luxVertical;
        },
        set: function (vertical) {
            this._luxVertical = vertical;
            this.updateHostClasses();
        },
        enumerable: true,
        configurable: true
    });
    LuxDividerComponent.prototype.ngOnInit = function () {
        this.updateHostClasses();
    };
    LuxDividerComponent.prototype.updateHostClasses = function () {
        this.classes = this.luxVertical ? 'lux-vertical-divider' : 'lux-horizontal-divider';
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxDividerComponent.prototype, "luxInset", void 0);
    tslib_1.__decorate([
        HostBinding('class'),
        tslib_1.__metadata("design:type", Object)
    ], LuxDividerComponent.prototype, "classes", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxDividerComponent.prototype, "luxVertical", null);
    LuxDividerComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-divider',
            template: "<mat-divider [inset]=\"luxInset\" [vertical]=\"luxVertical\"> </mat-divider>\n",
            styles: [".mat-divider-vertical{width:0;height:100%}:host{display:block}:host.lux-vertical-divider{position:relative;margin-left:8px;margin-right:8px;min-width:1px}:host.lux-horizontal-divider{margin-top:8px;margin-bottom:8px;min-height:1px}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxDividerComponent);
    return LuxDividerComponent;
}());
export { LuxDividerComponent };
