// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-stepper/lux-stepper.component.js
import * as tslib_1 from "tslib";
import { AfterViewInit, ChangeDetectorRef, Component, ComponentFactory, ComponentFactoryResolver, ComponentRef, ContentChildren, ElementRef, EventEmitter, Input, Output, QueryList, ViewContainerRef } from '@angular/core';
import { LuxStepperHelperService } from './lux-stepper-helper.service';
import { LuxStepComponent } from './lux-stepper-subcomponents/lux-step.component';
import { LuxIconComponent } from '../../lux-icon/lux-icon/lux-icon.component';
import { LuxUtil } from '../../lux-util/lux-util';
var LuxStepperComponent = /** @class */ (function () {
    function LuxStepperComponent(stepperService, cdr, cfr, elementRef) {
        this.stepperService = stepperService;
        this.cdr = cdr;
        this.cfr = cfr;
        this.elementRef = elementRef;
        this._DEFAULT_PREV_BTN_CONF = {
            label: 'Zurück'
        };
        this._DEFAULT_NEXT_BTN_CONF = {
            label: 'Weiter'
        };
        this._DEFAULT_FIN_BTN_CONF = {
            label: 'Abschließen',
            color: 'primary'
        };
        this.luxFinishButtonClicked = new EventEmitter();
        this.luxStepChanged = new EventEmitter();
        this.luxCurrentStepNumberChange = new EventEmitter();
        this.stepperConfiguration = {
            luxCurrentStepNumber: 0,
            luxShowNavigationButtons: true,
            luxHorizontalStepAnimationActive: true,
            luxEditedIconName: 'fa-pencil'
        };
        // Die Default-Konfiguration präventiv als Startwert setzen
        this.luxPreviousButtonConfig = this._DEFAULT_PREV_BTN_CONF;
        this.luxNextButtonConfig = this._DEFAULT_NEXT_BTN_CONF;
        this.luxFinishButtonConfig = this._DEFAULT_FIN_BTN_CONF;
        // Den Stepper im Helper-Service bekannt machen
        this.stepperService.registerStepper(this);
    }
    LuxStepperComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        // Änderungen an den luxSteps sollten auch dem Konfigurationsobjekt bekannt gemacht werden
        this.luxSteps.changes.subscribe(function () {
            _this.stepperConfiguration.luxSteps = _this.luxSteps.toArray();
            _this.cdr.detectChanges();
        });
        // Initial die aktuellen luxSteps in die Konfiguration schreiben
        this.stepperConfiguration.luxSteps = this.luxSteps.toArray();
        this.cdr.detectChanges();
        // Falls initial bereits bestimmt wurde, dass individuelle Icons genutzt werden, diese generieren
        if (this.stepperConfiguration.luxUseCustomIcons) {
            this.generateCustomIcons();
        }
        // Auf next/previous Aufrufe aus dem Service horchen und entsprechend reagieren
        this.stepperService.getObservable(this).subscribe(function (next) {
            // Voraussetzung: Stepper nicht deaktiviert
            if (!_this.stepperConfiguration.luxDisabled) {
                if (next === true) {
                    _this.checkValidation();
                    _this.matStepper.next();
                    _this.matStepper._stepHeader.toArray()[_this.matStepper.selectedIndex].focus();
                }
                else if (next === false) {
                    _this.matStepper.previous();
                    _this.matStepper._stepHeader.toArray()[_this.matStepper.selectedIndex].focus();
                }
            }
        });
        // Änderungen an den Icons jedes einzelnen Steps führt zu Neugenerierung aller individuellen Icons
        // ==> Material erlaubt leider nur alle Icons identisch zu ändern, nicht für jeden Step einzeln, deshalb
        // generieren wir selbst die Icons.
        this.luxSteps.toArray().forEach(function (luxStep) {
            luxStep.getIconChangeObsv().subscribe(function (iconChange) {
                if (_this.stepperConfiguration.luxUseCustomIcons && iconChange) {
                    _this.clearCustomIcons();
                    _this.generateCustomIcons();
                }
            });
        });
        this.setFocusedCSS(this.luxCurrentStepNumber);
        this.cdr.detectChanges();
    };
    /**
     * Wird beim Wechsel des aktuellen Steps (Klick auf Tab oder .next()/.previous() Aufruf) aufgerufen.
     * @param $event
     */
    LuxStepperComponent.prototype.onStepChanged = function ($event) {
        this.luxCurrentStepNumber = $event.selectedIndex;
        this.luxStepChanged.emit($event);
        var matStepHeaders = this.elementRef.nativeElement.querySelectorAll('mat-step-header');
        if (matStepHeaders.item($event.selectedIndex).className.indexOf('lux-step-header-touched') === -1) {
            matStepHeaders.item($event.selectedIndex).className += ' lux-step-header-touched';
        }
        this.setFocusedCSS($event.selectedIndex);
    };
    /**
     * Generiert die individuellen Icons fuer die Steps.
     */
    LuxStepperComponent.prototype.generateCustomIcons = function () {
        var _this = this;
        var factory = this.cfr.resolveComponentFactory(LuxIconComponent);
        var index = 0;
        this.matStepLabels.forEach(function (stepLabel) {
            _this.generateCustomIconForStep(stepLabel, _this.luxSteps.toArray()[index], factory);
            index++;
        });
    };
    /**
     * Entfernt die eigenen Icons fuer die Steps.
     */
    LuxStepperComponent.prototype.clearCustomIcons = function () {
        this.matStepLabels.forEach(function (stepLabel) {
            stepLabel.clear();
        });
    };
    /**
     * Stößt die Validierungsprüfung für den aktuell sichtbaren Step und dessen StepControl (wenn vorhanden) an.
     */
    LuxStepperComponent.prototype.checkValidation = function () {
        var stepControl = this.luxSteps.toArray()[this.stepperConfiguration.luxCurrentStepNumber].luxStepControl;
        if (stepControl) {
            LuxUtil.showValidationErrors(stepControl);
        }
    };
    /**
     * Generiert die Icons fuer einen einzelnen Step
     * @param stepLabel
     * @param luxStep
     * @param factory
     */
    LuxStepperComponent.prototype.generateCustomIconForStep = function (stepLabel, luxStep, factory) {
        if (luxStep && luxStep.luxIconName) {
            // Das edited und normal Icon generieren
            var componentIconEdited = stepLabel.createComponent(factory);
            var instanceIconEdited = componentIconEdited.instance;
            instanceIconEdited.luxIconName = this.luxEditedIconName;
            instanceIconEdited.luxIconSize = '2x';
            instanceIconEdited.luxRounded = true;
            instanceIconEdited.luxMargin = '0 8px 0 0';
            componentIconEdited.location.nativeElement.className += ' lux-stepper-edited-icon';
            var componentIconNormal = stepLabel.createComponent(factory);
            var instanceIconNormal = componentIconNormal.instance;
            instanceIconNormal.luxIconName = luxStep.luxIconName;
            instanceIconNormal.luxIconSize = '2x';
            instanceIconNormal.luxRounded = true;
            instanceIconNormal.luxMargin = '0 8px 0 0';
            componentIconNormal.location.nativeElement.className += ' lux-stepper-normal-icon';
        }
    };
    /**
     * Aktualisiert die aktuellen Icons, entfernt zunächst die individuellen Icons und
     * versucht anschließend diese neu zu generieren (nötig bei Änderungen).
     */
    LuxStepperComponent.prototype.updateIcons = function () {
        if (this.matStepLabels) {
            this.clearCustomIcons();
            if (this.stepperConfiguration.luxUseCustomIcons) {
                this.generateCustomIcons();
            }
        }
    };
    LuxStepperComponent.prototype.setFocusedCSS = function (index) {
        var matStepHeaders = this.elementRef.nativeElement.querySelectorAll('mat-step-header');
        if (matStepHeaders.item(index).className.indexOf('lux-step-header-touched') === -1) {
            matStepHeaders.item(index).className += ' lux-step-header-touched';
        }
    };
    Object.defineProperty(LuxStepperComponent.prototype, "luxCurrentStepNumber", {
        // region ###### Accessoren, die auf die stepperConfiguration verweisen ######
        /**** Getter/Setter luxCurrentStepNumber ****/
        get: function () {
            return this.stepperConfiguration.luxCurrentStepNumber;
        },
        set: function (step) {
            var _this = this;
            if (step !== this.luxCurrentStepNumber) {
                setTimeout(function () {
                    // OutOfBound-Steps abfangen
                    step = step < 0 ? 0 : step;
                    step = step >= _this.luxSteps.length ? _this.luxSteps.length - 1 : step;
                    _this.stepperConfiguration.luxCurrentStepNumber = step;
                    _this.luxCurrentStepNumberChange.emit(_this.stepperConfiguration.luxCurrentStepNumber);
                });
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxUseCustomIcons", {
        /**** Getter/Setter luxUseCustomIcons ****/
        get: function () {
            return this.stepperConfiguration.luxUseCustomIcons;
        },
        set: function (use) {
            var _this = this;
            this.stepperConfiguration.luxUseCustomIcons = use;
            setTimeout(function () {
                _this.updateIcons();
            });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxEditedIconName", {
        /**** Getter/Setter luxEditedIconName ****/
        get: function () {
            return this.stepperConfiguration.luxEditedIconName;
        },
        set: function (iconName) {
            this.stepperConfiguration.luxEditedIconName = iconName;
            this.updateIcons();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxVerticalStepper", {
        /**** Getter/Setter luxVerticalStepper ****/
        get: function () {
            return this.stepperConfiguration.luxVerticalStepper;
        },
        set: function (vertical) {
            this.stepperConfiguration.luxVerticalStepper = vertical;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxLinear", {
        /**** Getter/Setter luxLinear ****/
        get: function () {
            return this.stepperConfiguration.luxLinear;
        },
        set: function (linear) {
            this.stepperConfiguration.luxLinear = linear;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxDisabled", {
        /**** Getter/Setter luxDisabled ****/
        get: function () {
            return this.stepperConfiguration.luxDisabled;
        },
        set: function (disabled) {
            this.stepperConfiguration.luxDisabled = disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxShowNavigationButtons", {
        /**** Getter/Setter luxShowNavigationButtons ****/
        get: function () {
            return this.stepperConfiguration.luxShowNavigationButtons;
        },
        set: function (showNavButtons) {
            this.stepperConfiguration.luxShowNavigationButtons = showNavButtons;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxHorizontalStepAnimationActive", {
        /**** Getter/Setter luxHorizontalStepAnimationActive ****/
        get: function () {
            return this.stepperConfiguration.luxHorizontalStepAnimationActive;
        },
        set: function (animationActive) {
            this.stepperConfiguration.luxHorizontalStepAnimationActive = animationActive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxPreviousButtonConfig", {
        /**** Getter/Setter luxPreviousButtonConfig ****/
        get: function () {
            return this.stepperConfiguration.luxPreviousButtonConfig;
        },
        set: function (config) {
            this.stepperConfiguration.luxPreviousButtonConfig = config ? config : this._DEFAULT_PREV_BTN_CONF;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxNextButtonConfig", {
        /**** Getter/Setter luxNextButtonConfig ****/
        get: function () {
            return this.stepperConfiguration.luxNextButtonConfig;
        },
        set: function (config) {
            this.stepperConfiguration.luxNextButtonConfig = config ? config : this._DEFAULT_NEXT_BTN_CONF;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxStepperComponent.prototype, "luxFinishButtonConfig", {
        /**** Getter/Setter luxFinishButtonConfig ****/
        get: function () {
            return this.stepperConfiguration.luxFinishButtonConfig;
        },
        set: function (config) {
            this.stepperConfiguration.luxFinishButtonConfig = config ? config : this._DEFAULT_FIN_BTN_CONF;
        },
        enumerable: true,
        configurable: true
    });
    LuxStepperComponent.ctorParameters = function () { return [
        { type: LuxStepperHelperService },
        { type: ChangeDetectorRef },
        { type: ComponentFactoryResolver },
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        ContentChildren(LuxStepComponent),
        tslib_1.__metadata("design:type", QueryList)
    ], LuxStepperComponent.prototype, "luxSteps", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperComponent.prototype, "luxFinishButtonClicked", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperComponent.prototype, "luxStepChanged", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperComponent.prototype, "luxCurrentStepNumberChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], LuxStepperComponent.prototype, "luxCurrentStepNumber", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxUseCustomIcons", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], LuxStepperComponent.prototype, "luxEditedIconName", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxVerticalStepper", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxLinear", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxDisabled", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxShowNavigationButtons", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], LuxStepperComponent.prototype, "luxHorizontalStepAnimationActive", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], LuxStepperComponent.prototype, "luxPreviousButtonConfig", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], LuxStepperComponent.prototype, "luxNextButtonConfig", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], LuxStepperComponent.prototype, "luxFinishButtonConfig", null);
    LuxStepperComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-stepper',
            template: "<ng-container [ngSwitch]=\"luxVerticalStepper\">\n  <!-- VERTIKALER STEPPER -->\n  <lux-stepper-vertical\n    [luxStepperConfig]=\"stepperConfiguration\"\n    (luxMatStepperLoaded)=\"matStepper = $event\"\n    (luxMatStepLabelsLoaded)=\"matStepLabels = $event\"\n    (luxStepChanged)=\"onStepChanged($event)\"\n    (luxFinButtonClicked)=\"luxFinishButtonClicked.emit($event)\"\n    (luxPrevButtonClicked)=\"stepperService.previousStep(this)\"\n    (luxNextButtonClicked)=\"stepperService.nextStep(this)\"\n    (luxCheckValidation)=\"checkValidation()\"\n    ngClass.sm=\"lux-stepper-mobile\"\n    ngClass.xs=\"lux-stepper-mobile\"\n    *ngSwitchCase=\"true\"\n  >\n  </lux-stepper-vertical>\n  <!-- HORIZONTALER STEPPER -->\n  <lux-stepper-horizontal\n    [luxStepperConfig]=\"stepperConfiguration\"\n    (luxMatStepperLoaded)=\"matStepper = $event\"\n    (luxMatStepLabelsLoaded)=\"matStepLabels = $event\"\n    (luxStepChanged)=\"onStepChanged($event)\"\n    (luxFinButtonClicked)=\"luxFinishButtonClicked.emit($event)\"\n    (luxPrevButtonClicked)=\"stepperService.previousStep(this)\"\n    (luxNextButtonClicked)=\"stepperService.nextStep(this)\"\n    (luxCheckValidation)=\"checkValidation()\"\n    ngClass.sm=\"lux-stepper-mobile\"\n    ngClass.xs=\"lux-stepper-mobile\"\n    *ngSwitchDefault\n  >\n  </lux-stepper-horizontal>\n</ng-container>\n\n<ng-content select=\"lux-step\"></ng-content>\n<div class=\"lux-stepper-disabled-overlay\" [class.lux-hidden]=\"!luxDisabled\"></div>\n",
            styles: ["@charset \"UTF-8\";lux-side-nav .lux-side-nav-item,lux-side-nav lux-link{box-sizing:border-box}lux-side-nav .lux-side-nav-item.lux-side-nav-item-selected,lux-side-nav lux-link.lux-side-nav-item-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}lux-side-nav .lux-side-nav-item.lux-side-nav-item-selected:after,lux-side-nav lux-link.lux-side-nav-item-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}lux-side-nav .lux-side-nav-item button .mat-button-focus-overlay,lux-side-nav lux-link button .mat-button-focus-overlay{display:none}lux-side-nav .lux-side-nav-item button:focus,lux-side-nav lux-link button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-side-nav .lux-side-nav-item button:hover,lux-side-nav lux-link button:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-list{border:4px solid transparent}lux-list:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-list lux-list-item:focus{outline-width:0}lux-list lux-list-item:focus mat-card{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-list lux-list-item lux-card.lux-list-item-selected mat-card{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}lux-list lux-list-item lux-card.lux-list-item-selected mat-card:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}lux-list lux-list-item lux-card mat-card{box-sizing:border-box}lux-list lux-list-item lux-card mat-card:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .lux-form-control-label>*{color:#335c85}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .cdk-focused .mat-radio-container,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-checkbox-inner-container,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-slide-toggle-thumb{position:relative}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .cdk-focused .mat-radio-container:after,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-checkbox-inner-container:after,lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .mat-slide-toggle-thumb:after{content:\"\";border:1px solid rgba(112,141,170,.2);background-color:rgba(112,141,170,.2);display:block;top:-4px;left:-4px;right:-4px;bottom:-4px;border-radius:15px;z-index:1;position:absolute}lux-form-control .lux-form-control:not(.lux-form-control-disabled).lux-focused .lux-form-control-container:after{content:'';display:block;box-sizing:border-box;left:0;right:0;position:absolute;height:2px;border-bottom:2px solid #0073b3}lux-datepicker mat-datepicker-toggle button:focus:after{content:\"\";border:1px solid rgba(112,141,170,.2);background-color:rgba(112,141,170,.2);display:block;top:-4px;left:-4px;right:-4px;bottom:-4px;border-radius:15px;z-index:1;position:absolute}.lux-datepicker-panel .mat-calendar-header button .mat-button-focus-overlay,lux-datepicker mat-datepicker-toggle .mat-button-focus-overlay{display:none}.lux-datepicker-panel .mat-calendar-header button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-datepicker-panel .mat-calendar-header button:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}.lux-datepicker-panel .mat-calendar-content td.mat-calendar-body-cell:hover>div{background-color:rgba(0,0,0,.2)!important;border-radius:999px}.lux-datepicker-panel .mat-calendar-content td.mat-calendar-body-cell.mat-calendar-body-active>div{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:999px}lux-table tr:hover>td,lux-table tr:hover>th{border-radius:0;background-color:#cfcfcf!important}lux-table th button.mat-sort-header-button:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-select-panel mat-option{border-left:0;box-sizing:border-box}.lux-select-panel mat-option.mat-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}.lux-select-panel mat-option.mat-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}.lux-select-panel mat-option:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-select-panel mat-option.mat-active,.lux-select-panel mat-option:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-button button:not([disabled]) .mat-button-focus-overlay{display:none}lux-button button:not([disabled]):not(.mat-fab):not(.mat-raised-button):focus{background-color:rgba(112,141,170,.2)!important}lux-button button:not([disabled]):not(.mat-fab):not(.mat-raised-button):hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-button button:not([disabled]).mat-fab:focus,lux-button button:not([disabled]).mat-raised-button:focus{background-color:rgba(112,141,170,.2)!important}lux-button button:not([disabled]).mat-fab:focus.mat-accent,lux-button button:not([disabled]).mat-raised-button:focus.mat-accent{background-color:#b6d184!important}lux-button button:not([disabled]).mat-fab:focus.mat-warn,lux-button button:not([disabled]).mat-raised-button:focus.mat-warn{background-color:#ea2524!important}lux-button button:not([disabled]).mat-fab:focus.mat-primary,lux-button button:not([disabled]).mat-raised-button:focus.mat-primary{background-color:#4b82ba!important}.lux-autocomplete-panel mat-option{border-left:0;box-sizing:border-box}.lux-autocomplete-panel mat-option:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-autocomplete-panel mat-option.mat-active,.lux-autocomplete-panel mat-option:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-tabs mat-ink-bar{background-color:#0073b3!important}lux-tabs .mat-tab-label:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-panel mat-expansion-panel-header:hover,lux-stepper mat-step-header:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-panel mat-expansion-panel-header:focus,lux-stepper mat-step-header:focus,lux-stepper mat-step-header[aria-selected=true],lux-tile mat-card:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}lux-master-detail lux-master-header .lux-master-header-collapse button.lux-button:not(.mat-fab):not(.mat-raised-button):focus{background-color:rgba(0,0,0,.2)!important}lux-file-list .lux-file-list{border:4px solid transparent}lux-file-list .lux-file-list:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-file-list .lux-file-list .lux-file-list-entry{border:4px solid transparent}lux-file-list .lux-file-list .lux-file-list-entry:focus{outline-width:0;border:4px solid rgba(112,141,170,.2)!important}lux-file-list .lux-file-list mat-card{margin:4px}.lux-menu-panel .lux-menu-item:not([disabled=true]):focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}.lux-menu-panel .lux-menu-item:not([disabled=true]):hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}lux-dialog-structure lux-dialog-content:focus,lux-dialog-structure lux-dialog-title:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}example-root .example-base-components-list .example-component-list-item,example-root .example-base-components-list mat-expansion-panel-header{box-sizing:border-box}example-root .example-base-components-list .example-component-list-item.example-component-list-item-selected,example-root .example-base-components-list mat-expansion-panel-header.example-component-list-item-selected{background-color:rgba(112,141,170,.2)!important;position:relative;border-radius:4px}example-root .example-base-components-list .example-component-list-item.example-component-list-item-selected:after,example-root .example-base-components-list mat-expansion-panel-header.example-component-list-item-selected:after{content:'';left:0;width:6px;top:0;bottom:0;position:absolute;border-left:6px solid #0073b3;border-radius:4px 0 0 4px}example-root .example-base-components-list .example-component-list-item:focus,example-root .example-base-components-list mat-expansion-panel-header:focus{outline-width:0!important;outline-style:unset;background-color:rgba(112,141,170,.2)!important;border-radius:4px}example-root .example-base-components-list .example-component-list-item:hover,example-root .example-base-components-list mat-expansion-panel-header:hover{background-color:rgba(0,0,0,.2)!important;border-radius:4px}.lux-app-container,body,html{height:100%}.lux-app-container{-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;box-sizing:border-box;display:-webkit-box;display:-ms-flexbox;display:flex}body{-webkit-overflow-scrolling:touch}div{font-family:Roboto,Helvetica Neue,sans-serif}.cdk-overlay-backdrop.cdk-overlay-backdrop-showing.lux-file-preview-backdrop{background:#000;opacity:.8!important}lux-card.lux-card-grow{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-flex:1;-ms-flex:1 1 auto;flex:1 1 auto;width:400px;min-width:0}lux-card.lux-card-grow mat-card,lux-card.lux-card-grow mat-card-content{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-flex:1;-ms-flex:1 1 auto;flex:1 1 auto;min-width:0}.lux-crop{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.lux-label{color:rgba(0,0,0,.54);font-size:75%}.lux-hint{color:rgba(0,0,0,.6);font-size:12px}.lux-display-none{display:none}.lux-hidden{visibility:hidden}.lux-nowrap{white-space:nowrap}.lux-cursor{cursor:pointer}.lux-block-pointer-events{pointer-events:none}.lux-vertical-align-baseline{vertical-align:baseline!important}.lux-black-semi-transparent{color:rgba(0,0,0,.87)}.lux-semi-transparent{opacity:.6}.lux-overflow-y-hidden{overflow-y:hidden}.lux-overflow-y-auto{overflow-y:auto}.lux-overflow-wrap-break-word{overflow-wrap:break-word}.lux-bold{font-weight:700!important}.lux-uppercase{text-transform:uppercase!important}.lux-lowercase{text-transform:lowercase!important}.lux-hyphenate{overflow-wrap:break-word;word-wrap:break-word;word-break:normal;-webkit-hyphens:auto;-ms-hyphens:auto;hyphens:auto;overflow-x:auto}.lux-p-footer,.lux-p-header{padding:8px}.lux-snackbar{top:68px;right:8px;position:absolute;margin:0!important;min-width:unset!important;box-shadow:0 3px 1px -2px rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.14),0 1px 5px 0 rgba(0,0,0,.12)}.mat-snack-bar-handset .lux-snackbar{left:8px;width:auto!important}.lux-mt-0{margin-top:0}.lux-pt-0{padding-top:0}.lux-mt-1{margin-top:2px}.lux-pt-1{padding-top:2px}.lux-mt-2{margin-top:4px}.lux-pt-2{padding-top:4px}.lux-mt-3{margin-top:8px}.lux-pt-3{padding-top:8px}.lux-mt-4{margin-top:16px}.lux-pt-4{padding-top:16px}.lux-mb-0{margin-bottom:0}.lux-pb-0{padding-bottom:0}.lux-mb-1{margin-bottom:2px}.lux-pb-1{padding-bottom:2px}.lux-mb-2{margin-bottom:4px}.lux-pb-2{padding-bottom:4px}.lux-mb-3{margin-bottom:8px}.lux-pb-3{padding-bottom:8px}.lux-mb-4{margin-bottom:16px}.lux-pb-4{padding-bottom:16px}.lux-ml-0{margin-left:0}.lux-pl-0{padding-left:0}.lux-ml-1{margin-left:2px}.lux-pl-1{padding-left:2px}.lux-ml-2{margin-left:4px}.lux-pl-2{padding-left:4px}.lux-ml-3{margin-left:8px}.lux-pl-3{padding-left:8px}.lux-ml-4{margin-left:16px}.lux-pl-4{padding-left:16px}.lux-m-0{margin:0}.lux-mr-0{margin-right:0}.lux-p-0{padding:0}.lux-pr-0{padding-right:0}.lux-m-1{margin:2px}.lux-mr-1{margin-right:2px}.lux-p-1{padding:2px}.lux-pr-1{padding-right:2px}.lux-m-2{margin:4px}.lux-mr-2{margin-right:4px}.lux-p-2{padding:4px}.lux-pr-2{padding-right:4px}.lux-m-3{margin:8px}.lux-mr-3{margin-right:8px}.lux-p-3{padding:8px}.lux-pr-3{padding-right:8px}.lux-m-4{margin:16px}.lux-mr-4{margin-right:16px}.lux-p-4{padding:16px}.lux-pr-4{padding-right:16px}.lux-flex-layout-gap-10[style*=row]>:not(:last-child){margin-right:10px}.lux-flex-layout-gap-10[style*=column]>:not(:last-child){margin-bottom:10px}.lux-flex-layout-gap-20[style*=row]>:not(:last-child){margin-right:20px}.lux-flex-layout-gap-20[style*=column]>:not(:last-child){margin-bottom:20px}.lux-flex-layout-gap-30[style*=row]>:not(:last-child){margin-right:30px}.lux-flex-layout-gap-30[style*=column]>:not(:last-child){margin-bottom:30px}.lux-z-index-100{z-index:100}.lux-z-index-500{z-index:500}.lux-z-index-1000{z-index:1000}.lux-uppercase>lux-button .lux-button-label,lux-button.lux-uppercase .lux-button-label{text-transform:uppercase}.lux-bg-color-red{background:#b01211}.lux-bg-color-green{background:#95bc4c}.lux-bg-color-yellow{background:#ebc000}.lux-bg-color-blue{background:#335c85}.lux-bg-color-gray{background:#c8c8c8}.lux-bg-color-orange{background:#ec9c1b}.lux-bg-color-brown{background:#783f04}.lux-font-color-black{color:#000}.lux-font-color-white{color:#fff}.lux-badge-notification.mat-badge-large .mat-badge-content{font-size:15px}.lux-badge-notification.lux-badge-color-default .mat-badge-content{color:#000;background:#fff;border:1px solid rgba(0,0,0,.87)}:host{position:relative;display:block}:host ::ng-deep .lux-stepper-content-section,:host ::ng-deep .mat-horizontal-stepper-content[aria-expanded=true],:host ::ng-deep .mat-vertical-stepper-content{padding:8px}:host ::ng-deep .mat-horizontal-stepper-header-container{padding:8px 8px 0}:host ::ng-deep .mat-vertical-stepper-header{margin:8px 8px 0}:host ::ng-deep mat-step-header .mat-step-icon,:host ::ng-deep mat-step-header .mat-step-icon-not-touched{width:40px;height:40px;font-size:2em;box-sizing:border-box}:host ::ng-deep mat-step-header .mat-step-icon .mat-icon{font-size:32px;width:32px;height:32px}:host ::ng-deep mat-step-header .mat-step-label{height:100%;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;position:relative;font-weight:400}:host ::ng-deep mat-step-header lux-step-header{color:rgba(0,0,0,.38)!important}:host ::ng-deep mat-step-header .mat-step-icon{background:rgba(0,0,0,.38);color:#fff}:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit+.mat-step-label lux-step-header,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] lux-step-header{color:rgba(0,0,0,.87)!important}:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit+.mat-step-label .lux-stepper-edited-icon i,:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit+.mat-step-label .lux-stepper-edited-icon mat-icon,:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit+.mat-step-label .lux-stepper-normal-icon i,:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit+.mat-step-label .lux-stepper-normal-icon mat-icon,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] .lux-stepper-edited-icon i,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] .lux-stepper-edited-icon mat-icon,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] .lux-stepper-normal-icon i,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] .lux-stepper-normal-icon mat-icon{background-color:#335c85}:host ::ng-deep mat-step-header.lux-step-header-touched .mat-step-icon-state-edit,:host ::ng-deep mat-step-header.lux-step-header-touched[aria-selected=true] .mat-step-icon-state-number{background:#335c85;color:#fff}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header .mat-step-icon,:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header .mat-step-icon-not-touched{display:none}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon{margin-right:8px!important}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon>.lux-material-icon,:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon>i{font-size:25px;width:32px;height:32px;text-align:center}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon.lux-stepper-edited-icon{display:none}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon.lux-stepper-normal-icon{display:block}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon>i,:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon>mat-icon{background-color:rgba(0,0,0,.38);color:#fff}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header lux-icon>i:before{padding-top:2px;display:block}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header .mat-step-icon-state-edit+.mat-step-label lux-icon.lux-stepper-edited-icon{display:block}:host ::ng-deep .lux-ignore-mat-step-icons ::ng-deep mat-step-header .mat-step-icon-state-edit+.mat-step-label lux-icon.lux-stepper-normal-icon{display:none}:host ::ng-deep .mat-step-optional{position:absolute;bottom:15px;right:0;font-style:italic}:host ::ng-deep lux-step-content>lux-card>mat-card{margin:0!important}.lux-stepper-disabled-overlay{position:absolute;top:0;left:0;right:0;bottom:0;background-color:rgba(0,0,0,.04)}.lux-stepper-mobile ::ng-deep .lux-stepper-content-section,.lux-stepper-mobile ::ng-deep .mat-horizontal-stepper-content[aria-expanded=true],.lux-stepper-mobile ::ng-deep .mat-vertical-stepper-content{padding:4px}.lux-stepper-mobile ::ng-deep .mat-vertical-stepper-content{padding-left:8px}.lux-stepper-mobile ::ng-deep .mat-horizontal-stepper-header-container{padding:4px 4px 0}.lux-stepper-mobile ::ng-deep .mat-vertical-stepper-header{margin:4px 4px 0}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper lux-step-header{display:none}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper .mat-stepper-horizontal-line{margin:0;min-width:16px}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper mat-step-header{padding:0 8px}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper.lux-ignore-mat-step-icons .mat-step-label{-webkit-box-pack:center;-ms-flex-pack:center;justify-content:center;min-width:40px}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper.lux-ignore-mat-step-icons .mat-step-label .lux-stepper-edited-icon,.lux-stepper-mobile ::ng-deep mat-horizontal-stepper.lux-ignore-mat-step-icons .mat-step-label .lux-stepper-normal-icon{margin:0!important}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper:not(.lux-ignore-mat-step-icons) .mat-step-label{display:none}.lux-stepper-mobile ::ng-deep mat-horizontal-stepper:not(.lux-ignore-mat-step-icons) .mat-step-icon{margin:0!important}"]
        }),
        tslib_1.__metadata("design:paramtypes", [LuxStepperHelperService,
            ChangeDetectorRef,
            ComponentFactoryResolver,
            ElementRef])
    ], LuxStepperComponent);
    return LuxStepperComponent;
}());
export { LuxStepperComponent };
