// @ihk-gfi/lux-components/esm5/src/app/modules/lux-file-preview/lux-file-preview-notsupportedviewer/lux-file-preview-notsupportedviewer.component.js
import * as tslib_1 from "tslib";
import { Component, Inject } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { LuxFilePreviewBase } from '../lux-file-preview-base/lux-file-preview-base';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
import { LuxFilePreviewRef } from '../lux-file-preview-ref';
var LuxFilePreviewNotSupportedViewerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LuxFilePreviewNotSupportedViewerComponent, _super);
    function LuxFilePreviewNotSupportedViewerComponent(previewRef, previewData, sanitizer) {
        var _this = _super.call(this, previewRef, previewData, sanitizer) || this;
        _this.previewRef = previewRef;
        _this.previewData = previewData;
        _this.sanitizer = sanitizer;
        _this.counter = 5;
        _this.downloadLabelDefault = 'Download... ';
        _this.downloadLabel = _this.downloadLabelDefault + _this.counter;
        return _this;
    }
    LuxFilePreviewNotSupportedViewerComponent.prototype.ngOnInit = function () {
        this.updateDownloadLabel();
    };
    LuxFilePreviewNotSupportedViewerComponent.prototype.onDownload = function () {
        clearTimeout(this.timer);
        _super.prototype.onDownload.call(this);
    };
    LuxFilePreviewNotSupportedViewerComponent.prototype.onClose = function () {
        clearTimeout(this.timer);
        _super.prototype.onClose.call(this);
    };
    LuxFilePreviewNotSupportedViewerComponent.prototype.updateDownloadLabel = function () {
        var _this = this;
        this.timer = setTimeout(function () {
            if (_this.counter > 0) {
                _this.counter--;
                _this.downloadLabel = _this.downloadLabelDefault + _this.counter;
                _this.updateDownloadLabel();
            }
            else {
                _this.onDownload();
            }
        }, 1000);
    };
    LuxFilePreviewNotSupportedViewerComponent.ctorParameters = function () { return [
        { type: LuxFilePreviewRef },
        { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] },
        { type: DomSanitizer }
    ]; };
    LuxFilePreviewNotSupportedViewerComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-file-preview-notsupportedviewer',
            template: "<lux-file-preview-toolbar>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-download\"\n    (luxClicked)=\"onDownload()\"\n    luxColor=\"accent\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-download-btn\"\n    luxAriaLabel=\"Datei herunterladen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-window-close\"\n    (luxClicked)=\"onClose()\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-close-btn\"\n    luxAriaLabel=\"Dateivorschau schlie\u00DFen\"\n  ></lux-button>\n</lux-file-preview-toolbar>\n<div class=\"message-box\" fxLayout=\"column\" fxLayoutAlign=\"center center\" fxLayoutGap=\"20px\">\n  <div>Diese Datei kann nicht angezeigt werden.</div>\n  <lux-button [luxLabel]=\"downloadLabel\" (luxClicked)=\"onDownload()\" [luxRaised]=\"true\" luxColor=\"accent\"></lux-button>\n</div>\n",
            styles: [""]
        }),
        tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
        tslib_1.__metadata("design:paramtypes", [LuxFilePreviewRef, Object, DomSanitizer])
    ], LuxFilePreviewNotSupportedViewerComponent);
    return LuxFilePreviewNotSupportedViewerComponent;
}(LuxFilePreviewBase));
export { LuxFilePreviewNotSupportedViewerComponent };
