// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-app-footer/lux-app-footer-button.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
let LuxAppFooterButtonService = class LuxAppFooterButtonService {
    constructor() {
        this._buttonInfos = [];
    }
    get buttonInfos() {
        return this._buttonInfos;
    }
    set buttonInfos(buttonInfos) {
        this._buttonInfos = buttonInfos ? buttonInfos : [];
    }
    pushButtonInfos(...value) {
        if (!this.buttonInfos) {
            this._buttonInfos = [];
        }
        this._buttonInfos.push(...value);
    }
    getButtonInfoByCMD(cmd) {
        return this.buttonInfos.find((buttonInfo) => buttonInfo.cmd === cmd);
    }
    removeButtonInfoAtIndex(i) {
        this._buttonInfos = this._buttonInfos.filter((info, index) => index !== i);
    }
    removeButtonInfoByCmd(cmd) {
        this._buttonInfos = this._buttonInfos.filter(info => info.cmd !== cmd);
    }
    clearButtonInfos() {
        this._buttonInfos = [];
    }
    sendButtonCommand(buttonCommand) {
        // Den angeklickten Button erhalten und dessen onClick-Funktion aufrufen.
        // Als Parameter den Button selbst mitgeben (falls dieser bearbeitet werden muss, z.B.).
        const clickedButton = this.getButtonInfoByCMD(buttonCommand);
        if (clickedButton) {
            clickedButton.onClick(clickedButton);
        }
    }
};
LuxAppFooterButtonService = tslib_1.__decorate([
    Injectable()
], LuxAppFooterButtonService);
export { LuxAppFooterButtonService };
