// @ihk-gfi/lux-components/esm5/src/app/modules/lux-common/lux-message-box/lux-message-box-subcomponents/lux-message.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LuxUtil } from '../../../lux-util/lux-util';
var LuxMessageComponent = /** @class */ (function () {
    function LuxMessageComponent() {
        this.backgroundCSSClass = 'lux-bg-color-blue';
        this.fontCSSClass = 'lux-font-color-white';
        this.luxMessageClosed = new EventEmitter();
    }
    Object.defineProperty(LuxMessageComponent.prototype, "luxMessage", {
        get: function () {
            return this._luxMessage;
        },
        set: function (message) {
            this._luxMessage = message;
            if (this.luxMessage) {
                this.updateColor();
            }
        },
        enumerable: true,
        configurable: true
    });
    LuxMessageComponent.prototype.ngOnInit = function () { };
    /**
     * Setzt die Messages auf ein leeres Array um so die MessageBox auszublenden.
     */
    LuxMessageComponent.prototype.close = function () {
        this.luxMessageClosed.emit(this.luxMessage);
    };
    /**
     * Aktualisiert die Farbe dieser Box passend zur Farbe der Nachricht.
     */
    LuxMessageComponent.prototype.updateColor = function () {
        var color = this.luxMessage.color;
        var result = LuxUtil.getColorsByBgColorsEnum(color);
        this.fontCSSClass = result.fontCSSClass;
        this.backgroundCSSClass = result.backgroundCSSClass;
    };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxMessageComponent.prototype, "luxMessageClosed", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], LuxMessageComponent.prototype, "luxMessage", null);
    LuxMessageComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-message',
            template: "<div\n  class=\"lux-message-container {{ backgroundCSSClass }} {{ fontCSSClass }}\"\n  fxLayout=\"row\"\n  fxLayoutAlign=\"start center\"\n  luxAriaRole=\"listitem\"\n  [luxAriaLabel]=\"luxMessage.text\"\n  tabindex=\"0\"\n  *ngIf=\"luxMessage\"\n>\n  <div class=\"lux-message-icon\" fxFlex=\"0 0 50px\" fxLayoutAlign=\"start center\">\n    <lux-icon [luxIconName]=\"luxMessage.iconName\" *ngIf=\"luxMessage.iconName\"></lux-icon>\n  </div>\n  <div fxFlex=\"100\" fxFlexFill fxLayoutAlign=\"space-between center\">\n    <div class=\"lux-message-text\" fxFlex=\"1 1 auto\" fxLayoutAlign=\"start\">\n      <p>{{ luxMessage.text }}</p>\n    </div>\n    <lux-button\n      class=\"lux-close-message-btn\"\n      luxIconName=\"fas fa-times\"\n      luxTagId=\"lux-message-box-close-btn\"\n      (luxClicked)=\"close()\"\n      luxAriaLabel=\"Nachricht schlie\u00DFen\"\n    >\n    </lux-button>\n  </div>\n</div>\n"
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxMessageComponent);
    return LuxMessageComponent;
}());
export { LuxMessageComponent };
