// @ihk-gfi/lux-components/esm5/src/app/modules/lux-pipes/lux-relative-timestamp/lux-relative-timestamp.pipe.js
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
export var day = 'Tagen', week = 'Wochen', month = 'Monaten', year = 'Jahren';
export var today = 'Heute', yesterday = 'Gestern', tomorrow = 'Morgen';
export var prefixFuture = 'in', prefixPast = 'vor';
export var MS_PER_DAY = 1000 * 60 * 60 * 24;
export var timeDeltas = [
    { name: year, days: 672, dayUnit: 336 },
    { name: month, days: 56, dayUnit: 28 },
    { name: week, days: 14, dayUnit: 7 },
    { name: day, days: 2, dayUnit: 1 }
];
export var timeDeltasRelative = [
    { name: tomorrow, days: 1 },
    { name: yesterday, days: -1 },
    { name: today, days: 0 }
];
var LuxRelativeTimestampPipe = /** @class */ (function () {
    function LuxRelativeTimestampPipe() {
    }
    LuxRelativeTimestampPipe.prototype.transform = function (timestamp, defaultText, prefix) {
        var e_1, _a, e_2, _b;
        if (defaultText === void 0) { defaultText = ''; }
        if (!timestamp) {
            return defaultText;
        }
        var now = new Date();
        var then = new Date(timestamp);
        var delta = this.calcDiff(now, then);
        var timeName = null;
        try {
            for (var timeDeltas_1 = tslib_1.__values(timeDeltas), timeDeltas_1_1 = timeDeltas_1.next(); !timeDeltas_1_1.done; timeDeltas_1_1 = timeDeltas_1.next()) {
                var timeDelta = timeDeltas_1_1.value;
                var tempDelta = delta < 0 ? delta * -1 : delta;
                if (tempDelta >= timeDelta.days) {
                    if (!prefix) {
                        prefix = delta < 0 ? prefixPast : prefixFuture;
                    }
                    var timeUnits = timeDelta.name === day ? tempDelta : Math.floor(tempDelta / timeDelta.dayUnit);
                    timeUnits *= timeUnits < 0 ? -1 : 1;
                    timeName = prefix + " " + timeUnits + " " + timeDelta.name;
                    break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (timeDeltas_1_1 && !timeDeltas_1_1.done && (_a = timeDeltas_1.return)) _a.call(timeDeltas_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (timeName === null) {
            try {
                for (var timeDeltasRelative_1 = tslib_1.__values(timeDeltasRelative), timeDeltasRelative_1_1 = timeDeltasRelative_1.next(); !timeDeltasRelative_1_1.done; timeDeltasRelative_1_1 = timeDeltasRelative_1.next()) {
                    var timeDelta = timeDeltasRelative_1_1.value;
                    if (delta === timeDelta.days) {
                        timeName = timeDelta.name;
                        break;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (timeDeltasRelative_1_1 && !timeDeltasRelative_1_1.done && (_b = timeDeltasRelative_1.return)) _b.call(timeDeltasRelative_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return timeName;
    };
    LuxRelativeTimestampPipe.prototype.calcDiff = function (a, b) {
        var utc1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
        var utc2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());
        return Math.floor((utc2 - utc1) / MS_PER_DAY);
    };
    LuxRelativeTimestampPipe = tslib_1.__decorate([
        Pipe({
            name: 'luxRelativeTimestamp'
        })
    ], LuxRelativeTimestampPipe);
    return LuxRelativeTimestampPipe;
}());
export { LuxRelativeTimestampPipe };
