// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-stepper/lux-stepper-subcomponents/lux-stepper-nav-buttons/lux-stepper-nav-buttons.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LuxStepComponent } from '../lux-step.component';
var LuxStepperNavButtonsComponent = /** @class */ (function () {
    function LuxStepperNavButtonsComponent() {
        this.luxPrevClick = new EventEmitter();
        this.luxNextClick = new EventEmitter();
        this.luxFinClick = new EventEmitter();
    }
    LuxStepperNavButtonsComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperNavButtonsComponent.prototype, "luxPrevClick", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperNavButtonsComponent.prototype, "luxNextClick", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxStepperNavButtonsComponent.prototype, "luxFinClick", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], LuxStepperNavButtonsComponent.prototype, "luxIndex", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", LuxStepComponent)
    ], LuxStepperNavButtonsComponent.prototype, "luxStep", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxStepperNavButtonsComponent.prototype, "luxStepperConfig", void 0);
    LuxStepperNavButtonsComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-stepper-nav-buttons',
            template: "<div class=\"lux-mt-3\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n  <!-- Previous-Button -->\n  <lux-button\n    [luxLabel]=\"luxStepperConfig.luxPreviousButtonConfig?.label\"\n    [luxColor]=\"luxStepperConfig.luxPreviousButtonConfig?.color\"\n    [luxIconName]=\"luxStepperConfig.luxPreviousButtonConfig?.iconName\"\n    [luxIconAlignWithLabel]=\"luxStepperConfig.luxPreviousButtonConfig?.alignIconWithLabel\"\n    (luxClicked)=\"luxPrevClick.emit()\"\n    luxTagId=\"luxStepper.previousStep\"\n    *ngIf=\"luxIndex > 0\"\n  >\n  </lux-button>\n  <!-- Next-Button -->\n  <lux-button\n    [luxLabel]=\"luxStepperConfig.luxNextButtonConfig?.label\"\n    [luxColor]=\"luxStepperConfig.luxNextButtonConfig?.color\"\n    [luxIconName]=\"luxStepperConfig.luxNextButtonConfig?.iconName\"\n    [luxIconAlignWithLabel]=\"luxStepperConfig.luxNextButtonConfig?.alignIconWithLabel\"\n    (luxClicked)=\"luxNextClick.emit()\"\n    [luxDisabled]=\"!luxStep.isCompleted() && luxStepperConfig.luxLinear\"\n    luxTagId=\"luxStepper.nextStep\"\n    *ngIf=\"luxIndex < luxStepperConfig.luxSteps.length - 1\"\n  >\n  </lux-button>\n  <!-- Finish-Button -->\n  <lux-button\n    [luxLabel]=\"luxStepperConfig.luxFinishButtonConfig?.label\"\n    [luxColor]=\"luxStepperConfig.luxFinishButtonConfig?.color\"\n    [luxIconName]=\"luxStepperConfig.luxFinishButtonConfig?.iconName\"\n    [luxIconAlignWithLabel]=\"luxStepperConfig.luxFinishButtonConfig?.alignIconWithLabel\"\n    (luxClicked)=\"luxFinClick.emit()\"\n    [luxDisabled]=\"!luxStep.isCompleted() && luxStepperConfig.luxLinear\"\n    luxTagId=\"luxStepper.finished\"\n    *ngIf=\"luxIndex === luxStepperConfig.luxSteps.length - 1\"\n  >\n  </lux-button>\n</div>\n",
            styles: [""]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxStepperNavButtonsComponent);
    return LuxStepperNavButtonsComponent;
}());
export { LuxStepperNavButtonsComponent };
