// @ihk-gfi/lux-components/esm5/src/app/modules/lux-lookup/lux-lookup-model/lux-lookup-parameters.js
/**
 * Enums, welche die auswaehlbaren Felder fuer den Response einer Lookup-Service-Abfrage beinhaltet.
 */
export var LuxFieldValues;
(function (LuxFieldValues) {
    LuxFieldValues["kurz"] = "kurz";
    LuxFieldValues["lang1"] = "lang1";
    LuxFieldValues["lang2"] = "lang2";
    LuxFieldValues["gueltig_von"] = "gueltig_von";
    LuxFieldValues["gueltig_bis"] = "gueltig_bis";
})(LuxFieldValues || (LuxFieldValues = {}));
export var LuxBehandlungsOptionenUngueltige;
(function (LuxBehandlungsOptionenUngueltige) {
    LuxBehandlungsOptionenUngueltige["anzeigen"] = "anzeigen";
    LuxBehandlungsOptionenUngueltige["ausgrauen"] = "ausgrauen";
    LuxBehandlungsOptionenUngueltige["ausblenden"] = "ausblenden";
})(LuxBehandlungsOptionenUngueltige || (LuxBehandlungsOptionenUngueltige = {}));
/**
 * Class, welche benutzt wird um die Abfrage an den Lookup-Service zu modifizeren.
 */
var LuxLookupParameters = /** @class */ (function () {
    function LuxLookupParameters(data) {
        this.knr = data.knr;
        this.keys = !!data.keys ? data.keys : [];
        this.fields = !!data.fields
            ? data.fields
            : [
                LuxFieldValues.kurz,
                LuxFieldValues.lang1,
                LuxFieldValues.lang2,
                LuxFieldValues.gueltig_bis,
                LuxFieldValues.gueltig_von
            ];
        this.raw = !!data.raw ? data.raw : false;
    }
    return LuxLookupParameters;
}());
export { LuxLookupParameters };
