// @ihk-gfi/lux-components/esm5/src/app/modules/lux-common/lux-table/lux-table-subcomponents/lux-table-column.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, Input } from '@angular/core';
import { LuxTableColumnHeaderComponent } from './lux-table-column-header.component';
import { LuxTableColumnContentComponent } from './lux-table-column-content.component';
import { LuxTableColumnFooterComponent } from './lux-table-column-footer.component';
import { Subject } from 'rxjs';
var LuxTableColumnComponent = /** @class */ (function () {
    function LuxTableColumnComponent() {
        this.change$ = new Subject();
        this.luxColumnDef = undefined;
        this.luxSortable = false;
        this.luxSticky = false;
    }
    LuxTableColumnComponent.prototype.ngOnInit = function () { };
    LuxTableColumnComponent.prototype.ngOnChanges = function (simpleChanges) {
        this.change$.next();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTableColumnComponent.prototype, "luxColumnDef", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTableColumnComponent.prototype, "luxSortable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTableColumnComponent.prototype, "luxSticky", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTableColumnComponent.prototype, "luxResponsiveBehaviour", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxTableColumnComponent.prototype, "luxResponsiveAt", void 0);
    tslib_1.__decorate([
        ContentChild(LuxTableColumnHeaderComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxTableColumnHeaderComponent)
    ], LuxTableColumnComponent.prototype, "header", void 0);
    tslib_1.__decorate([
        ContentChild(LuxTableColumnContentComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxTableColumnContentComponent)
    ], LuxTableColumnComponent.prototype, "content", void 0);
    tslib_1.__decorate([
        ContentChild(LuxTableColumnFooterComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxTableColumnFooterComponent)
    ], LuxTableColumnComponent.prototype, "footer", void 0);
    LuxTableColumnComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-table-column',
            template: '<ng-content select="lux-table-column-header"></ng-content>' +
                '<ng-content select="lux-table-column-content"></ng-content>' +
                '<ng-content select="lux-table-column-footer"></ng-content>'
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxTableColumnComponent);
    return LuxTableColumnComponent;
}());
export { LuxTableColumnComponent };
