// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-file-preview/lux-file-preview.service.js
import * as tslib_1 from "tslib";
import { Overlay, OverlayConfig, OverlayRef } from '@angular/cdk/overlay';
import { ComponentPortal, PortalInjector } from '@angular/cdk/portal';
import { ComponentRef, Injectable, Injector } from '@angular/core';
import { LUX_FILE_PREVIEW_DATA } from './lux-file-preview-config';
import { LuxFilePreviewRef } from './lux-file-preview-ref';
import { LuxFilePreviewComponent } from './lux-file-preview.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
export const DEFAULT_FILE_PREVIEW_CONFIG = {
    hasBackdrop: true,
    backdropClass: 'lux-file-preview-backdrop',
    panelClass: 'lux-file-preview-panel',
    previewData: {
        fileComponent: null,
        fileObject: null
    }
};
let LuxFilePreviewService = class LuxFilePreviewService {
    constructor(injector, overlay) {
        this.injector = injector;
        this.overlay = overlay;
    }
    open(config) {
        const previewConfig = Object.assign({}, DEFAULT_FILE_PREVIEW_CONFIG, config);
        const overlayRef = this.createOverlay(previewConfig);
        const previewRef = new LuxFilePreviewRef(overlayRef);
        const previewComponent = this.attachDialogContainer(overlayRef, previewRef, previewConfig);
        overlayRef.backdropClick().subscribe(() => previewComponent.onClose());
        return previewRef;
    }
    attachDialogContainer(overlayRef, dialogRef, config) {
        const injector = this.createInjector(config, dialogRef);
        const containerPortal = new ComponentPortal(LuxFilePreviewComponent, null, injector);
        const containerRef = overlayRef.attach(containerPortal);
        return containerRef.instance;
    }
    createOverlay(config) {
        return this.overlay.create(this.getOverlayConfig(config));
    }
    createInjector(config, dialogRef) {
        const injectionTokens = new WeakMap();
        injectionTokens.set(LuxFilePreviewRef, dialogRef);
        injectionTokens.set(LUX_FILE_PREVIEW_DATA, config.previewData);
        return new PortalInjector(this.injector, injectionTokens);
    }
    getOverlayConfig(config) {
        const positionStrategy = this.overlay
            .position()
            .global()
            .centerHorizontally()
            .centerVertically();
        const overlayConfig = new OverlayConfig({
            hasBackdrop: config.hasBackdrop,
            backdropClass: config.backdropClass,
            panelClass: config.panelClass,
            scrollStrategy: this.overlay.scrollStrategies.block(),
            positionStrategy
        });
        return overlayConfig;
    }
};
LuxFilePreviewService.ctorParameters = () => [
    { type: Injector },
    { type: Overlay }
];
LuxFilePreviewService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LuxFilePreviewService_Factory() { return new LuxFilePreviewService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Overlay)); }, token: LuxFilePreviewService, providedIn: "root" });
LuxFilePreviewService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__metadata("design:paramtypes", [Injector, Overlay])
], LuxFilePreviewService);
export { LuxFilePreviewService };
