// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-lookup/lux-lookup-label/lux-lookup-label.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { LuxConsoleService } from '../../lux-util/lux-console.service';
import { LuxLookupParameters } from '../lux-lookup-model/lux-lookup-parameters';
import { LuxLookupHandlerService } from '../lux-lookup-service/lux-lookup-handler.service';
import { LuxLookupService } from '../lux-lookup-service/lux-lookup.service';
let LuxLookupLabelComponent = class LuxLookupLabelComponent {
    constructor(lookupService, lookupHandler, luxConsoleLogger) {
        this.luxLookupUrl = '/lookup/';
        this.luxBezeichnung = 'kurz';
        this.lookupService = lookupService;
        this.lookupHandler = lookupHandler;
        this.logger = luxConsoleLogger;
    }
    ngOnInit() {
        if (!this.luxLookupId) {
            console.error('Das Lookup-Label mit der Tabellen-Nr. ${this.luxTableNo} besitzt keine LookupId.' +
                'Bitte tragen Sie diese nach.');
        }
        if (!this.luxTableNo) {
            console.error('Das Lookup-Label mit der LookupId ${this.luxLookupId} besitzt keine Tabellen-Nummer. ' +
                'Bitte tragen Sie diese nach.');
        }
        if (!this.luxTableKey) {
            console.error('Das Lookup-Label mit der Tabellen-Nr. ${this.luxTableNo} besitzt keinen Tabellen-Key.' +
                'Bitte tragen Sie diese nach.');
        }
        this.fetchLookupData();
        this.lookupHandler.addLookupElement(this.luxLookupId);
        this.lookupHandler.getLookupElementObsv(this.luxLookupId).subscribe(() => {
            this.fetchLookupData();
        });
    }
    fetchLookupData() {
        const keys = [this.luxTableKey];
        this.lookupParameters = new LuxLookupParameters({ knr: this.luxLookupKnr, keys });
        this.lookupService
            .getLookupTable(this.luxTableNo, this.lookupParameters, this.luxLookupUrl)
            .subscribe((entries) => {
            if (typeof entries !== 'undefined' && entries.length === 1) {
                this.entry = entries[0];
            }
        });
    }
    /**
     * liefert die Bezeichnung (Kurz- oder Langbezeichnung) des Entries für den Key zur Tabelle.
     *
     * @returns string
     */
    getBezeichnung() {
        let bezeichnung = '';
        if (this.entry) {
            if ('kurz' === this.luxBezeichnung) {
                bezeichnung = this.entry.kurzText;
            }
            else if ('lang' === this.luxBezeichnung) {
                bezeichnung = this.entry.langText1;
                if (!bezeichnung) {
                    bezeichnung = this.entry.kurzText;
                }
            }
        }
        return bezeichnung;
    }
};
LuxLookupLabelComponent.ctorParameters = () => [
    { type: LuxLookupService },
    { type: LuxLookupHandlerService },
    { type: LuxConsoleService }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], LuxLookupLabelComponent.prototype, "luxLookupKnr", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLookupLabelComponent.prototype, "luxLookupId", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLookupLabelComponent.prototype, "luxLookupUrl", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLookupLabelComponent.prototype, "luxTableNo", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLookupLabelComponent.prototype, "luxTableKey", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLookupLabelComponent.prototype, "luxBezeichnung", void 0);
LuxLookupLabelComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-lookup-label',
        template: "<span>{{ getBezeichnung() }}</span>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [LuxLookupService,
        LuxLookupHandlerService,
        LuxConsoleService])
], LuxLookupLabelComponent);
export { LuxLookupLabelComponent };
