// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-action/lux-menu/lux-menu-subcomponents/lux-menu-item.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { LuxActionComponentBaseClass } from '../../lux-action-model/lux-action-component-base.class';
// @dynamic Erklärung steht in der Datei "lux-decorators.ts".
let LuxMenuItemComponent = class LuxMenuItemComponent extends LuxActionComponentBaseClass {
    constructor() {
        super();
        this.luxHiddenChange = new EventEmitter();
        this.luxHideLabelIfExtendedChange = new EventEmitter();
        this.luxAlwaysVisibleChange = new EventEmitter();
        this._luxAlwaysVisible = true;
        this._luxHideLabelIfExtended = false;
        this._luxHidden = false;
        // Wird vom LuxMenuComponent mit dem berechneten Breitenwert belegt
        this.width = 0;
        // Wird vom LuxMenuComponent mit dem passenden Zustand belegt
        this.extended = false;
    }
    get luxAlwaysVisible() {
        return this._luxAlwaysVisible;
    }
    set luxAlwaysVisible(value) {
        this._luxAlwaysVisible = value;
        this.luxAlwaysVisibleChange.emit(value);
    }
    get luxHideLabelIfExtended() {
        return this._luxHideLabelIfExtended;
    }
    set luxHideLabelIfExtended(value) {
        this._luxHideLabelIfExtended = value;
        this.luxHideLabelIfExtendedChange.emit(value);
    }
    get luxHidden() {
        return this._luxHidden;
    }
    set luxHidden(value) {
        this._luxHidden = value;
        this.luxHiddenChange.emit(value);
    }
    clicked($event) {
        this.luxClicked.emit($event);
    }
};
tslib_1.__decorate([
    ViewChild(TemplateRef, { static: false }),
    tslib_1.__metadata("design:type", TemplateRef)
], LuxMenuItemComponent.prototype, "templateRef", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxMenuItemComponent.prototype, "luxHiddenChange", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxMenuItemComponent.prototype, "luxHideLabelIfExtendedChange", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxMenuItemComponent.prototype, "luxAlwaysVisibleChange", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], LuxMenuItemComponent.prototype, "luxAlwaysVisible", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], LuxMenuItemComponent.prototype, "luxHideLabelIfExtended", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], LuxMenuItemComponent.prototype, "luxHidden", null);
LuxMenuItemComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-menu-item',
        template: ''
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxMenuItemComponent);
export { LuxMenuItemComponent };
