// @ihk-gfi/lux-components/esm5/src/app/modules/lux-action/lux-link/lux-link.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, OnDestroy, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { LuxActionComponentBaseClass } from '../lux-action-model/lux-action-component-base.class';
var LuxLinkComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LuxLinkComponent, _super);
    function LuxLinkComponent(router, cdr) {
        var _this = _super.call(this) || this;
        _this.router = router;
        _this.cdr = cdr;
        _this.iconSize = '2x';
        _this.luxBlank = false;
        return _this;
    }
    LuxLinkComponent.prototype.ngOnInit = function () { };
    LuxLinkComponent.prototype.ngOnDestroy = function () { };
    LuxLinkComponent.prototype.redirectToHref = function ($event) {
        this.luxClicked.emit($event);
        if (this.luxHref) {
            this.luxHref = this.luxHref.trim();
            if (!this.luxHref.startsWith('http')) {
                this.router.navigate([this.luxHref]).then(function () { });
            }
            else {
                window.open(this.luxHref, this.luxBlank ? '_blank' : '_self');
            }
        }
    };
    LuxLinkComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ChangeDetectorRef }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxLinkComponent.prototype, "luxHref", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxLinkComponent.prototype, "luxBlank", void 0);
    LuxLinkComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-link',
            template: "<lux-button\n  [luxRaised]=\"luxRaised\"\n  [luxColor]=\"luxColor\"\n  [luxLabel]=\"luxLabel\"\n  [luxIconName]=\"luxIconName\"\n  [luxIconAlignWithLabel]=\"luxIconAlignWithLabel\"\n  [luxDisabled]=\"luxDisabled\"\n  [luxRounded]=\"luxRounded\"\n  [luxTagId]=\"luxTagId\"\n  class=\"lux-link\"\n  luxAriaRole=\"link\"\n  luxAriaRoleSelector=\"button\"\n  (luxClicked)=\"redirectToHref($event)\"\n>\n</lux-button>\n",
            styles: [""]
        }),
        tslib_1.__metadata("design:paramtypes", [Router, ChangeDetectorRef])
    ], LuxLinkComponent);
    return LuxLinkComponent;
}(LuxActionComponentBaseClass));
export { LuxLinkComponent };
