// @ihk-gfi/lux-components/esm5/src/app/modules/lux-popups/lux-popups.module.js
import * as tslib_1 from "tslib";
// tslint:disable:max-line-length
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FlexLayoutModule } from '@angular/flex-layout';
import { MatButtonModule, MatDialogModule, MatSnackBarModule } from '@angular/material';
import { LuxIconModule } from '../lux-icon/lux-icon.module';
import { LuxSnackbarComponent } from './lux-snackbar/lux-snackbar-component/lux-snackbar.component';
import { LuxActionModule } from '../lux-action/lux-action.module';
import { LuxComponentsConfigModule } from '../lux-components-config/lux-components-config.module';
import { LuxDialogPresetComponent } from './lux-dialog/lux-dialog-preset/lux-dialog-preset.component';
import { LuxLayoutModule } from '../lux-layout/lux-layout.module';
import { A11yModule } from '@angular/cdk/a11y';
import { LuxDirectivesModule } from '../lux-directives/lux-directives.module';
import { LuxDialogStructureComponent } from './lux-dialog/lux-dialog-structure/lux-dialog-structure.component';
import { LuxDialogTitleComponent } from './lux-dialog/lux-dialog-structure/lux-dialog-structure-subcomponents/lux-dialog-title.component';
import { LuxDialogContentComponent } from './lux-dialog/lux-dialog-structure/lux-dialog-structure-subcomponents/lux-dialog-content.component';
import { LuxDialogActionsComponent } from './lux-dialog/lux-dialog-structure/lux-dialog-structure-subcomponents/lux-dialog-actions.component';
var LuxPopupsModule = /** @class */ (function () {
    function LuxPopupsModule() {
    }
    LuxPopupsModule = tslib_1.__decorate([
        NgModule({
            imports: [
                CommonModule,
                FlexLayoutModule,
                LuxIconModule,
                LuxActionModule,
                MatSnackBarModule,
                MatButtonModule,
                LuxComponentsConfigModule,
                LuxLayoutModule,
                MatDialogModule,
                A11yModule,
                LuxDirectivesModule
            ],
            declarations: [
                LuxSnackbarComponent,
                LuxDialogPresetComponent,
                LuxDialogStructureComponent,
                LuxDialogTitleComponent,
                LuxDialogContentComponent,
                LuxDialogActionsComponent
            ],
            exports: [
                LuxSnackbarComponent,
                LuxDialogPresetComponent,
                LuxDialogStructureComponent,
                LuxDialogTitleComponent,
                LuxDialogContentComponent,
                LuxDialogActionsComponent
            ],
            providers: [],
            entryComponents: [LuxDialogPresetComponent]
        })
    ], LuxPopupsModule);
    return LuxPopupsModule;
}());
export { LuxPopupsModule };
