// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-stepper/lux-stepper-helper.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject, of } from 'rxjs';
let LuxStepperHelperService = class LuxStepperHelperService {
    constructor() {
        this.switchStep = new Map();
    }
    /**
     * Springt im Stepper zum naechsten Step.
     * Alternativ (ohne Stepper als Parameter) laesst die Funktion alle
     * bekannten Stepper einen Schritt nach vorne springen.
     *
     * Voraussetzung: luxLinear = false oder aktuell Form ist valide.
     */
    nextStep(stepper) {
        if (stepper) {
            const subject = this.switchStep.get(stepper);
            if (subject) {
                subject.next(true);
            }
        }
        else {
            this.switchStep.forEach((switchStep) => switchStep.next(true));
        }
    }
    /**
     * Springt im Stepper zum vorherigen Step.
     * Alternativ (ohne Stepper als Parameter) laesst die Funktion alle
     * Bekannten Stepper einen Schritt zurueckspringen.
     */
    previousStep(stepper) {
        if (stepper) {
            const subject = this.switchStep.get(stepper);
            if (subject) {
                subject.next(false);
            }
        }
        else {
            this.switchStep.forEach((switchStep) => switchStep.next(false));
        }
    }
    /**
     * Gibt das Observable der internen Map fuer den Stepper zurueck.
     * @param stepper
     * @returns Observable<boolean> | Observable<any>
     */
    getObservable(stepper) {
        const subject = this.switchStep.get(stepper);
        return subject ? subject.asObservable() : of(null);
    }
    /**
     * Traegt den Stepper in die Map der dem Service bekannten Stepper ein.
     * @param stepper
     */
    registerStepper(stepper) {
        if (!this.switchStep.get(stepper)) {
            this.switchStep.set(stepper, new BehaviorSubject(false));
        }
    }
};
LuxStepperHelperService = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [])
], LuxStepperHelperService);
export { LuxStepperHelperService };
