// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-stepper/lux-stepper-subcomponents/lux-step-header.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { LuxComponentsConfigService } from '../../../lux-components-config/lux-components-config.service';
var LuxStepHeaderComponent = /** @class */ (function () {
    function LuxStepHeaderComponent(componentsConfigService) {
        this.componentsConfigService = componentsConfigService;
    }
    LuxStepHeaderComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.configSubscription = this.componentsConfigService.config.subscribe(function () {
            _this.labelUppercase = _this.componentsConfigService.isLabelUppercaseForSelector('lux-step');
        });
    };
    LuxStepHeaderComponent.prototype.ngOnDestroy = function () {
        this.configSubscription.unsubscribe();
    };
    LuxStepHeaderComponent.ctorParameters = function () { return [
        { type: LuxComponentsConfigService }
    ]; };
    LuxStepHeaderComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-step-header',
            template: "\n    <div [ngClass]=\"{ 'lux-uppercase': labelUppercase }\">\n      <ng-content></ng-content>\n    </div>\n  ",
            styles: ['']
        }),
        tslib_1.__metadata("design:paramtypes", [LuxComponentsConfigService])
    ], LuxStepHeaderComponent);
    return LuxStepHeaderComponent;
}());
export { LuxStepHeaderComponent };
