// @ihk-gfi/lux-components-update/src/lux-version-1.7/1.7.11/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const project_1 = require("@schematics/angular/utility/project");
const validation_1 = require("../../utility/validation");
const dependencies_1 = require("../../utility/dependencies");
const logging_1 = require("../../utility/logging");
const chalk_1 = require("chalk");
const util_1 = require("../../utility/util");
const files_1 = require("../../utility/files");
/**
 * Haupt-Rule für diesen Schematic-Generator.
 * @param options
 */
exports.luxVersion = (options) => {
    return schematics_1.chain([
        setupProject(options),
        checkVersions(),
        updatePackageJson(),
        renameLuxButtonProperty(options),
        renameLuxCardContentExpanded(options),
        todosForUser()
    ]);
};
/**
 * Prüft, ob die Property "project" gesetzt ist und
 * erstellt wenn nötig einen Standard-Pfad zum Projekt, wenn keiner bekannt ist.
 * @param options
 */
function setupProject(options) {
    return (tree, context) => {
        logging_1.logNewUpdate('1.7.11');
        logging_1.logInfoWithDescriptor('Starte Konfiguration der Schematic.');
        return util_1.waitForTreeCallback(tree, () => {
            if (!options.project) {
                throw logging_1.formattedSchematicsException('Option "project" wird benötigt.');
            }
            const project = project_1.getProject(tree, options.project);
            if (options.path === undefined) {
                options.path = project.root;
            }
            util_1.checkSmoketestScriptExists(tree, context);
            logging_1.logSuccess(`Schematic-Konfiguration für Projekt "${options.project}" erfolgreich.`);
            return tree;
        });
    };
}
exports.setupProject = setupProject;
/**
 * Prüft ob die Versionen des Projekts mit den erforderlichen Versionen dieses Updates übereinstimmen.
 */
function checkVersions() {
    return (tree, context) => {
        logging_1.logInfoWithDescriptor('Starte die Versionsprüfung.');
        return util_1.waitForTreeCallback(tree, () => {
            const minimumLuxComponentsVersion = '1.7.10';
            validation_1.validateLuxComponentsVersion(tree, context, minimumLuxComponentsVersion);
            const minimumNodeVersion = '8.0.0';
            validation_1.validateNodeVersion(context, minimumNodeVersion);
            logging_1.logSuccess(`Versionen erfolgreich geprüft.`);
            return tree;
        });
    };
}
exports.checkVersions = checkVersions;
/**
 * Aktualisiert die package.json des Projekts.
 */
function updatePackageJson() {
    return (tree, context) => {
        logging_1.logInfoWithDescriptor('Aktualisiere LUX-Components Version auf 1.7.11.');
        return util_1.waitForTreeCallback(tree, () => {
            const newDependency = {
                type: dependencies_1.NodeDependencyType.Default,
                version: '1.7.11',
                name: 'lux-components'
            };
            dependencies_1.updatePackageJsonDependency(tree, context, newDependency);
            logging_1.logSuccess(`package.json erfolgreich aktualisiert.`);
            return tree;
        });
    };
}
exports.updatePackageJson = updatePackageJson;
function renameLuxButtonProperty(options) {
    return (tree, context) => {
        logging_1.logInfoWithDescriptor(`Benenne die Property ${chalk_1.default.redBright('luxRoundButton')} zu ${chalk_1.default.redBright('luxRounded')} um.`);
        return util_1.waitForTreeCallback(tree, () => {
            let fileCount = 0;
            files_1.iterateFilesAndModifyContent(tree, options.path + '/src/app', (filePath, content) => {
                // Alle Vorkommnisse des Konstruktor-Aufrufs finden
                const modifiedContent = content.replace(/luxRoundButton/gm, 'luxRounded');
                // Den angepassten Content einsetzen
                if (modifiedContent !== content) {
                    tree.overwrite(filePath, modifiedContent);
                    fileCount++;
                }
            }, '.html');
            logging_1.logInfo(`${fileCount} Template-Dateien angepasst.`);
            logging_1.logSuccess(`Properties umbenannt.`);
            return tree;
        });
    };
}
exports.renameLuxButtonProperty = renameLuxButtonProperty;
function renameLuxCardContentExpanded(options) {
    return (tree, context) => {
        logging_1.logInfoWithDescriptor(`Benenne das Tag ${chalk_1.default.redBright('lux-card-content-expanded')} zu ${chalk_1.default.redBright('lux-card-content-switched')} um.`);
        return util_1.waitForTreeCallback(tree, () => {
            let fileCount = 0;
            files_1.iterateFilesAndModifyContent(tree, options.path + '/src/app', (filePath, content) => {
                // Alle Vorkommnisse des Konstruktor-Aufrufs finden
                const modifiedContent = content.replace(/lux-card-content-expanded/g, 'lux-card-content-switched');
                // Den angepassten Content einsetzen
                if (modifiedContent !== content) {
                    tree.overwrite(filePath, modifiedContent);
                    fileCount++;
                }
            }, '.html');
            logging_1.logInfo(`${fileCount} Template-Dateien angepasst.`);
            logging_1.logSuccess(`Properties umbenannt.`);
            return tree;
        });
    };
}
exports.renameLuxCardContentExpanded = renameLuxCardContentExpanded;
/**
 * Gibt die offen stehenden ToDos (Aufgaben, die der Generator nicht übernehmen konnte) für den User aus.
 */
function todosForUser() {
    return (tree, context) => {
        util_1.runInstallAndLogToDos(context, `Alle Components unter ${chalk_1.default.blueBright('LuxFormModule')} wurden auf eine Baseline gebracht.\r\n${logging_1.TAB}Bitte prüfen Sie in Ihren Views, ob die Ansichten noch korrekt sind.`, `Die Components ${chalk_1.default.blueBright('LuxTile')}, ${chalk_1.default.blueBright('LuxCard')}, ${chalk_1.default.blueBright('LuxPanel')} und ${chalk_1.default.blueBright('LuxStepper')} haben in der mobilen Ansicht ca. 50% ihrer Margins und Paddings verloren.\r\n` +
            `${logging_1.TAB}Bitte prüfen Sie sicherheitshalber Ihre Applikation in den xs- und sm-Queries.`, `Die Breite der Component ${chalk_1.default.blueBright('LuxTile')} hat in der mobilen Ansicht nun eine Breite von 140px (vorher 150px), bitte den entsprechenden Queries überprüfen.`, `Die Properties ${chalk_1.default.redBright('luxDashboardLink')}, ${chalk_1.default.redBright('luxDashboardLinkTitle')} und ${chalk_1.default.redBright('luxOpenLinkBlank')} wurden von ${chalk_1.default.blueBright('LuxAppHeader')} entfernt.`, `Bitte starten Sie ${chalk_1.default.redBright('npm run smoketest')} um möglichen Fehlern vorzugreifen.`, `Weitere Informationen: https://confluence.gfi.ihk.de/display/EVA/Update+Guide#UpdateGuide-UmstellungaufVersion1.7.11`);
        return tree;
    };
}
exports.todosForUser = todosForUser;
