// @ihk-gfi/lux-components/esm5/src/app/modules/lux-directives/lux-tooltip/lux-tooltip.directive.js
import * as tslib_1 from "tslib";
import { AriaDescriber, FocusMonitor } from '@angular/cdk/a11y';
import { Directionality } from '@angular/cdk/bidi';
import { Overlay } from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { ScrollDispatcher } from '@angular/cdk/scrolling';
import { Directive, ElementRef, HostListener, Inject, Input, NgZone, OnChanges, Optional, SimpleChanges, ViewContainerRef } from '@angular/core';
import { MAT_TOOLTIP_DEFAULT_OPTIONS, MAT_TOOLTIP_SCROLL_STRATEGY, MatTooltip } from '@angular/material';
var LuxTooltipDirective = /** @class */ (function (_super) {
    tslib_1.__extends(LuxTooltipDirective, _super);
    function LuxTooltipDirective(luxOverlay, luxElementRef, luxScrollDispatcher, luxViewContainerRef, luxNgZone, luxPlatform, luxAriaDescriber, luxFocusMonitor, luxScrollStrategy, luxDir, luxDefaultOptions) {
        var _this = _super.call(this, luxOverlay, luxElementRef, luxScrollDispatcher, luxViewContainerRef, luxNgZone, luxPlatform, luxAriaDescriber, luxFocusMonitor, luxScrollStrategy, luxDir, luxDefaultOptions) || this;
        _this.luxOverlay = luxOverlay;
        _this.luxElementRef = luxElementRef;
        _this.luxScrollDispatcher = luxScrollDispatcher;
        _this.luxViewContainerRef = luxViewContainerRef;
        _this.luxNgZone = luxNgZone;
        _this.luxPlatform = luxPlatform;
        _this.luxAriaDescriber = luxAriaDescriber;
        _this.luxFocusMonitor = luxFocusMonitor;
        _this.luxScrollStrategy = luxScrollStrategy;
        _this.luxDir = luxDir;
        _this.luxDefaultOptions = luxDefaultOptions;
        _this.luxTooltip = '???';
        _this.luxTooltipHideDelay = 0;
        _this.luxTooltipShowDelay = 0;
        _this.luxTooltipPosition = 'above';
        _this.luxTooltipDisabled = false;
        return _this;
    }
    LuxTooltipDirective.prototype._handleLongPress = function () {
        _super.prototype.show.call(this, this.luxTooltipShowDelay);
    };
    LuxTooltipDirective.prototype._handleKeydown = function (event) {
        _super.prototype._handleKeydown.call(this, event);
    };
    LuxTooltipDirective.prototype._handleTouchend = function () {
        _super.prototype._handleTouchend.call(this);
    };
    LuxTooltipDirective.prototype.ngOnChanges = function (simpleChanges) {
        this.message = this.luxTooltip;
        this.hideDelay = this.luxTooltipHideDelay;
        this.showDelay = this.luxTooltipShowDelay;
        this.position = this.luxTooltipPosition;
        this.disabled = this.luxTooltipDisabled;
    };
    LuxTooltipDirective.ctorParameters = function () { return [
        { type: Overlay },
        { type: ElementRef },
        { type: ScrollDispatcher },
        { type: ViewContainerRef },
        { type: NgZone },
        { type: Platform },
        { type: AriaDescriber },
        { type: FocusMonitor },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_TOOLTIP_SCROLL_STRATEGY,] }] },
        { type: Directionality, decorators: [{ type: Optional }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_TOOLTIP_DEFAULT_OPTIONS,] }] }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTooltipDirective.prototype, "luxTooltip", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], LuxTooltipDirective.prototype, "luxTooltipHideDelay", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], LuxTooltipDirective.prototype, "luxTooltipShowDelay", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTooltipDirective.prototype, "luxTooltipPosition", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTooltipDirective.prototype, "luxTooltipDisabled", void 0);
    tslib_1.__decorate([
        HostListener('longpress'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], LuxTooltipDirective.prototype, "_handleLongPress", null);
    tslib_1.__decorate([
        HostListener('keydown', ['$event']),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], LuxTooltipDirective.prototype, "_handleKeydown", null);
    tslib_1.__decorate([
        HostListener('touchend'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], LuxTooltipDirective.prototype, "_handleTouchend", null);
    LuxTooltipDirective = tslib_1.__decorate([
        Directive({
            selector: '[luxTooltip]',
            exportAs: 'luxTooltip'
        }),
        tslib_1.__param(8, Inject(MAT_TOOLTIP_SCROLL_STRATEGY)),
        tslib_1.__param(9, Optional()),
        tslib_1.__param(10, Optional()),
        tslib_1.__param(10, Inject(MAT_TOOLTIP_DEFAULT_OPTIONS)),
        tslib_1.__metadata("design:paramtypes", [Overlay,
            ElementRef,
            ScrollDispatcher,
            ViewContainerRef,
            NgZone,
            Platform,
            AriaDescriber,
            FocusMonitor, Object, Directionality, Object])
    ], LuxTooltipDirective);
    return LuxTooltipDirective;
}(MatTooltip));
export { LuxTooltipDirective };
