// @ihk-gfi/lux-components-update/src/utility/test-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
/**
 * Standard-Optionen für das Test-Workspace.
 */
exports.workspaceOptions = {
    name: 'workspace',
    newProjectRoot: 'projects',
    version: '6.0.0'
};
/**
 * Standard-Optionen für die Test-Applikation.
 */
exports.appOptions = {
    name: 'bar',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: 'scss',
    skipTests: false,
    skipPackageJson: false
};
/**
 * Fügt eine Dependency in die erste Zeile der Property "dependencies" aus der package.json.
 * @param appTree
 * @param dependency
 * @param version
 */
exports.addDependencyToPackageJson = (appTree, dependency, version) => {
    let packageJsonString = appTree.readContent('/package.json');
    const newPackageJsonString = packageJsonString.replace('"dependencies": {\n', `"dependencies": {\n\t"${dependency}": "${version}",\n`);
    appTree.overwrite('/package.json', newPackageJsonString);
};
/**
 * Erzeugt einen Mock-Context für den Aufruf über callRule.
 */
exports.createMockContext = () => {
    return {
        strategy: schematics_1.MergeStrategy.Default,
        logger: {
            info: (...args) => {
            },
            warn: (...args) => {
            },
            error: (...args) => {
            },
        }
    };
};
