// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-action/lux-link/lux-link.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, Input, OnDestroy, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { LuxActionComponentBaseClass } from '../lux-action-model/lux-action-component-base.class';
let LuxLinkComponent = class LuxLinkComponent extends LuxActionComponentBaseClass {
    constructor(router, cdr) {
        super();
        this.router = router;
        this.cdr = cdr;
        this.iconSize = '2x';
        this.luxBlank = false;
    }
    ngOnInit() { }
    ngOnDestroy() { }
    redirectToHref($event) {
        this.luxClicked.emit($event);
        if (this.luxHref) {
            this.luxHref = this.luxHref.trim();
            if (!this.luxHref.startsWith('http')) {
                this.router.navigate([this.luxHref]).then(() => { });
            }
            else {
                window.open(this.luxHref, this.luxBlank ? '_blank' : '_self');
            }
        }
    }
};
LuxLinkComponent.ctorParameters = () => [
    { type: Router },
    { type: ChangeDetectorRef }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxLinkComponent.prototype, "luxHref", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxLinkComponent.prototype, "luxBlank", void 0);
LuxLinkComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-link',
        template: "<lux-button\n  [luxRaised]=\"luxRaised\"\n  [luxColor]=\"luxColor\"\n  [luxLabel]=\"luxLabel\"\n  [luxIconName]=\"luxIconName\"\n  [luxIconAlignWithLabel]=\"luxIconAlignWithLabel\"\n  [luxDisabled]=\"luxDisabled\"\n  [luxRounded]=\"luxRounded\"\n  [luxTagId]=\"luxTagId\"\n  class=\"lux-link\"\n  luxAriaRole=\"link\"\n  luxAriaRoleSelector=\"button\"\n  (luxClicked)=\"redirectToHref($event)\"\n>\n</lux-button>\n",
        styles: [""]
    }),
    tslib_1.__metadata("design:paramtypes", [Router, ChangeDetectorRef])
], LuxLinkComponent);
export { LuxLinkComponent };
