// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-master-detail/lux-master-detail-subcomponents/lux-detail-view/lux-detail-wrapper.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
let LuxDetailWrapperComponent = class LuxDetailWrapperComponent {
    constructor() {
        this.luxDetailRendered = new EventEmitter();
    }
    set luxDetailTemplate(ref) {
        this._luxDetailTemplate = ref;
    }
    get luxDetailTemplate() {
        return this._luxDetailTemplate;
    }
    ngOnInit() { }
    ngAfterViewInit() {
        this.luxDetailRendered.emit();
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxDetailWrapperComponent.prototype, "luxDetailRendered", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], LuxDetailWrapperComponent.prototype, "luxDetailContext", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", TemplateRef),
    tslib_1.__metadata("design:paramtypes", [TemplateRef])
], LuxDetailWrapperComponent.prototype, "luxDetailTemplate", null);
LuxDetailWrapperComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-detail-wrapper',
        template: '<ng-container *ngTemplateOutlet="luxDetailTemplate; context: luxDetailContext"></ng-container>'
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxDetailWrapperComponent);
export { LuxDetailWrapperComponent };
