// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-file-preview/lux-file-preview-notsupportedviewer/lux-file-preview-notsupportedviewer.component.js
import * as tslib_1 from "tslib";
import { Component, Inject } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { LuxFilePreviewBase } from '../lux-file-preview-base/lux-file-preview-base';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
import { LuxFilePreviewRef } from '../lux-file-preview-ref';
let LuxFilePreviewNotSupportedViewerComponent = class LuxFilePreviewNotSupportedViewerComponent extends LuxFilePreviewBase {
    constructor(previewRef, previewData, sanitizer) {
        super(previewRef, previewData, sanitizer);
        this.previewRef = previewRef;
        this.previewData = previewData;
        this.sanitizer = sanitizer;
        this.counter = 5;
        this.downloadLabelDefault = 'Download... ';
        this.downloadLabel = this.downloadLabelDefault + this.counter;
    }
    ngOnInit() {
        this.updateDownloadLabel();
    }
    onDownload() {
        clearTimeout(this.timer);
        super.onDownload();
    }
    onClose() {
        clearTimeout(this.timer);
        super.onClose();
    }
    updateDownloadLabel() {
        this.timer = setTimeout(() => {
            if (this.counter > 0) {
                this.counter--;
                this.downloadLabel = this.downloadLabelDefault + this.counter;
                this.updateDownloadLabel();
            }
            else {
                this.onDownload();
            }
        }, 1000);
    }
};
LuxFilePreviewNotSupportedViewerComponent.ctorParameters = () => [
    { type: LuxFilePreviewRef },
    { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] },
    { type: DomSanitizer }
];
LuxFilePreviewNotSupportedViewerComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-file-preview-notsupportedviewer',
        template: "<lux-file-preview-toolbar>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-download\"\n    (luxClicked)=\"onDownload()\"\n    luxColor=\"accent\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-download-btn\"\n    luxAriaLabel=\"Datei herunterladen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-window-close\"\n    (luxClicked)=\"onClose()\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-close-btn\"\n    luxAriaLabel=\"Dateivorschau schlie\u00DFen\"\n  ></lux-button>\n</lux-file-preview-toolbar>\n<div class=\"message-box\" fxLayout=\"column\" fxLayoutAlign=\"center center\" fxLayoutGap=\"20px\">\n  <div>Diese Datei kann nicht angezeigt werden.</div>\n  <lux-button [luxLabel]=\"downloadLabel\" (luxClicked)=\"onDownload()\" [luxRaised]=\"true\" luxColor=\"accent\"></lux-button>\n</div>\n",
        styles: [""]
    }),
    tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
    tslib_1.__metadata("design:paramtypes", [LuxFilePreviewRef, Object, DomSanitizer])
], LuxFilePreviewNotSupportedViewerComponent);
export { LuxFilePreviewNotSupportedViewerComponent };
