// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-labelledby.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaLabelledbyDirective = class LuxAriaLabelledbyDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-labelledby');
        this.elementRef = elementRef;
        this.renderer = renderer;
        if (!this.luxAriaLabelledbySelector) {
            const tagName = elementRef.nativeElement.tagName.toLowerCase();
            if (tagName === 'lux-button') {
                this.luxAriaLabelledbySelector = 'button';
            }
            else if (tagName === 'lux-app-header-action-nav-item') {
                this.luxAriaLabelledbySelector = 'button';
            }
        }
    }
    get luxAriaLabelledby() {
        return this._luxAriaLabelledby;
    }
    set luxAriaLabelledby(labelledby) {
        this._luxAriaLabelledby = labelledby;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaLabelledbySelector;
    }
    getValue() {
        return this._luxAriaLabelledby;
    }
};
LuxAriaLabelledbyDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaLabelledbyDirective.prototype, "luxAriaLabelledbySelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaLabelledbyDirective.prototype, "luxAriaLabelledby", null);
LuxAriaLabelledbyDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaLabelledby]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaLabelledbyDirective);
export { LuxAriaLabelledbyDirective };
