// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-label.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaLabelDirective = class LuxAriaLabelDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-label');
        this.elementRef = elementRef;
        this.renderer = renderer;
        if (!this.luxAriaLabelSelector) {
            const tagName = elementRef.nativeElement.tagName.toLowerCase();
            if (tagName === 'lux-button') {
                this.luxAriaLabelSelector = 'button';
            }
            else if (tagName === 'lux-app-header-action-nav-item') {
                this.luxAriaLabelSelector = 'button';
            }
        }
    }
    get luxAriaLabel() {
        return this._luxAriaLabel;
    }
    set luxAriaLabel(label) {
        this._luxAriaLabel = label;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaLabelSelector;
    }
    getValue() {
        return this._luxAriaLabel;
    }
};
LuxAriaLabelDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaLabelDirective.prototype, "luxAriaLabelSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaLabelDirective.prototype, "luxAriaLabel", null);
LuxAriaLabelDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaLabel]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaLabelDirective);
export { LuxAriaLabelDirective };
