// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-required.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaRequiredDirective = class LuxAriaRequiredDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-required');
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    get luxAriaRequired() {
        return this._luxAriaRequired;
    }
    set luxAriaRequired(required) {
        this._luxAriaRequired = required;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaRequiredSelector;
    }
    getValue() {
        return this._luxAriaRequired;
    }
};
LuxAriaRequiredDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaRequiredDirective.prototype, "luxAriaRequiredSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaRequiredDirective.prototype, "luxAriaRequired", null);
LuxAriaRequiredDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaRequired]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaRequiredDirective);
export { LuxAriaRequiredDirective };
