// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-hidden.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaHiddenDirective = class LuxAriaHiddenDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-hidden');
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    get luxAriaHidden() {
        return this._luxAriaHidden;
    }
    set luxAriaHidden(hidden) {
        this._luxAriaHidden = hidden;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaHiddenSelector;
    }
    getValue() {
        return this._luxAriaHidden;
    }
};
LuxAriaHiddenDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaHiddenDirective.prototype, "luxAriaHiddenSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaHiddenDirective.prototype, "luxAriaHidden", null);
LuxAriaHiddenDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaHidden]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaHiddenDirective);
export { LuxAriaHiddenDirective };
