// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-role.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaRoleDirective = class LuxAriaRoleDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'role');
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    get luxAriaRole() {
        return this._luxAriaRole;
    }
    set luxAriaRole(role) {
        this._luxAriaRole = role;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaRoleSelector;
    }
    getValue() {
        return this._luxAriaRole;
    }
};
LuxAriaRoleDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaRoleDirective.prototype, "luxAriaRoleSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaRoleDirective.prototype, "luxAriaRole", null);
LuxAriaRoleDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaRole]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaRoleDirective);
export { LuxAriaRoleDirective };
