// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-expanded.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaExpandedDirective = class LuxAriaExpandedDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-expanded');
        this.elementRef = elementRef;
        this.renderer = renderer;
        if (!this.luxAriaExpandedSelector) {
            const tagName = elementRef.nativeElement.tagName.toLowerCase();
            if (tagName === 'lux-button') {
                this.luxAriaExpandedSelector = 'button';
            }
        }
    }
    get luxAriaExpanded() {
        return this._luxAriaExpanded;
    }
    set luxAriaExpanded(expanded) {
        this._luxAriaExpanded = expanded;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaExpandedSelector;
    }
    getValue() {
        return this._luxAriaExpanded;
    }
};
LuxAriaExpandedDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaExpandedDirective.prototype, "luxAriaExpandedSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaExpandedDirective.prototype, "luxAriaExpanded", null);
LuxAriaExpandedDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaExpanded]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaExpandedDirective);
export { LuxAriaExpandedDirective };
