// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-invalid.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaInvalidDirective = class LuxAriaInvalidDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-invalid');
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    get luxAriaInvalid() {
        return this._luxAriaInvalid;
    }
    set luxAriaInvalid(invalid) {
        this._luxAriaInvalid = invalid;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaInvalidSelector;
    }
    getValue() {
        return this._luxAriaInvalid;
    }
};
LuxAriaInvalidDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaInvalidDirective.prototype, "luxAriaInvalidSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaInvalidDirective.prototype, "luxAriaInvalid", null);
LuxAriaInvalidDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaInvalid]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaInvalidDirective);
export { LuxAriaInvalidDirective };
