// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-haspopup.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaHaspopupDirective = class LuxAriaHaspopupDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-haspopup');
        this.elementRef = elementRef;
        this.renderer = renderer;
        if (!this.luxAriaHasPopupSelector) {
            const tagName = elementRef.nativeElement.tagName.toLowerCase();
            if (tagName === 'lux-button') {
                this.luxAriaHasPopupSelector = 'button';
            }
            else if (tagName === 'lux-app-header-action-nav-item') {
                this.luxAriaHasPopupSelector = 'button';
            }
        }
    }
    get luxAriaHasPopup() {
        return this._luxAriaHasPopup;
    }
    set luxAriaHasPopup(hasPopup) {
        this._luxAriaHasPopup = hasPopup;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaHasPopupSelector;
    }
    getValue() {
        return this._luxAriaHasPopup;
    }
};
LuxAriaHaspopupDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaHaspopupDirective.prototype, "luxAriaHasPopupSelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaHaspopupDirective.prototype, "luxAriaHasPopup", null);
LuxAriaHaspopupDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaHasPopup]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaHaspopupDirective);
export { LuxAriaHaspopupDirective };
