// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-directives/lux-aria/lux-aria-describedby.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { LuxAriaBase } from './lux-aria-base';
let LuxAriaDescribedbyDirective = class LuxAriaDescribedbyDirective extends LuxAriaBase {
    constructor(elementRef, renderer) {
        super(elementRef, renderer, 'aria-describedby');
        this.elementRef = elementRef;
        this.renderer = renderer;
        if (!this.luxAriaDescribedbySelector) {
            const tagName = elementRef.nativeElement.tagName.toLowerCase();
            if (tagName === 'lux-button') {
                this.luxAriaDescribedbySelector = 'button';
            }
            else if (tagName === 'lux-app-header-action-nav-item') {
                this.luxAriaDescribedbySelector = 'button';
            }
        }
    }
    get luxAriaDescribedby() {
        return this._luxAriaDescribedby;
    }
    set luxAriaDescribedby(describedby) {
        this._luxAriaDescribedby = describedby;
        this.renderAria();
    }
    getSelector() {
        return this.luxAriaDescribedbySelector;
    }
    getValue() {
        return this._luxAriaDescribedby;
    }
};
LuxAriaDescribedbyDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAriaDescribedbyDirective.prototype, "luxAriaDescribedbySelector", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAriaDescribedbyDirective.prototype, "luxAriaDescribedby", null);
LuxAriaDescribedbyDirective = tslib_1.__decorate([
    Directive({
        selector: '[luxAriaDescribedby]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], LuxAriaDescribedbyDirective);
export { LuxAriaDescribedbyDirective };
