// @ihk-gfi/lux-components/esm5/src/app/modules/lux-directives/lux-tabindex/lux-tab-index.directive.js
import * as tslib_1 from "tslib";
import { AfterViewInit, Directive, ElementRef, Input, OnChanges, OnDestroy, Renderer2, SimpleChanges } from '@angular/core';
import { ReplaySubject } from 'rxjs';
var LuxTabIndexDirective = /** @class */ (function () {
    function LuxTabIndexDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.changes$ = new ReplaySubject(1);
        this.luxTabIndex = '0';
        this.luxApplyToParent = false;
        this.luxApplyToChildren = true;
        this.luxPotentialChildren = ['input', 'textarea', 'a', 'button', 'mat-select'];
    }
    LuxTabIndexDirective.prototype.ngOnChanges = function (simpleChanges) {
        this.changes$.next(simpleChanges);
    };
    LuxTabIndexDirective.prototype.ngAfterViewInit = function () {
        this.updateElementsWithTabIndex();
        // Dadurch verhindern wir, dass Changes vor afterViewInit zu etwaigen Fehlern führen könnten und erst danach aufgelöst werden
        this.changesSubscription = this.changes$.subscribe(this.onChanges.bind(this));
    };
    LuxTabIndexDirective.prototype.ngOnDestroy = function () {
        if (this.changesSubscription) {
            this.changesSubscription.unsubscribe();
        }
    };
    /**
     * Eigener Listener für Änderungen, der erst nach dem AfterViewInit-Lifecycle greift.
     * Sicherheitshalber, um evtl. noch unaufgelösten Elementen zu entgehen.
     * @param simpleChanges
     */
    LuxTabIndexDirective.prototype.onChanges = function (simpleChanges) {
        if (simpleChanges.luxTabIndex) {
            this.updateElementsWithTabIndex();
        }
    };
    /**
     * Aktualisiert den Tabindex für
     *  - das Zielelement, wenn luxApplyToParent === true (default = false) ist
     *  - potentielle Kindelemente, wenn luxApplyToChildren === true (default) ist
     */
    LuxTabIndexDirective.prototype.updateElementsWithTabIndex = function () {
        var _this = this;
        if (this.luxApplyToParent) {
            this.setTabIndexForElement(this.elementRef.nativeElement);
        }
        if (this.luxApplyToChildren) {
            this.luxPotentialChildren.forEach(function (childQuery) { return _this.setTabIndexByQuery(childQuery); });
        }
    };
    LuxTabIndexDirective.prototype.setTabIndexByQuery = function (elementQuery) {
        var elements = this.elementRef.nativeElement.querySelectorAll(elementQuery);
        for (var i = 0; i < elements.length; i++) {
            var element = elements[i];
            this.setTabIndexForElement(element);
        }
    };
    LuxTabIndexDirective.prototype.setTabIndexForElement = function (element) {
        this.renderer.setAttribute(element, 'tabIndex', this.luxTabIndex);
    };
    LuxTabIndexDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], LuxTabIndexDirective.prototype, "luxTabIndex", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTabIndexDirective.prototype, "luxApplyToParent", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean)
    ], LuxTabIndexDirective.prototype, "luxApplyToChildren", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], LuxTabIndexDirective.prototype, "luxPotentialChildren", void 0);
    LuxTabIndexDirective = tslib_1.__decorate([
        Directive({
            selector: '[luxTabIndex]'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], LuxTabIndexDirective);
    return LuxTabIndexDirective;
}());
export { LuxTabIndexDirective };
