// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-accordion/lux-accordion.component.js
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { MatAccordion } from '@angular/material';
import { Subject } from 'rxjs';
let LuxAccordionComponent = class LuxAccordionComponent {
    constructor() {
        this.changed$ = new Subject();
        this.luxMode = 'default';
        this.luxMulti = false;
        this._luxDisabled = false;
        this._luxHideToggle = false;
    }
    get luxDisabled() {
        return this._luxDisabled;
    }
    set luxDisabled(disabled) {
        this._luxDisabled = disabled;
        this.changed$.next('luxDisabled');
    }
    get luxHideToggle() {
        return this._luxHideToggle;
    }
    set luxHideToggle(hideToggle) {
        this._luxHideToggle = hideToggle;
        this.changed$.next('luxHideToggle');
    }
    get luxExpandedHeaderHeight() {
        return this._luxExpandedHeaderHeight;
    }
    set luxExpandedHeaderHeight(height) {
        this._luxExpandedHeaderHeight = height;
        this.changed$.next('luxExpandedHeaderHeight');
    }
    get luxCollapsedHeaderHeight() {
        return this._luxCollapsedHeaderHeight;
    }
    set luxCollapsedHeaderHeight(height) {
        this._luxCollapsedHeaderHeight = height;
        this.changed$.next('luxCollapsedHeaderHeight');
    }
    ngOnInit() { }
    ngOnDestroy() {
        this.changed$.complete();
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxAccordionComponent.prototype, "luxMode", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxAccordionComponent.prototype, "luxMulti", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], LuxAccordionComponent.prototype, "luxDisabled", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean),
    tslib_1.__metadata("design:paramtypes", [Boolean])
], LuxAccordionComponent.prototype, "luxHideToggle", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAccordionComponent.prototype, "luxExpandedHeaderHeight", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxAccordionComponent.prototype, "luxCollapsedHeaderHeight", null);
tslib_1.__decorate([
    ViewChild(MatAccordion, { static: true }),
    tslib_1.__metadata("design:type", MatAccordion)
], LuxAccordionComponent.prototype, "matAccordion", void 0);
LuxAccordionComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-accordion',
        template: "<mat-accordion [displayMode]=\"luxMode\" [multi]=\"luxMulti\">\n  <ng-content select=\"lux-panel\"></ng-content>\n</mat-accordion>\n"
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxAccordionComponent);
export { LuxAccordionComponent };
