// @ihk-gfi/lux-components/esm5/src/app/modules/lux-components-config/lux-components-config.module.js
import * as tslib_1 from "tslib";
import { InjectionToken, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { LuxComponentsHammerConfig } from './lux-components-hammer-config.class';
// Diesen bitte !nicht! injecten, sondern den LuxComponentsConfigService.
export var LuxConfigTokenService = new InjectionToken('luxConfig');
var LuxComponentsConfigModule = /** @class */ (function () {
    function LuxComponentsConfigModule() {
    }
    LuxComponentsConfigModule_1 = LuxComponentsConfigModule;
    // Den InjectionToken mit der übergebenen Konfiguration überschreiben, damit die Komponenten diese nutzen können
    LuxComponentsConfigModule.forRoot = function (config) {
        return {
            ngModule: LuxComponentsConfigModule_1,
            providers: [
                {
                    provide: LuxConfigTokenService,
                    useValue: config
                }
            ]
        };
    };
    var LuxComponentsConfigModule_1;
    LuxComponentsConfigModule = LuxComponentsConfigModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [],
            providers: [
                {
                    provide: HAMMER_GESTURE_CONFIG,
                    useClass: LuxComponentsHammerConfig
                }
            ]
        })
    ], LuxComponentsConfigModule);
    return LuxComponentsConfigModule;
}());
export { LuxComponentsConfigModule };
