// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-util/lux-media-query-observer.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { LuxConsoleService } from './lux-console.service';
import { BehaviorSubject } from 'rxjs';
import { BreakpointObserver, Breakpoints, BreakpointState } from '@angular/cdk/layout';
let LuxMediaQueryObserverService = class LuxMediaQueryObserverService {
    constructor(breakpointObserver, logger) {
        this.breakpointObserver = breakpointObserver;
        this.logger = logger;
        this._mediaQueryChanged = new BehaviorSubject('');
        this._subscriptions = [];
        this.addQuerySubscription(Breakpoints.XSmall, 'xs');
        this.addQuerySubscription(Breakpoints.Small, 'sm');
        this.addQuerySubscription(Breakpoints.Medium, 'md');
        this.addQuerySubscription(Breakpoints.Large, 'lg');
        this.addQuerySubscription(Breakpoints.XLarge, 'xl');
    }
    get activeMediaQuery() {
        return this._mediaQueryChanged.getValue();
    }
    ngOnDestroy() {
        this._subscriptions.forEach((subscription) => subscription.unsubscribe());
        this._mediaQueryChanged.complete();
    }
    getMediaQueryChangedAsObservable() {
        return this._mediaQueryChanged.asObservable();
    }
    isXS() {
        return this.activeMediaQuery === 'xs';
    }
    isSM() {
        return this.activeMediaQuery === 'sm';
    }
    isMD() {
        return this.activeMediaQuery === 'md';
    }
    isLG() {
        return this.activeMediaQuery === 'lg';
    }
    isXL() {
        return this.activeMediaQuery === 'xl';
    }
    addQuerySubscription(breakpoint, breakpointString) {
        this._subscriptions.push(this.breakpointObserver.observe([breakpoint]).subscribe((state) => {
            if (state.matches) {
                this._mediaQueryChanged.next(breakpointString);
                this.logger.log(`MediaQuery [${this.activeMediaQuery}] activated.`);
            }
        }));
    }
};
LuxMediaQueryObserverService.ctorParameters = () => [
    { type: BreakpointObserver },
    { type: LuxConsoleService }
];
LuxMediaQueryObserverService = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [BreakpointObserver, LuxConsoleService])
], LuxMediaQueryObserverService);
export { LuxMediaQueryObserverService };
