// @ihk-gfi/lux-components/esm5/src/app/modules/lux-popups/lux-snackbar/lux-snackbar.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { MatSnackBar, MatSnackBarDismiss } from '@angular/material';
import { LuxSnackbarComponent } from './lux-snackbar-component/lux-snackbar.component';
/**
 * Über den LuxSnackbarService können einfach Snackbarinfos angezeigt werden.
 */
var LuxSnackbarService = /** @class */ (function () {
    function LuxSnackbarService(snackBar) {
        this.snackBar = snackBar;
    }
    LuxSnackbarService_1 = LuxSnackbarService;
    /**
     * Diese Methode öffnet eine Snackbar mit einem Text.
     *
     * @param message Ein Text.
     * @param duration Eine Dauer in Msec (z.B. 2000 = 2 Sekunden). Wenn die Snackbar permanent angezeigt werden soll,
     * muss der Wert 0 angegeben werden.
     * @param actionName Eine Bezeichnung (z.B. Wiederherstellen) für die Action. Dieser Parameter ist optional und
     * muss nicht angegeben werden.
     */
    LuxSnackbarService.prototype.openText = function (message, duration, actionName) {
        this.snackBar.open(message, actionName, {
            duration: duration,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    };
    /**
     * Diese Methode öffnet eine Snackbar, in der die übergebene Komponente angezeigt wird.
     *
     * @param component Die Komponente, die angezeigt werden soll.
     * @param duration Eine Dauer in Msec (z.B. 2000 = 2 Sekunden). Wenn die Snackbar permanent angezeigt werden soll,
     * muss der Wert 0 angegeben werden.
     * @param data Ein Datenobjekt, das an die Komponente weitergereicht wird. Mit dem Code folgenden Code, können die
     * Daten verwendet werden. constructor(@Inject(MAT_SNACK_BAR_DATA) public data: any).
     */
    LuxSnackbarService.prototype.openComponent = function (component, duration, data) {
        if (duration === void 0) { duration = 0; }
        this.snackBar.openFromComponent(component, {
            duration: duration,
            data: data,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    };
    /**
     * Oeffnet eine Snackbar anhand der uebergebenen Konfiguration.
     * Ermoeglicht eine genaue Konfiguration der Snackbar.
     * @param config
     */
    LuxSnackbarService.prototype.open = function (duration, config) {
        this.snackBar.openFromComponent(LuxSnackbarComponent, {
            duration: duration,
            data: config,
            verticalPosition: LuxSnackbarService_1.VERTICAL_POSITION,
            horizontalPosition: LuxSnackbarService_1.HORIZONTAL_POSITION,
            panelClass: 'lux-snackbar'
        });
    };
    /**
     * Diese Methode liefert ein Observable zurück, das den Aufrufer benachrichtigt, wenn die Action in der Snackbar
     * geklickt wird.
     */
    LuxSnackbarService.prototype.onAction = function () {
        if (this.snackBar._openedSnackBarRef.instance instanceof LuxSnackbarComponent) {
            return this.snackBar._openedSnackBarRef.instance.onAction();
        }
        return this.snackBar._openedSnackBarRef.onAction();
    };
    /**
     * Diese Methode liefert ein Observable zurück, das den Aufrufer benachrichtigt, wenn die Action in der Snackbar
     * geklickt wird.
     */
    LuxSnackbarService.prototype.afterDismissed = function () {
        return this.snackBar._openedSnackBarRef.afterDismissed();
    };
    /**
     * Diese Methode blendet die Snackbar aus.
     */
    LuxSnackbarService.prototype.dismiss = function () {
        this.snackBar.dismiss();
    };
    var LuxSnackbarService_1;
    LuxSnackbarService.VERTICAL_POSITION = 'top';
    LuxSnackbarService.HORIZONTAL_POSITION = 'end';
    LuxSnackbarService.ctorParameters = function () { return [
        { type: MatSnackBar }
    ]; };
    LuxSnackbarService = LuxSnackbarService_1 = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [MatSnackBar])
    ], LuxSnackbarService);
    return LuxSnackbarService;
}());
export { LuxSnackbarService };
