// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-form/lux-form-model/lux-form-selectable-base.class.js
import * as tslib_1 from "tslib";
import { LuxFormComponentBase } from './lux-form-component-base.class';
import { EventEmitter, Input, Output } from '@angular/core';
/**
 * Basis-Klasse für FormComponents, die einen ähnlichen Grundaufbau für die Auswahl von
 * Optionen aus einem Array anbieten (Radio-Buttons und Selects z.B.).
 */
export class LuxFormSelectableBase extends LuxFormComponentBase {
    constructor(controlContainer, cdr, logger, config) {
        super(controlContainer, cdr, logger, config);
        this._luxOptions = [];
        this.luxSelectedChange = new EventEmitter();
        this.luxOptionLabelProp = '';
        this.luxCompareWith = (o1, o2) => o1 === o2;
        /**
         * Kapselung von der übergebenen luxCompareWith-Funktion.
         * Fängt undefinierte Objekte ab und returned stattdessen false.
         * @param o1
         * @param o2
         */
        this.compareObjects = (o1, o2) => {
            if ((!o1 && o2) || (o1 && !o2)) {
                return false;
            }
            else if (this.luxCompareWith && !!o1 && !!o2) {
                return this.luxCompareWith(o1, o2);
            }
            else {
                return o1 === o2;
            }
        };
    }
    get luxSelected() {
        return this.getValue();
    }
    set luxSelected(selected) {
        this.setValue(selected);
    }
    get luxOptions() {
        return this._luxOptions;
    }
    set luxOptions(options) {
        this._luxOptions = options;
        this.checkSelectedAndUpdate(this.luxSelected);
    }
    // region Overridden methods
    notifyFormValueChanged(formValue) {
        this.checkSelectedAndUpdate(formValue);
    }
    // endregion
    /**
     * Versucht, wenn Options und FormControl vorhanden sind, den selected-Wert mit den Options
     * zu vergleichen und dann auch wenn möglich als luxSelected-Wert zu sichern.
     * @param selected
     */
    checkSelectedAndUpdate(selected) {
        if (this.luxOptions && this.luxOptions.length > 0 && this.formControl) {
            if (this.luxPickValue && selected instanceof Object && !Array.isArray(selected)) {
                // Wenn der Wert zufälligerweise noch ein Objekt sein sollte, versuchen den Key auszulesen
                selected = this.luxPickValue(selected);
                // Da der Wert neu gesetzt wurde, diesen im nächsten Zyklus erst in die Werte schreiben
                setTimeout(() => {
                    this.checkSelectedAndUpdate(selected);
                });
            }
            else {
                // Für den Fall, das der eingesetzte Wert sich doch noch vom FormControl-Value unterscheidet,
                // diesen ergänzen
                if (this.luxSelected !== selected) {
                    this.luxSelected = selected;
                }
                this.checkSelectedInOptions(selected);
                setTimeout(() => {
                    this.luxSelectedChange.emit(selected);
                });
            }
        }
    }
    /**
     * Prüft ob der übergebene Wert in den luxOptions ist.
     * Kann auch ein Array als "selected" enthalten.
     * Wenn der Wert nicht gefunden werden konnte wird eine Fehlermeldung in der console ausgegeben.
     * @param selected
     */
    checkSelectedInOptions(selected) {
        const selectedAsArray = Array.isArray(selected) ? selected : [selected];
        if (selected && this.luxOptions && !this.allSelectedInOptions(selectedAsArray)) {
            // Selected nicht in Options = einen Fehler in die Console loggen
            this.logSelectedNotFound(selected);
        }
    }
    /**
     * Prüft ob die übergebenen Select-Objekte in den luxOptions enthalten sind.
     * @param selectedAsArray
     */
    allSelectedInOptions(selectedAsArray) {
        // Prüfen ob ein Unterarray von Elementen existiert.
        const optionsHasSubarray = this.luxOptions.length === 0
            ? false
            : // Dieses kann direkt in den Options stehen oder als value-Property
                Array.isArray(this.luxOptions[0]) ||
                    (!!this.luxPickValue && Array.isArray(this.luxPickValue(this.luxOptions[0])));
        const targetLength = optionsHasSubarray ? 1 : selectedAsArray.length;
        const length = this.luxOptions.filter((optionEntry) => {
            if (optionsHasSubarray) {
                return this.luxPickValue
                    ? this.luxPickValue(optionEntry) === selectedAsArray
                    : this.compareObjects(optionEntry, selectedAsArray);
            }
            else {
                return selectedAsArray.find((selectedEntry) => {
                    return this.luxPickValue
                        ? this.luxPickValue(optionEntry) === selectedEntry
                        : this.compareObjects(optionEntry, selectedEntry);
                });
            }
        }).length;
        return length === targetLength;
    }
    /**
     * Loggt die Fehlermeldung in die Console, wenn das neue Value-Objekt nicht in den Options gefunden
     * werden konnte.
     * @param selected
     */
    logSelectedNotFound(selected) {
        this.logger.error(`\n### Das Objekt ${JSON.stringify(selected)} ist nicht Teil der möglichen Optionen.\n` +
            `\n### Komponente: "${this}"\n` +
            `Prüfen Sie evtl. die luxCompareWith-Funktion, um Properties anstelle ganzer Objekte zu vergleichen.`);
    }
    triggerOutputPatternCheck() {
        this.checkOutputPatternViolation(this.luxSelectedChange.observers);
    }
    triggerInputPatternCheck(simpleChanges) {
        this.checkInputPatternViolation(simpleChanges.luxSelected);
    }
}
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxFormSelectableBase.prototype, "luxSelectedChange", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxFormSelectableBase.prototype, "luxOptionLabelProp", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxFormSelectableBase.prototype, "luxTagId", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Function)
], LuxFormSelectableBase.prototype, "luxPickValue", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxFormSelectableBase.prototype, "luxReadonly", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], LuxFormSelectableBase.prototype, "luxCompareWith", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], LuxFormSelectableBase.prototype, "luxSelected", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array),
    tslib_1.__metadata("design:paramtypes", [Array])
], LuxFormSelectableBase.prototype, "luxOptions", null);
