// @ihk-gfi/lux-components/esm5/src/app/modules/lux-error/lux-error-page/lux-error-services/lux-error-store.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Dieser Service dient dazu, die aktuellen Fehlermeldungen und die Error-Page Konfiguration zu speichern.
 *
 * Er wird von LuxErrorService und LuxErrorPageComponent benutzt, ohne ihn gäbe es Cycle-Dependency Fehler.
 */
var LuxErrorStoreService = /** @class */ (function () {
    function LuxErrorStoreService() {
        /**
         * Enthaelt die normale Konfiguration der Fehlerseite, kann bei Bedarf mit setConfig ueberschrieben werden.
         */
        this._defaultConfig = {
            iconName: 'far fa-times-circle',
            iconSize: '5x',
            errorText: 'Uups... da ist etwas schief gelaufen. Wir kennen die Fehlerdetails bereits und kümmern uns darum.',
            homeRedirectText: 'Zurück zur Startseite',
            homeRedirectUrl: '',
            errorPageUrl: 'errorpage',
            skipLocationChange: true
        };
    }
    Object.defineProperty(LuxErrorStoreService.prototype, "config", {
        /**
         * Gibt die aktuelle Konfiguration zurueck.
         * @returns ILuxErrorPageConfig
         */
        get: function () {
            return this._config;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxErrorStoreService.prototype, "lastErrors", {
        /**
         * Gibt ein Array der letzten Fehler zurueck.
         * @returns Array<ILuxError[]>
         */
        get: function () {
            return Array.of(this._lastErrors);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxErrorStoreService.prototype, "error", {
        /**
         * Gibt den aktuellen Fehler zurueck.
         * @returns ILuxError
         */
        get: function () {
            return this._error;
        },
        /**
         * Sichert den Fehler in der Property und fügt ihn der "lastErrors"-Liste hinzu.
         * @param newError
         */
        set: function (newError) {
            this._error = newError;
            this._lastErrors.push(newError);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initialisiert den Service.
     */
    LuxErrorStoreService.prototype.init = function () {
        this._lastErrors = [];
        this._error = null;
        this._config = {};
        this.safeNewConfig(this._defaultConfig);
    };
    /**
     *
     * @param luxErrorPageConfig
     */
    LuxErrorStoreService.prototype.safeNewConfig = function (luxErrorPageConfig) {
        var _this = this;
        // wenn keine Config übergeben wird, ein leeres Objekt erzeugen
        // das sorgt dafür, dass wenigstens die default-Werte genommen werden
        luxErrorPageConfig = luxErrorPageConfig ? luxErrorPageConfig : {};
        // alle Werte der übergebenen Config prüfen und ggf. default-Werte nutzen
        Object.keys(luxErrorPageConfig).forEach(function (e) {
            if (luxErrorPageConfig[e] !== undefined && luxErrorPageConfig[e] !== null) {
                _this._config[e] = luxErrorPageConfig[e];
            }
            else {
                _this._config[e] = _this._defaultConfig[e];
            }
        });
    };
    LuxErrorStoreService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LuxErrorStoreService_Factory() { return new LuxErrorStoreService(); }, token: LuxErrorStoreService, providedIn: "root" });
    LuxErrorStoreService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxErrorStoreService);
    return LuxErrorStoreService;
}());
export { LuxErrorStoreService };
