// @ihk-gfi/lux-components/esm5/src/app/modules/lux-error/lux-http-error/lux-http-error.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { NavigationStart, Router } from '@angular/router';
import { LuxHttpErrorInterceptor } from './lux-http-error-interceptor';
import { LuxBackgroundColorsEnum } from '../../lux-util/lux-colors.enum';
var LuxHttpErrorComponent = /** @class */ (function () {
    function LuxHttpErrorComponent(router) {
        this.subs = [];
        // Beim Ansteuern einer neuen Route, die aktuellen Fehler resetten.
        router.events.forEach(function (event) {
            if (event instanceof NavigationStart) {
                LuxHttpErrorInterceptor.dataStream.next([]);
            }
        });
    }
    LuxHttpErrorComponent.prototype.ngOnInit = function () {
        this.errors = [];
    };
    LuxHttpErrorComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        // Wenn neue Fehler-Objekte kommen, diese umformatieren und in der LuxMessageBoxComponent anzeigen.
        this.subs.push(LuxHttpErrorInterceptor.dataStream$().subscribe(function (errors) {
            _this.updateErrors(errors);
        }));
    };
    LuxHttpErrorComponent.prototype.ngOnDestroy = function () {
        // Alle bekannten Subscriptions auflösen.
        this.subs.forEach(function (sub) {
            sub.unsubscribe();
        });
    };
    /**
     * Updated das aktuelle Errors-Array mit dem übergebenen Wert.
     * Liest dabei die Message aus dem Fehler und erzeugt LuxMessage-Objekte für die LuxMessageBoxComponent.
     * @param errors
     */
    LuxHttpErrorComponent.prototype.updateErrors = function (errors) {
        var _this = this;
        var errorMessages = [];
        if (errors && errors.length > 0) {
            errors.forEach(function (error) {
                errorMessages.push({
                    text: _this.readErrorMessage(error),
                    color: LuxBackgroundColorsEnum.red,
                    iconName: 'fa-bug'
                });
            });
        }
        setTimeout(function () {
            _this.errors = errorMessages;
        });
    };
    /**
     * Versucht eine Fehlermeldung aus dem Fehler zu lesen.
     * Zuerst wird geschaut, ob der Fehler selbst nur ein String ist ==> Wenn ja, diesen zurückgeben.
     * Dann wird geprüft, ob der Fehler eine "message"-Property besitzt ==> Wenn ja, diese zurückgeben
     * Als letzter Ausweg wird das "error"-Objekt selbst über die toString-Methode zurückgegeben.
     * @param error
     */
    LuxHttpErrorComponent.prototype.readErrorMessage = function (error) {
        if (typeof error === 'string') {
            return error;
        }
        else if (error.hasOwnProperty('message')) {
            return error['message'];
        }
        else {
            return error.toString();
        }
    };
    LuxHttpErrorComponent.ctorParameters = function () { return [
        { type: Router }
    ]; };
    LuxHttpErrorComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-http-error',
            template: "<lux-message-box [luxMessages]=\"errors\"> </lux-message-box>\n",
            styles: [""]
        }),
        tslib_1.__metadata("design:paramtypes", [Router])
    ], LuxHttpErrorComponent);
    return LuxHttpErrorComponent;
}());
export { LuxHttpErrorComponent };
