// @ihk-gfi/lux-components/esm5/src/app/modules/lux-directives/lux-aria/lux-aria-base.js
var LuxAriaBase = /** @class */ (function () {
    function LuxAriaBase(elementRef, renderer, ariaTagName) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.ariaTagName = ariaTagName;
        this.init = false;
    }
    LuxAriaBase.prototype.ngAfterViewInit = function () {
        this.init = true;
        this.renderAria();
    };
    LuxAriaBase.prototype.renderAria = function () {
        if (this.init) {
            var selector = this.getSelector();
            var el = void 0;
            if (selector) {
                el = this.elementRef.nativeElement.querySelector(selector);
            }
            else {
                el = this.elementRef.nativeElement;
            }
            if (el) {
                var value = this.getValue();
                if (value === null || value === undefined) {
                    this.renderer.removeAttribute(el, this.ariaTagName);
                }
                else {
                    this.renderer.setAttribute(el, this.ariaTagName, value);
                }
            }
        }
    };
    return LuxAriaBase;
}());
export { LuxAriaBase };
