// @ihk-gfi/lux-components/esm5/src/app/modules/lux-form/lux-datepicker/lux-datepicker-adapter.js
import * as tslib_1 from "tslib";
import { NativeDateAdapter } from '@angular/material';
import { Injectable } from '@angular/core';
import { LuxUtil } from '../../lux-util/lux-util';
var LuxDatepickerAdapter = /** @class */ (function (_super) {
    tslib_1.__extends(LuxDatepickerAdapter, _super);
    function LuxDatepickerAdapter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        // dd.MM.yyyy
        _this.dotRegExp = new RegExp(/\d{1,2}\.\d{1,2}\.\d{4}/);
        // MM/dd/yyyy
        _this.backslashRegExp = new RegExp(/\d{1,2}\/\d{1,2}\/\d{4}/);
        // dd-MM-yyyy
        _this.hyphenRegExp = new RegExp(/\d{1,2}-\d{1,2}-\d{4}/);
        // yyyy-MM-dd
        _this.hyphenRegExp_1 = new RegExp(/\d{4}-\d{1,2}-\d{1,2}/);
        return _this;
    }
    LuxDatepickerAdapter.prototype.format = function (date, displayFormat) {
        if (date) {
            if (displayFormat) {
                if (typeof date === 'string') {
                    date = new Date(date);
                }
                return date.toLocaleDateString(this.locale, displayFormat);
            }
            else {
                return date.toLocaleDateString(this.locale);
            }
        }
        else {
            return '';
        }
    };
    LuxDatepickerAdapter.prototype.parse = function (value) {
        if (value) {
            // Pr端fen, ob der Wert ein ISO-String ist
            if (LuxUtil.ISO_8601_FULL.test(value)) {
                return new Date(value);
            }
            // Hat der String das Format dd.MM.YYYY ?
            if (this.dotRegExp.test(value)) {
                return this.getUTCNulled_ddMMYYYY(value, '.');
            }
            else if (this.backslashRegExp.test(value)) {
                return this.getUTCNulled_MMddYYY(value, '/');
            }
            else if (this.hyphenRegExp.test(value)) {
                return this.getUTCNulled_ddMMYYYY(value, '-');
            }
            else if (this.hyphenRegExp_1.test(value)) {
                return this.getUTCNulled_YYYYMMdd(value, '-');
            }
            return value;
        }
        return null;
    };
    /**
     * UTC Date mit 0-Werten f端r Time aus einem ddMMYYYY-String erhalten.
     * @param dateString
     * @param separator
     */
    LuxDatepickerAdapter.prototype.getUTCNulled_ddMMYYYY = function (dateString, separator) {
        var splitDate = dateString.split(separator);
        var tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[2], +splitDate[1] - 1, +splitDate[0]);
        return tempDate;
    };
    /**
     * UTC Date mit 0-Werten f端r Time aus einem YYYYMMdd-String erhalten.
     * @param dateString
     * @param separator
     */
    LuxDatepickerAdapter.prototype.getUTCNulled_YYYYMMdd = function (dateString, separator) {
        var splitDate = dateString.split(separator);
        var tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[0], +splitDate[1] - 1, +splitDate[2]);
        return tempDate;
    };
    /**
     * UTC Date mit 0-Werten f端r Time aus einem MMddYYYY-String erhalten.
     * @param dateString
     * @param separator
     */
    LuxDatepickerAdapter.prototype.getUTCNulled_MMddYYY = function (dateString, separator) {
        var splitDate = dateString.split(separator);
        var tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[2], +splitDate[0] - 1, +splitDate[1]);
        return tempDate;
    };
    LuxDatepickerAdapter.prototype.isValid = function (date) {
        return LuxUtil.isDate(date);
    };
    LuxDatepickerAdapter = tslib_1.__decorate([
        Injectable()
    ], LuxDatepickerAdapter);
    return LuxDatepickerAdapter;
}(NativeDateAdapter));
export { LuxDatepickerAdapter };
