// @ihk-gfi/lux-components/esm5/src/app/modules/lux-file-preview/lux-file-preview-toolbar/lux-file-preview-toolbar.component.js
import * as tslib_1 from "tslib";
import { Component, Inject } from '@angular/core';
import { LuxMediaQueryObserverService } from '../../lux-util/lux-media-query-observer.service';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
var LuxFilePreviewToolbarComponent = /** @class */ (function () {
    function LuxFilePreviewToolbarComponent(mediaQueryService, data) {
        this.mediaQueryService = mediaQueryService;
        this.data = data;
    }
    LuxFilePreviewToolbarComponent.ctorParameters = function () { return [
        { type: LuxMediaQueryObserverService },
        { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] }
    ]; };
    LuxFilePreviewToolbarComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-file-preview-toolbar',
            template: "<div fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n  <div fxFlex=\"calc(100% - 350px)\" fxLayout=\"row\" fxLayoutAlign=\"start center\">\n    <div class=\"lux-crop\" *ngIf=\"!mediaQueryService.isXS()\">\n      <lux-icon luxIconName=\"fa-file\"></lux-icon>\n      {{ data.fileObject.name }}\n    </div>\n  </div>\n  <div fxFlex=\"none\" fxLayout=\"row\" fxLayoutGap=\"5px\" fxLayoutAlign=\"end center\">\n    <ng-content></ng-content>\n  </div>\n</div>\n",
            styles: [":host{position:absolute;top:0;left:0;right:0;color:#fff;z-index:1;padding:8px}:host ::ng-deep lux-button button{padding:5px!important;min-width:45px!important}:host ::ng-deep lux-select lux-form-control div{background-color:#fff;outline-width:0!important;outline-style:none}:host ::ng-deep lux-select lux-form-control div div.lux-form-control-label{display:none}:host ::ng-deep .lux-button-disabled button{background-color:rgba(128,128,128,.5)!important}"]
        }),
        tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
        tslib_1.__metadata("design:paramtypes", [LuxMediaQueryObserverService, Object])
    ], LuxFilePreviewToolbarComponent);
    return LuxFilePreviewToolbarComponent;
}());
export { LuxFilePreviewToolbarComponent };
