// @ihk-gfi/lux-components/esm5/src/app/modules/lux-lookup/lux-lookup-service/lux-lookup-handler.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
var LuxLookupHandlerService = /** @class */ (function () {
    function LuxLookupHandlerService() {
        this.lookupElements = new Map();
    }
    /**
     * Fügt ein LookupComponent zu der Map hinzu.
     * @param string name
     */
    LuxLookupHandlerService.prototype.addLookupElement = function (name) {
        this.lookupElements.set(name, new Subject());
    };
    /**
     * Gibt das Subject einer LookupComponent als Observable zurück (oder null).
     * @param string name
     * @returns Observable<any> | null
     */
    LuxLookupHandlerService.prototype.getLookupElementObsv = function (name) {
        var lookupEl = this.getLookupElementSubject(name);
        if (lookupEl) {
            return lookupEl.asObservable();
        }
        return null;
    };
    /**
     * Gibt das Subject einer LookupComponent zurück (oder null).
     * @param string name
     * @returns Observable<any> | null
     */
    LuxLookupHandlerService.prototype.getLookupElementSubject = function (name) {
        var lookupEl = this.lookupElements.get(name);
        if (lookupEl) {
            return lookupEl;
        }
        return null;
    };
    /**
     * Stößt das Neuladen von Schlüsseltabellendaten einer LookupComponent an.
     * @param string name
     */
    LuxLookupHandlerService.prototype.reloadData = function (lookupId) {
        var lookupEl = this.getLookupElementSubject(lookupId);
        if (lookupEl) {
            lookupEl.next();
        }
        else {
            console.error('Die Lookup-Komponente ' + lookupId + ' konnte nicht gefunden werden.');
        }
    };
    LuxLookupHandlerService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxLookupHandlerService);
    return LuxLookupHandlerService;
}());
export { LuxLookupHandlerService };
