// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-components-config/lux-components-config.module.js
import * as tslib_1 from "tslib";
var LuxComponentsConfigModule_1;
import { InjectionToken, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { LuxComponentsHammerConfig } from './lux-components-hammer-config.class';
// Diesen bitte !nicht! injecten, sondern den LuxComponentsConfigService.
export const LuxConfigTokenService = new InjectionToken('luxConfig');
let LuxComponentsConfigModule = LuxComponentsConfigModule_1 = class LuxComponentsConfigModule {
    // Den InjectionToken mit der übergebenen Konfiguration überschreiben, damit die Komponenten diese nutzen können
    static forRoot(config) {
        return {
            ngModule: LuxComponentsConfigModule_1,
            providers: [
                {
                    provide: LuxConfigTokenService,
                    useValue: config
                }
            ]
        };
    }
};
LuxComponentsConfigModule = LuxComponentsConfigModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [],
        providers: [
            {
                provide: HAMMER_GESTURE_CONFIG,
                useClass: LuxComponentsHammerConfig
            }
        ]
    })
], LuxComponentsConfigModule);
export { LuxComponentsConfigModule };
