// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-form/lux-datepicker/lux-datepicker-adapter.js
import * as tslib_1 from "tslib";
import { NativeDateAdapter } from '@angular/material';
import { Injectable } from '@angular/core';
import { LuxUtil } from '../../lux-util/lux-util';
let LuxDatepickerAdapter = class LuxDatepickerAdapter extends NativeDateAdapter {
    constructor() {
        super(...arguments);
        // dd.MM.yyyy
        this.dotRegExp = new RegExp(/\d{1,2}\.\d{1,2}\.\d{4}/);
        // MM/dd/yyyy
        this.backslashRegExp = new RegExp(/\d{1,2}\/\d{1,2}\/\d{4}/);
        // dd-MM-yyyy
        this.hyphenRegExp = new RegExp(/\d{1,2}-\d{1,2}-\d{4}/);
        // yyyy-MM-dd
        this.hyphenRegExp_1 = new RegExp(/\d{4}-\d{1,2}-\d{1,2}/);
    }
    format(date, displayFormat) {
        if (date) {
            if (displayFormat) {
                if (typeof date === 'string') {
                    date = new Date(date);
                }
                return date.toLocaleDateString(this.locale, displayFormat);
            }
            else {
                return date.toLocaleDateString(this.locale);
            }
        }
        else {
            return '';
        }
    }
    parse(value) {
        if (value) {
            // Pr端fen, ob der Wert ein ISO-String ist
            if (LuxUtil.ISO_8601_FULL.test(value)) {
                return new Date(value);
            }
            // Hat der String das Format dd.MM.YYYY ?
            if (this.dotRegExp.test(value)) {
                return this.getUTCNulled_ddMMYYYY(value, '.');
            }
            else if (this.backslashRegExp.test(value)) {
                return this.getUTCNulled_MMddYYY(value, '/');
            }
            else if (this.hyphenRegExp.test(value)) {
                return this.getUTCNulled_ddMMYYYY(value, '-');
            }
            else if (this.hyphenRegExp_1.test(value)) {
                return this.getUTCNulled_YYYYMMdd(value, '-');
            }
            return value;
        }
        return null;
    }
    /**
     * UTC Date mit 0-Werten f端r Time aus einem ddMMYYYY-String erhalten.
     * @param dateString
     * @param separator
     */
    getUTCNulled_ddMMYYYY(dateString, separator) {
        const splitDate = dateString.split(separator);
        const tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[2], +splitDate[1] - 1, +splitDate[0]);
        return tempDate;
    }
    /**
     * UTC Date mit 0-Werten f端r Time aus einem YYYYMMdd-String erhalten.
     * @param dateString
     * @param separator
     */
    getUTCNulled_YYYYMMdd(dateString, separator) {
        const splitDate = dateString.split(separator);
        const tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[0], +splitDate[1] - 1, +splitDate[2]);
        return tempDate;
    }
    /**
     * UTC Date mit 0-Werten f端r Time aus einem MMddYYYY-String erhalten.
     * @param dateString
     * @param separator
     */
    getUTCNulled_MMddYYY(dateString, separator) {
        const splitDate = dateString.split(separator);
        const tempDate = new Date(0);
        tempDate.setUTCFullYear(+splitDate[2], +splitDate[0] - 1, +splitDate[1]);
        return tempDate;
    }
    isValid(date) {
        return LuxUtil.isDate(date);
    }
};
LuxDatepickerAdapter = tslib_1.__decorate([
    Injectable()
], LuxDatepickerAdapter);
export { LuxDatepickerAdapter };
