// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-popups/lux-dialog/lux-dialog.service.js
import * as tslib_1 from "tslib";
var LuxDialogService_1;
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material';
import { LuxConsoleService } from '../../lux-util/lux-console.service';
import { LuxDialogRef } from './lux-dialog-model/lux-dialog-ref.class';
import { DEFAULT_DIALOG_CONF } from './lux-dialog-model/lux-dialog-config.interface';
import { DEFAULT_DIALOG_PRESET_CONF } from './lux-dialog-model/lux-dialog-preset-config.interface';
import { LuxDialogPresetComponent } from './lux-dialog-preset/lux-dialog-preset.component';
let LuxDialogService = LuxDialogService_1 = class LuxDialogService {
    constructor(matDialog, logger, luxDialogRef) {
        this.matDialog = matDialog;
        this.logger = logger;
        this.luxDialogRef = luxDialogRef;
        this.dialogOpened = false;
    }
    /**
     * Öffnet einen Dialog basierend auf der übergebenen Component und den entsprechenden Daten.
     * @param component
     * @param config
     * @param data
     */
    openComponent(component, config, data) {
        this.handleOpen(component, config, data, DEFAULT_DIALOG_CONF);
        return this.luxDialogRef;
    }
    /**
     * Öffnet einen Dialog basierend auf der LuxDialogComponent und der LuxDialogConfig.
     * @param config
     */
    open(config) {
        // Eine Dialog-Instanz erzeugen, als Data übergeben wir hier noch einmal die Config
        this.handleOpen(LuxDialogPresetComponent, config, config, DEFAULT_DIALOG_PRESET_CONF);
        return this.luxDialogRef;
    }
    /**
     * Prüft ob bereits ein Dialog geöffnet ist und etwaige CSS-Klassen für den Dialog gegeben sind.
     * Anschließend wird der Dialog mit den übergebenen Config-Optionen und Data-Informationen geöffnet.
     * @param component
     * @param config
     * @param data
     * @param defaultConfig
     */
    handleOpen(component, config, data, defaultConfig) {
        if (this.dialogOpened) {
            this.logger.error(LuxDialogService_1.ALREADY_OPENED_ERROR);
            return null;
        }
        // Wenn keine Config übergeben ist, die defaultConfig nehmen
        config = config ? config : defaultConfig;
        // Die CSS-Klassen fürs Panel herausfinden
        const panelClass = ['lux-dialog'];
        if (config.panelClass) {
            if (Array.isArray(config.panelClass)) {
                panelClass.push(...config.panelClass);
            }
            else {
                panelClass.push(config.panelClass);
            }
        }
        // Dialog öffnen und Konfiguration übergeben
        const matDialogRef = this.matDialog.open(component, {
            width: config.width,
            height: config.height,
            autoFocus: false,
            restoreFocus: true,
            disableClose: config.disableClose,
            panelClass: panelClass
        });
        this.luxDialogRef.init(matDialogRef, data);
    }
};
LuxDialogService.ALREADY_OPENED_ERROR = 'Aktuell ist bereits ein Dialog geöffnet';
LuxDialogService.ctorParameters = () => [
    { type: MatDialog },
    { type: LuxConsoleService },
    { type: LuxDialogRef }
];
LuxDialogService = LuxDialogService_1 = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [MatDialog, LuxConsoleService, LuxDialogRef])
], LuxDialogService);
export { LuxDialogService };
