// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-master-detail/lux-master-detail-subcomponents/lux-detail-view/lux-detail-wrapper.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
var LuxDetailWrapperComponent = /** @class */ (function () {
    function LuxDetailWrapperComponent() {
        this.luxDetailRendered = new EventEmitter();
    }
    Object.defineProperty(LuxDetailWrapperComponent.prototype, "luxDetailTemplate", {
        get: function () {
            return this._luxDetailTemplate;
        },
        set: function (ref) {
            this._luxDetailTemplate = ref;
        },
        enumerable: true,
        configurable: true
    });
    LuxDetailWrapperComponent.prototype.ngOnInit = function () { };
    LuxDetailWrapperComponent.prototype.ngAfterViewInit = function () {
        this.luxDetailRendered.emit();
    };
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], LuxDetailWrapperComponent.prototype, "luxDetailRendered", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], LuxDetailWrapperComponent.prototype, "luxDetailContext", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", TemplateRef),
        tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], LuxDetailWrapperComponent.prototype, "luxDetailTemplate", null);
    LuxDetailWrapperComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-detail-wrapper',
            template: '<ng-container *ngTemplateOutlet="luxDetailTemplate; context: luxDetailContext"></ng-container>'
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], LuxDetailWrapperComponent);
    return LuxDetailWrapperComponent;
}());
export { LuxDetailWrapperComponent };
