// @ihk-gfi/lux-components/esm5/src/app/modules/lux-file-preview/lux-file-preview.service.js
import * as tslib_1 from "tslib";
import { Overlay, OverlayConfig, OverlayRef } from '@angular/cdk/overlay';
import { ComponentPortal, PortalInjector } from '@angular/cdk/portal';
import { ComponentRef, Injectable, Injector } from '@angular/core';
import { LUX_FILE_PREVIEW_DATA } from './lux-file-preview-config';
import { LuxFilePreviewRef } from './lux-file-preview-ref';
import { LuxFilePreviewComponent } from './lux-file-preview.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
export var DEFAULT_FILE_PREVIEW_CONFIG = {
    hasBackdrop: true,
    backdropClass: 'lux-file-preview-backdrop',
    panelClass: 'lux-file-preview-panel',
    previewData: {
        fileComponent: null,
        fileObject: null
    }
};
var LuxFilePreviewService = /** @class */ (function () {
    function LuxFilePreviewService(injector, overlay) {
        this.injector = injector;
        this.overlay = overlay;
    }
    LuxFilePreviewService.prototype.open = function (config) {
        var previewConfig = tslib_1.__assign({}, DEFAULT_FILE_PREVIEW_CONFIG, config);
        var overlayRef = this.createOverlay(previewConfig);
        var previewRef = new LuxFilePreviewRef(overlayRef);
        var previewComponent = this.attachDialogContainer(overlayRef, previewRef, previewConfig);
        overlayRef.backdropClick().subscribe(function () { return previewComponent.onClose(); });
        return previewRef;
    };
    LuxFilePreviewService.prototype.attachDialogContainer = function (overlayRef, dialogRef, config) {
        var injector = this.createInjector(config, dialogRef);
        var containerPortal = new ComponentPortal(LuxFilePreviewComponent, null, injector);
        var containerRef = overlayRef.attach(containerPortal);
        return containerRef.instance;
    };
    LuxFilePreviewService.prototype.createOverlay = function (config) {
        return this.overlay.create(this.getOverlayConfig(config));
    };
    LuxFilePreviewService.prototype.createInjector = function (config, dialogRef) {
        var injectionTokens = new WeakMap();
        injectionTokens.set(LuxFilePreviewRef, dialogRef);
        injectionTokens.set(LUX_FILE_PREVIEW_DATA, config.previewData);
        return new PortalInjector(this.injector, injectionTokens);
    };
    LuxFilePreviewService.prototype.getOverlayConfig = function (config) {
        var positionStrategy = this.overlay
            .position()
            .global()
            .centerHorizontally()
            .centerVertically();
        var overlayConfig = new OverlayConfig({
            hasBackdrop: config.hasBackdrop,
            backdropClass: config.backdropClass,
            panelClass: config.panelClass,
            scrollStrategy: this.overlay.scrollStrategies.block(),
            positionStrategy: positionStrategy
        });
        return overlayConfig;
    };
    LuxFilePreviewService.ctorParameters = function () { return [
        { type: Injector },
        { type: Overlay }
    ]; };
    LuxFilePreviewService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LuxFilePreviewService_Factory() { return new LuxFilePreviewService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Overlay)); }, token: LuxFilePreviewService, providedIn: "root" });
    LuxFilePreviewService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        }),
        tslib_1.__metadata("design:paramtypes", [Injector, Overlay])
    ], LuxFilePreviewService);
    return LuxFilePreviewService;
}());
export { LuxFilePreviewService };
