// @ihk-gfi/lux-components/esm5/src/app/modules/lux-file-preview/lux-file-preview.component.js
import * as tslib_1 from "tslib";
import { Component, Inject, ViewChild } from '@angular/core';
import { LUX_FILE_PREVIEW_DATA } from './lux-file-preview-config';
import { LuxFilePreviewImgViewerComponent } from './lux-file-preview-imgviewer/lux-file-preview-imgviewer.component';
import { LuxFilePreviewNotSupportedViewerComponent } from './lux-file-preview-notsupportedviewer/lux-file-preview-notsupportedviewer.component';
import { LuxFilePreviewPdfViewerComponent } from './lux-file-preview-pdfviewer/lux-file-preview-pdfviewer.component';
var LuxFilePreviewComponent = /** @class */ (function () {
    function LuxFilePreviewComponent(data) {
        this.data = data;
        this.fileType = 'notsupported';
    }
    LuxFilePreviewComponent.prototype.ngOnInit = function () {
        if (this.data.fileObject.type.indexOf('image/') > -1) {
            this.fileType = 'img';
        }
        else if (this.data.fileObject.type.indexOf('application/pdf') > -1) {
            this.fileType = 'pdf';
        }
        else {
            this.fileType = 'notsupported';
        }
    };
    LuxFilePreviewComponent.prototype.onClose = function () {
        if (this.pdfViewer) {
            this.pdfViewer.onClose();
        }
        else if (this.imgViewer) {
            this.imgViewer.onClose();
        }
        else {
            this.notSupportedViewer.onClose();
        }
    };
    LuxFilePreviewComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] }
    ]; };
    tslib_1.__decorate([
        ViewChild(LuxFilePreviewPdfViewerComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxFilePreviewPdfViewerComponent)
    ], LuxFilePreviewComponent.prototype, "pdfViewer", void 0);
    tslib_1.__decorate([
        ViewChild(LuxFilePreviewImgViewerComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxFilePreviewImgViewerComponent)
    ], LuxFilePreviewComponent.prototype, "imgViewer", void 0);
    tslib_1.__decorate([
        ViewChild(LuxFilePreviewNotSupportedViewerComponent, { static: false }),
        tslib_1.__metadata("design:type", LuxFilePreviewNotSupportedViewerComponent)
    ], LuxFilePreviewComponent.prototype, "notSupportedViewer", void 0);
    LuxFilePreviewComponent = tslib_1.__decorate([
        Component({
            selector: 'lux-file-preview',
            template: "<lux-file-preview-pdfviewer *ngIf=\"fileType === 'pdf'\"></lux-file-preview-pdfviewer>\n<lux-file-preview-imgviewer *ngIf=\"fileType === 'img'\"></lux-file-preview-imgviewer>\n<lux-file-preview-notsupportedviewer *ngIf=\"fileType === 'notsupported'\"></lux-file-preview-notsupportedviewer>\n",
            styles: [":host ::ng-deep .message-box{border-radius:8px;background-color:#fff;padding:20px}"]
        }),
        tslib_1.__param(0, Inject(LUX_FILE_PREVIEW_DATA)),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], LuxFilePreviewComponent);
    return LuxFilePreviewComponent;
}());
export { LuxFilePreviewComponent };
