// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-master-detail/lux-master-detail-mobile-helper.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { LuxMediaQueryObserverService } from '../../lux-util/lux-media-query-observer.service';
let LuxMasterDetailMobileHelperService = class LuxMasterDetailMobileHelperService {
    constructor(mediaObserver) {
        this.mediaObserver = mediaObserver;
        this._mdRegistered$ = new BehaviorSubject(false);
        this._masterCollapsed$ = new BehaviorSubject(true);
        this._hasValue$ = new BehaviorSubject(false);
        this.mediaObserver.getMediaQueryChangedAsObservable().subscribe(() => {
            setTimeout(() => {
                this.handleMasterCollapseState();
            });
        });
    }
    get masterCollapsed$() {
        return this._masterCollapsed$.getValue();
    }
    get masterCollapsedObservable() {
        return this._masterCollapsed$.asObservable();
    }
    get isRegisteredObservable() {
        return this._mdRegistered$.asObservable();
    }
    get isRegistered() {
        return this._mdRegistered$.getValue();
    }
    get hasValueObservable() {
        return this._hasValue$.asObservable();
    }
    get hasValue() {
        return this._hasValue$.getValue();
    }
    set hasValue(hasValue) {
        this._hasValue$.next(hasValue);
    }
    /**
     * Sendet ueber den masterCollapsed$-Subject ein Event mit dem Wert true.
     * Kennzeichnet das der Master geoeffnet werden soll.
     */
    openMaster() {
        this._masterCollapsed$.next(true);
    }
    /**
     * Sendet ueber den masterCollapsed$-Subject ein Event mit dem Wert false.
     * Kennzeichnet das der Master geschlossen werden soll.
     */
    closeMaster() {
        this._masterCollapsed$.next(false);
    }
    /**
     * Gibt zurueck ob eine Media-Query fuer XS oder SM aktuell aktiv ist.
     * @returns boolean
     */
    isMobile() {
        return this.mediaObserver.isXS() || this.mediaObserver.isSM();
    }
    /**
     * Meldet eine Master-Detail-Komponente fuer diesen Service an.
     */
    register() {
        if (!this.isRegistered) {
            this._mdRegistered$.next(true);
        }
        else {
            console.error('Fehler: Ein Master-Detail-Helper-Service kann nur eine registrierte Master-Detail-Komponente haben.');
        }
    }
    /**
     * Meldet die aktuelle Master-Detail-Komponente dieses Services ab.
     */
    unregister() {
        if (this.isRegistered) {
            this._mdRegistered$.next(false);
        }
    }
    /**
     * Sendet ein Event um den Master zu oeffnen bzw. zu schliessen
     * wenn sich der Zustand der Media-Query geaendert hat.
     */
    handleMasterCollapseState() {
        setTimeout(() => {
            if (this.hasValue && (this.mediaObserver.isXS() || this.mediaObserver.isSM())) {
                this.closeMaster();
            }
            else {
                this.openMaster();
            }
        });
    }
};
LuxMasterDetailMobileHelperService.ctorParameters = () => [
    { type: LuxMediaQueryObserverService }
];
LuxMasterDetailMobileHelperService = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [LuxMediaQueryObserverService])
], LuxMasterDetailMobileHelperService);
export { LuxMasterDetailMobileHelperService };
