// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-popups/lux-dialog/lux-dialog-model/lux-dialog-ref.class.js
import * as tslib_1 from "tslib";
import { ReplaySubject } from 'rxjs';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
let LuxDialogRef = class LuxDialogRef {
    constructor() {
        this._dialogConfirmed = new ReplaySubject(1);
        this._dialogDeclined = new ReplaySubject(1);
        this._dialogClosed = new ReplaySubject(1);
    }
    /**
     * Gibt die Component, die in dem Dialog angezeigt wird wieder.
     */
    get componentInstance() {
        return this._matDialogRef.componentInstance;
    }
    /**
     * Damit hier ein Wert abgegeben wird, muss "closeDialog" ein Result mit dem Wert "true" übergeben bekommen.
     */
    get dialogConfirmed() {
        return this._dialogConfirmed.asObservable();
    }
    /**
     * Damit hier ein Wert abgegeben wird, muss "closeDialog" ein Result mit dem Wert "false" übergeben bekommen.
     */
    get dialogDeclined() {
        return this._dialogDeclined.asObservable();
    }
    /**
     * Wird bei Beenden des Dialogs ausgelöst.
     */
    get dialogClosed() {
        return this._dialogClosed.asObservable();
    }
    /**
     * Enthält die Daten für die Dialog-Component.
     */
    get data() {
        return this._data;
    }
    /**
     * (Re-)Initialisiert diese Dialog-Referenz neu.
     * @param matDialogRef
     * @param data
     */
    init(matDialogRef, data) {
        this._matDialogRef = matDialogRef;
        this._dialogConfirmed = new ReplaySubject(1);
        this._dialogDeclined = new ReplaySubject(1);
        this._dialogClosed = new ReplaySubject(1);
        this._data = data;
    }
    /**
     * Schließt den Dialog und gibt nach Beenden die entsprechenden Events aus.
     * @param result
     */
    closeDialog(result) {
        this._matDialogRef.close(result);
        this._matDialogRef.afterClosed().subscribe((dialogResult) => {
            if (dialogResult === true) {
                this._dialogConfirmed.next();
            }
            else if (dialogResult === false) {
                this._dialogDeclined.next();
            }
            this._dialogClosed.next(result);
        });
    }
};
LuxDialogRef.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LuxDialogRef_Factory() { return new LuxDialogRef(); }, token: LuxDialogRef, providedIn: "root" });
LuxDialogRef = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxDialogRef);
export { LuxDialogRef };
