// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-error/lux-http-error/lux-http-error.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { NavigationStart, Router } from '@angular/router';
import { LuxHttpErrorInterceptor } from './lux-http-error-interceptor';
import { LuxBackgroundColorsEnum } from '../../lux-util/lux-colors.enum';
let LuxHttpErrorComponent = class LuxHttpErrorComponent {
    constructor(router) {
        this.subs = [];
        // Beim Ansteuern einer neuen Route, die aktuellen Fehler resetten.
        router.events.forEach(event => {
            if (event instanceof NavigationStart) {
                LuxHttpErrorInterceptor.dataStream.next([]);
            }
        });
    }
    ngOnInit() {
        this.errors = [];
    }
    ngAfterViewInit() {
        // Wenn neue Fehler-Objekte kommen, diese umformatieren und in der LuxMessageBoxComponent anzeigen.
        this.subs.push(LuxHttpErrorInterceptor.dataStream$().subscribe((errors) => {
            this.updateErrors(errors);
        }));
    }
    ngOnDestroy() {
        // Alle bekannten Subscriptions auflösen.
        this.subs.forEach(sub => {
            sub.unsubscribe();
        });
    }
    /**
     * Updated das aktuelle Errors-Array mit dem übergebenen Wert.
     * Liest dabei die Message aus dem Fehler und erzeugt LuxMessage-Objekte für die LuxMessageBoxComponent.
     * @param errors
     */
    updateErrors(errors) {
        const errorMessages = [];
        if (errors && errors.length > 0) {
            errors.forEach((error) => {
                errorMessages.push({
                    text: this.readErrorMessage(error),
                    color: LuxBackgroundColorsEnum.red,
                    iconName: 'fa-bug'
                });
            });
        }
        setTimeout(() => {
            this.errors = errorMessages;
        });
    }
    /**
     * Versucht eine Fehlermeldung aus dem Fehler zu lesen.
     * Zuerst wird geschaut, ob der Fehler selbst nur ein String ist ==> Wenn ja, diesen zurückgeben.
     * Dann wird geprüft, ob der Fehler eine "message"-Property besitzt ==> Wenn ja, diese zurückgeben
     * Als letzter Ausweg wird das "error"-Objekt selbst über die toString-Methode zurückgegeben.
     * @param error
     */
    readErrorMessage(error) {
        if (typeof error === 'string') {
            return error;
        }
        else if (error.hasOwnProperty('message')) {
            return error['message'];
        }
        else {
            return error.toString();
        }
    }
};
LuxHttpErrorComponent.ctorParameters = () => [
    { type: Router }
];
LuxHttpErrorComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-http-error',
        template: "<lux-message-box [luxMessages]=\"errors\"> </lux-message-box>\n",
        styles: [""]
    }),
    tslib_1.__metadata("design:paramtypes", [Router])
], LuxHttpErrorComponent);
export { LuxHttpErrorComponent };
