// @ihk-gfi/lux-components/esm5/src/app/modules/lux-util/lux-media-query-observer.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { LuxConsoleService } from './lux-console.service';
import { BehaviorSubject } from 'rxjs';
import { BreakpointObserver, Breakpoints, BreakpointState } from '@angular/cdk/layout';
var LuxMediaQueryObserverService = /** @class */ (function () {
    function LuxMediaQueryObserverService(breakpointObserver, logger) {
        this.breakpointObserver = breakpointObserver;
        this.logger = logger;
        this._mediaQueryChanged = new BehaviorSubject('');
        this._subscriptions = [];
        this.addQuerySubscription(Breakpoints.XSmall, 'xs');
        this.addQuerySubscription(Breakpoints.Small, 'sm');
        this.addQuerySubscription(Breakpoints.Medium, 'md');
        this.addQuerySubscription(Breakpoints.Large, 'lg');
        this.addQuerySubscription(Breakpoints.XLarge, 'xl');
    }
    Object.defineProperty(LuxMediaQueryObserverService.prototype, "activeMediaQuery", {
        get: function () {
            return this._mediaQueryChanged.getValue();
        },
        enumerable: true,
        configurable: true
    });
    LuxMediaQueryObserverService.prototype.ngOnDestroy = function () {
        this._subscriptions.forEach(function (subscription) { return subscription.unsubscribe(); });
        this._mediaQueryChanged.complete();
    };
    LuxMediaQueryObserverService.prototype.getMediaQueryChangedAsObservable = function () {
        return this._mediaQueryChanged.asObservable();
    };
    LuxMediaQueryObserverService.prototype.isXS = function () {
        return this.activeMediaQuery === 'xs';
    };
    LuxMediaQueryObserverService.prototype.isSM = function () {
        return this.activeMediaQuery === 'sm';
    };
    LuxMediaQueryObserverService.prototype.isMD = function () {
        return this.activeMediaQuery === 'md';
    };
    LuxMediaQueryObserverService.prototype.isLG = function () {
        return this.activeMediaQuery === 'lg';
    };
    LuxMediaQueryObserverService.prototype.isXL = function () {
        return this.activeMediaQuery === 'xl';
    };
    LuxMediaQueryObserverService.prototype.addQuerySubscription = function (breakpoint, breakpointString) {
        var _this = this;
        this._subscriptions.push(this.breakpointObserver.observe([breakpoint]).subscribe(function (state) {
            if (state.matches) {
                _this._mediaQueryChanged.next(breakpointString);
                _this.logger.log("MediaQuery [" + _this.activeMediaQuery + "] activated.");
            }
        }));
    };
    LuxMediaQueryObserverService.ctorParameters = function () { return [
        { type: BreakpointObserver },
        { type: LuxConsoleService }
    ]; };
    LuxMediaQueryObserverService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [BreakpointObserver, LuxConsoleService])
    ], LuxMediaQueryObserverService);
    return LuxMediaQueryObserverService;
}());
export { LuxMediaQueryObserverService };
