// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-file-preview/lux-file-preview-pdfviewer/lux-file-preview-pdfviewer.component.js
import * as tslib_1 from "tslib";
import { Component, Inject } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { PdfViewerComponent } from 'ng2-pdf-viewer';
import { LuxFilePreviewBase } from '../lux-file-preview-base/lux-file-preview-base';
import { LUX_FILE_PREVIEW_DATA } from '../lux-file-preview-config';
import { LuxFilePreviewRef } from '../lux-file-preview-ref';
let LuxFilePreviewPdfViewerComponent = class LuxFilePreviewPdfViewerComponent extends LuxFilePreviewBase {
    constructor(previewRef, previewData, sanitizer) {
        super(previewRef, previewData, sanitizer);
        this.previewRef = previewRef;
        this.previewData = previewData;
        this.sanitizer = sanitizer;
        this.showAll = true;
        this.page = 1;
        this.numPages = 0;
        this.loadingDivWidth = 100;
        this.loadingDivHeight = 50;
        this.options = [
            { label: ' 15%', value: 0.15 },
            { label: ' 25%', value: 0.25 },
            { label: ' 50%', value: 0.5 },
            { label: ' 75%', value: 0.75 },
            { label: '100%', value: 1.0 },
            { label: '150%', value: 1.5 },
            { label: '200%', value: 2.0 }
        ];
        this.zoom = this.options[4];
    }
    ngOnInit() {
        super.ngOnInit();
        this.loadingDivLeft = window.innerWidth / 2 - this.loadingDivWidth / 2;
        this.loadingDivTop = window.innerHeight / 2 - this.loadingDivHeight / 2;
    }
    onPrevPage() {
        this.page--;
        this.clearFocus();
    }
    onNextPage() {
        this.page++;
        this.clearFocus();
    }
    pageRendered(e) {
        if (e && e.pageNumber === 1) {
            this.loadingFinished();
        }
    }
    callBackFn(pdf) {
        this.numPages = pdf.numPages;
        pdf.getPage(1).then((page) => {
            let viewportWidth = 0;
            let currentIndex = 0;
            while (viewportWidth === 0 || viewportWidth > this.width) {
                viewportWidth =
                    page.getViewport({
                        scale: this.zoom.value
                    }).width * PdfViewerComponent.CSS_UNITS;
                currentIndex = this.options.findIndex(currentZoom => currentZoom === this.zoom);
                if (currentIndex === 0) {
                    break;
                }
                if (viewportWidth > this.width) {
                    this.zoom = this.options[Math.max(currentIndex - 1, 0)];
                }
            }
        });
    }
    onError(event) {
        console.log(event);
    }
    onProgress(progressData) { }
};
LuxFilePreviewPdfViewerComponent.ctorParameters = () => [
    { type: LuxFilePreviewRef },
    { type: undefined, decorators: [{ type: Inject, args: [LUX_FILE_PREVIEW_DATA,] }] },
    { type: DomSanitizer }
];
LuxFilePreviewPdfViewerComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-file-preview-pdfviewer',
        template: "<lux-file-preview-toolbar>\n  <lux-select\n    fxFlex=\"80px\"\n    [luxOptions]=\"options\"\n    [(luxSelected)]=\"zoom\"\n    luxOptionLabelProp=\"label\"\n    luxAriaLabel=\"Zoom\"\n  ></lux-select>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-arrow-left\"\n    *ngIf=\"!showAll\"\n    (luxClicked)=\"onPrevPage()\"\n    [luxDisabled]=\"page === 1\"\n    [ngClass]=\"{ 'lux-button-disabled': page === 1 }\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-zoomin-btn\"\n    luxAriaLabel=\"Hinein zoomen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    luxIconName=\"fas fa-arrow-right\"\n    (luxClicked)=\"onNextPage()\"\n    [luxDisabled]=\"page === numPages\"\n    [ngClass]=\"{ 'lux-button-disabled': page === numPages }\"\n    *ngIf=\"!showAll\"\n    luxRaised=\"true\"\n    luxTagId=\"file-preview-zoomout-btn\"\n    luxAriaLabel=\"Heraus zoomen\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    [luxIconName]=\"downloadIconName\"\n    (luxClicked)=\"onDownload()\"\n    luxColor=\"accent\"\n    luxRaised=\"true\"\n    [luxTagId]=\"downloadTagId\"\n    [luxAriaLabel]=\"downloadAriaLabel\"\n  ></lux-button>\n  <lux-button\n    fxFlex=\"none\"\n    [luxIconName]=\"closeIconName\"\n    (luxClicked)=\"onClose()\"\n    luxRaised=\"true\"\n    [luxTagId]=\"closeTagId\"\n    [luxAriaLabel]=\"closeAriaLabel\"\n  ></lux-button>\n</lux-file-preview-toolbar>\n\n<div class=\"overlay-content\" [ngStyle]=\"{ width: width + 'px', height: height + 'px' }\">\n  <div\n    fxLayout=\"column\"\n    fxLayoutAlign=\"center center\"\n    *ngIf=\"loading\"\n    class=\"message-box message-box-pdf\"\n    [ngStyle]=\"{\n      left: this.loadingDivLeft + 'px',\n      top: this.loadingDivTop + 'px',\n      width: loadingDivWidth + 'px',\n      height: loadingDivHeight + 'px'\n    }\"\n  >\n    <p>Lade...</p>\n  </div>\n  <pdf-viewer\n    *ngIf=\"url\"\n    [src]=\"url\"\n    [show-all]=\"showAll\"\n    [page]=\"page\"\n    [zoom]=\"zoom.value\"\n    [render-text]=\"false\"\n    [c-maps-url]=\"'assets/cmaps/'\"\n    (page-rendered)=\"pageRendered($event)\"\n    (after-load-complete)=\"callBackFn($event)\"\n    (error)=\"onError($event)\"\n    (on-progress)=\"onProgress($event)\"\n    style=\"display: block;\"\n  ></pdf-viewer>\n</div>\n",
        styles: [".overlay-content{overflow:auto;background-color:gray;padding:8px;border-radius:4px}pdf-viewer{height:100%}.message-box-pdf{position:absolute;overflow:hidden}"]
    }),
    tslib_1.__param(1, Inject(LUX_FILE_PREVIEW_DATA)),
    tslib_1.__metadata("design:paramtypes", [LuxFilePreviewRef, Object, DomSanitizer])
], LuxFilePreviewPdfViewerComponent);
export { LuxFilePreviewPdfViewerComponent };
