// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-app-footer/lux-app-footer-button-info.js
var LuxAppFooterButtonInfo = /** @class */ (function () {
    function LuxAppFooterButtonInfo(label, cmd, color, disabled, hidden, raised, iconName, alwaysVisible, onClick) {
        this.label = label;
        this.color = color;
        this.disabled = disabled;
        this.cmd = cmd;
        this.hidden = hidden === true ? true : false;
        this.raised = raised === undefined || raised === null || raised === true ? true : false;
        this.iconName = iconName;
        this.alwaysVisible =
            alwaysVisible === undefined || alwaysVisible === null || alwaysVisible === false ? false : true;
        this.onClick = onClick ? onClick : function (that) { };
    }
    /**
     * Statische Methode um ein Info-Objekt zu generieren.
     * Nimmt ein Objekt vom Typ ILuxAppFooterButtonInfo entgegen.
     * @param data
     * @returns eine Button Info
     */
    LuxAppFooterButtonInfo.generateInfo = function (data) {
        var info = new LuxAppFooterButtonInfo(data.label, data.cmd, data.color, data.disabled, data.hidden, data.raised, data.iconName, data.alwaysVisible, data.onClick);
        return info;
    };
    return LuxAppFooterButtonInfo;
}());
export { LuxAppFooterButtonInfo };
