// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-app-footer/lux-app-footer-button.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var LuxAppFooterButtonService = /** @class */ (function () {
    function LuxAppFooterButtonService() {
        this._buttonInfos = [];
    }
    Object.defineProperty(LuxAppFooterButtonService.prototype, "buttonInfos", {
        get: function () {
            return this._buttonInfos;
        },
        set: function (buttonInfos) {
            this._buttonInfos = buttonInfos ? buttonInfos : [];
        },
        enumerable: true,
        configurable: true
    });
    LuxAppFooterButtonService.prototype.pushButtonInfos = function () {
        var _a;
        var value = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            value[_i] = arguments[_i];
        }
        if (!this.buttonInfos) {
            this._buttonInfos = [];
        }
        (_a = this._buttonInfos).push.apply(_a, tslib_1.__spread(value));
    };
    LuxAppFooterButtonService.prototype.getButtonInfoByCMD = function (cmd) {
        return this.buttonInfos.find(function (buttonInfo) { return buttonInfo.cmd === cmd; });
    };
    LuxAppFooterButtonService.prototype.removeButtonInfoAtIndex = function (i) {
        this._buttonInfos = this._buttonInfos.filter(function (info, index) { return index !== i; });
    };
    LuxAppFooterButtonService.prototype.removeButtonInfoByCmd = function (cmd) {
        this._buttonInfos = this._buttonInfos.filter(function (info) { return info.cmd !== cmd; });
    };
    LuxAppFooterButtonService.prototype.clearButtonInfos = function () {
        this._buttonInfos = [];
    };
    LuxAppFooterButtonService.prototype.sendButtonCommand = function (buttonCommand) {
        // Den angeklickten Button erhalten und dessen onClick-Funktion aufrufen.
        // Als Parameter den Button selbst mitgeben (falls dieser bearbeitet werden muss, z.B.).
        var clickedButton = this.getButtonInfoByCMD(buttonCommand);
        if (clickedButton) {
            clickedButton.onClick(clickedButton);
        }
    };
    LuxAppFooterButtonService = tslib_1.__decorate([
        Injectable()
    ], LuxAppFooterButtonService);
    return LuxAppFooterButtonService;
}());
export { LuxAppFooterButtonService };
