// @ihk-gfi/lux-components/esm5/src/app/modules/lux-layout/lux-master-detail/lux-master-detail-mobile-helper.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { LuxMediaQueryObserverService } from '../../lux-util/lux-media-query-observer.service';
var LuxMasterDetailMobileHelperService = /** @class */ (function () {
    function LuxMasterDetailMobileHelperService(mediaObserver) {
        var _this = this;
        this.mediaObserver = mediaObserver;
        this._mdRegistered$ = new BehaviorSubject(false);
        this._masterCollapsed$ = new BehaviorSubject(true);
        this._hasValue$ = new BehaviorSubject(false);
        this.mediaObserver.getMediaQueryChangedAsObservable().subscribe(function () {
            setTimeout(function () {
                _this.handleMasterCollapseState();
            });
        });
    }
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "masterCollapsed$", {
        get: function () {
            return this._masterCollapsed$.getValue();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "masterCollapsedObservable", {
        get: function () {
            return this._masterCollapsed$.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "isRegisteredObservable", {
        get: function () {
            return this._mdRegistered$.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "isRegistered", {
        get: function () {
            return this._mdRegistered$.getValue();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "hasValueObservable", {
        get: function () {
            return this._hasValue$.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LuxMasterDetailMobileHelperService.prototype, "hasValue", {
        get: function () {
            return this._hasValue$.getValue();
        },
        set: function (hasValue) {
            this._hasValue$.next(hasValue);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Sendet ueber den masterCollapsed$-Subject ein Event mit dem Wert true.
     * Kennzeichnet das der Master geoeffnet werden soll.
     */
    LuxMasterDetailMobileHelperService.prototype.openMaster = function () {
        this._masterCollapsed$.next(true);
    };
    /**
     * Sendet ueber den masterCollapsed$-Subject ein Event mit dem Wert false.
     * Kennzeichnet das der Master geschlossen werden soll.
     */
    LuxMasterDetailMobileHelperService.prototype.closeMaster = function () {
        this._masterCollapsed$.next(false);
    };
    /**
     * Gibt zurueck ob eine Media-Query fuer XS oder SM aktuell aktiv ist.
     * @returns boolean
     */
    LuxMasterDetailMobileHelperService.prototype.isMobile = function () {
        return this.mediaObserver.isXS() || this.mediaObserver.isSM();
    };
    /**
     * Meldet eine Master-Detail-Komponente fuer diesen Service an.
     */
    LuxMasterDetailMobileHelperService.prototype.register = function () {
        if (!this.isRegistered) {
            this._mdRegistered$.next(true);
        }
        else {
            console.error('Fehler: Ein Master-Detail-Helper-Service kann nur eine registrierte Master-Detail-Komponente haben.');
        }
    };
    /**
     * Meldet die aktuelle Master-Detail-Komponente dieses Services ab.
     */
    LuxMasterDetailMobileHelperService.prototype.unregister = function () {
        if (this.isRegistered) {
            this._mdRegistered$.next(false);
        }
    };
    /**
     * Sendet ein Event um den Master zu oeffnen bzw. zu schliessen
     * wenn sich der Zustand der Media-Query geaendert hat.
     */
    LuxMasterDetailMobileHelperService.prototype.handleMasterCollapseState = function () {
        var _this = this;
        setTimeout(function () {
            if (_this.hasValue && (_this.mediaObserver.isXS() || _this.mediaObserver.isSM())) {
                _this.closeMaster();
            }
            else {
                _this.openMaster();
            }
        });
    };
    LuxMasterDetailMobileHelperService.ctorParameters = function () { return [
        { type: LuxMediaQueryObserverService }
    ]; };
    LuxMasterDetailMobileHelperService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [LuxMediaQueryObserverService])
    ], LuxMasterDetailMobileHelperService);
    return LuxMasterDetailMobileHelperService;
}());
export { LuxMasterDetailMobileHelperService };
