// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-form/lux-chips/lux-chips-subcomponents/lux-chip-group.component.js
import * as tslib_1 from "tslib";
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
let LuxChipGroupComponent = class LuxChipGroupComponent {
    constructor() {
        this._luxColor = 'primary';
        this.luxChipClicked = new EventEmitter();
        this.luxChipAdded = new EventEmitter();
        this.luxChipRemoved = new EventEmitter();
        this.luxChipSelected = new EventEmitter();
        this.luxLabels = [];
        this.luxDisabled = false;
        this.luxRemovable = true;
        this.luxSelected = true;
    }
    get luxColor() {
        return this._luxColor;
    }
    set luxColor(color) {
        if (color !== 'primary' && color !== 'accent' && color !== 'warn') {
            color = undefined;
        }
        this._luxColor = color;
    }
    ngOnInit() { }
    add(label) {
        this.luxLabels.push(label);
        this.luxChipAdded.emit(label);
    }
    remove(index) {
        this.luxChipRemoved.emit(index);
        this.luxLabels.splice(index, 1);
    }
    click(index) {
        this.luxChipClicked.emit(index);
    }
    select(selected, index) {
        this.luxChipSelected.emit({ index, selected });
    }
};
tslib_1.__decorate([
    ContentChild(TemplateRef, { static: false }),
    tslib_1.__metadata("design:type", TemplateRef)
], LuxChipGroupComponent.prototype, "tempRef", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], LuxChipGroupComponent.prototype, "luxChipClicked", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], LuxChipGroupComponent.prototype, "luxChipAdded", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], LuxChipGroupComponent.prototype, "luxChipRemoved", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], LuxChipGroupComponent.prototype, "luxChipSelected", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], LuxChipGroupComponent.prototype, "luxLabels", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxChipGroupComponent.prototype, "luxDisabled", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxChipGroupComponent.prototype, "luxRemovable", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxChipGroupComponent.prototype, "luxSelected", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], LuxChipGroupComponent.prototype, "luxColor", null);
LuxChipGroupComponent = tslib_1.__decorate([
    Component({
        selector: 'lux-chip-group',
        template: ''
    }),
    tslib_1.__metadata("design:paramtypes", [])
], LuxChipGroupComponent);
export { LuxChipGroupComponent };
