// @ihk-gfi/lux-components/esm2015/src/app/modules/lux-layout/lux-app-header/lux-app-header-subcomponents/lux-side-nav/lux-side-nav-subcomponents/lux-side-nav-item.component.js
import * as tslib_1 from "tslib";
var LuxSideNavItemComponent_1;
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChildren, EventEmitter, Input, OnChanges, Output, QueryList, TemplateRef, ViewChild } from '@angular/core';
let LuxSideNavItemComponent = LuxSideNavItemComponent_1 = class LuxSideNavItemComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.luxDisabled = false;
        this.luxSelected = false;
        this.luxCloseOnClick = true;
        this.luxExpandable = false;
        this.luxExpanded = true;
        this.luxClicked = new EventEmitter();
    }
    ngOnChanges() {
        // Bei Input Änderungen die CD anstossen
        this.cdr.detectChanges();
    }
    ngAfterViewInit() {
        // Nach Abschluss der Initialisierung die CD anstossen
        this.cdr.detectChanges();
    }
    onClick($event) {
        this.luxClicked.emit($event);
        if (this.luxExpandable) {
            this.luxExpanded = !this.luxExpanded;
        }
    }
};
LuxSideNavItemComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
tslib_1.__decorate([
    ViewChild(TemplateRef, { static: false }),
    tslib_1.__metadata("design:type", TemplateRef)
], LuxSideNavItemComponent.prototype, "templateRef", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxSideNavItemComponent.prototype, "luxLabel", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxSideNavItemComponent.prototype, "luxDisabled", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxSideNavItemComponent.prototype, "luxTagId", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxSideNavItemComponent.prototype, "luxSelected", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxSideNavItemComponent.prototype, "luxCloseOnClick", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], LuxSideNavItemComponent.prototype, "luxIconName", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxSideNavItemComponent.prototype, "luxExpandable", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], LuxSideNavItemComponent.prototype, "luxExpanded", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], LuxSideNavItemComponent.prototype, "luxClicked", void 0);
tslib_1.__decorate([
    ContentChildren(LuxSideNavItemComponent_1),
    tslib_1.__metadata("design:type", QueryList)
], LuxSideNavItemComponent.prototype, "sideNavItems", void 0);
LuxSideNavItemComponent = LuxSideNavItemComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'lux-side-nav-item',
        template: "<ng-template>\n  <div class=\"lux-side-nav-item-container\">\n    <div class=\"lux-side-nav-item-button\" [ngClass]=\"{ 'lux-side-nav-item-expandable': luxExpandable }\">\n      <lux-button\n        (luxClicked)=\"onClick($event)\"\n        [luxLabel]=\"luxLabel\"\n        [luxIconName]=\"luxIconName\"\n        [luxDisabled]=\"luxDisabled\"\n        [luxTagId]=\"luxTagId\"\n        class=\"lux-side-nav-item\"\n        [ngClass]=\"{ 'lux-side-nav-item-selected': luxSelected }\"\n      >\n      </lux-button>\n      <ng-container *ngIf=\"luxExpandable\">\n        <lux-icon\n          class=\"lux-side-nav-item-expansion-icon\"\n          [luxIconName]=\"luxExpanded ? 'fas fa-chevron-up' : 'fas fa-chevron-down'\"\n        >\n        </lux-icon>\n      </ng-container>\n    </div>\n    <ng-container *ngIf=\"luxExpanded\">\n      <ng-container *ngFor=\"let sideNavItem of sideNavItems; let i = index\">\n        <div class=\"lux-side-nav-sub-item\" *ngIf=\"i > 0\">\n          <ng-container [ngTemplateOutlet]=\"sideNavItem.templateRef\"></ng-container>\n        </div>\n      </ng-container>\n    </ng-container>\n  </div>\n  <ng-content select=\"lux-divider\"></ng-content>\n</ng-template>\n<ng-content select=\"lux-side-nav-item\"></ng-content>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
], LuxSideNavItemComponent);
export { LuxSideNavItemComponent };
